// @caiu/library/esm5/lib/forms/actions.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var FormActions = /** @class */ (function () {
    function FormActions() {
    }
    /**
     * @param {?} control
     * @return {?}
     */
    FormActions.addForm = /**
     * @param {?} control
     * @return {?}
     */
    function (control) {
        return {
            type: FormActions.ADD,
            payload: control
        };
    };
    /**
     * @param {?} control
     * @param {?} changes
     * @return {?}
     */
    FormActions.changeStatus = /**
     * @param {?} control
     * @param {?} changes
     * @return {?}
     */
    function (control, changes) {
        return {
            type: FormActions.STATUS_CHANGES,
            payload: {
                control: control,
                changes: changes
            }
        };
    };
    /**
     * @param {?} control
     * @param {?} changes
     * @return {?}
     */
    FormActions.changeValue = /**
     * @param {?} control
     * @param {?} changes
     * @return {?}
     */
    function (control, changes) {
        return {
            type: FormActions.VALUE_CHANGES,
            payload: {
                control: control,
                changes: changes
            }
        };
    };
    /**
     * @param {?} route
     * @return {?}
     */
    FormActions.navigateToForm = /**
     * @param {?} route
     * @return {?}
     */
    function (route) {
        return {
            type: FormActions.NAVIGATE,
            payload: route
        };
    };
    /**
     * @param {?} forms
     * @return {?}
     */
    FormActions.resetForms = /**
     * @param {?} forms
     * @return {?}
     */
    function (forms) {
        return {
            type: FormActions.RESET,
            payload: forms
        };
    };
    FormActions.ADD = '[forms] Add Form';
    FormActions.RESET = '[forms] Reset Forms';
    FormActions.NAVIGATE = '[forms] Navigate to Form';
    FormActions.STATUS_CHANGES = '[form] Status Changes';
    FormActions.VALUE_CHANGES = '[form] Value Changes';
    return FormActions;
}());
export { FormActions };
function FormActions_tsickle_Closure_declarations() {
    /** @type {?} */
    FormActions.ADD;
    /** @type {?} */
    FormActions.RESET;
    /** @type {?} */
    FormActions.NAVIGATE;
    /** @type {?} */
    FormActions.STATUS_CHANGES;
    /** @type {?} */
    FormActions.VALUE_CHANGES;
}
