// @caiu/library/esm2015/lib/store/actions.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
export class AppActions {
    /**
     * @return {?}
     */
    static noAction() {
        return {
            type: AppActions.NO_ACTION
        };
    }
}
AppActions.INIT_STORE = '[Storage] Init Store';
AppActions.NO_ACTION = '[App] No Action';
function AppActions_tsickle_Closure_declarations() {
    /** @type {?} */
    AppActions.INIT_STORE;
    /** @type {?} */
    AppActions.NO_ACTION;
}
export class ConfigActions {
    /**
     * @param {?} config
     * @return {?}
     */
    static initialize(config) {
        return {
            type: ConfigActions.INITIALIZE,
            payload: config
        };
    }
}
ConfigActions.INITIALIZE = '[Config] Initialize Configuration';
function ConfigActions_tsickle_Closure_declarations() {
    /** @type {?} */
    ConfigActions.INITIALIZE;
}
export class SidenavActions {
    /**
     * @return {?}
     */
    static close() {
        return {
            type: SidenavActions.CLOSE
        };
    }
    /**
     * @return {?}
     */
    static open() {
        return {
            type: SidenavActions.OPEN
        };
    }
    /**
     * @return {?}
     */
    static toggle() {
        return {
            type: SidenavActions.TOGGLE
        };
    }
}
SidenavActions.CLOSE = '[Sidenav] Close';
SidenavActions.OPEN = '[Sidenav] Open';
SidenavActions.TOGGLE = '[Sidenav] Toggle';
function SidenavActions_tsickle_Closure_declarations() {
    /** @type {?} */
    SidenavActions.CLOSE;
    /** @type {?} */
    SidenavActions.OPEN;
    /** @type {?} */
    SidenavActions.TOGGLE;
}
export class WindowActions {
    /**
     * @param {?} payload
     * @return {?}
     */
    static resize(payload) {
        return {
            type: WindowActions.RESIZE,
            payload
        };
    }
}
WindowActions.RESIZE = '[Window] Resize';
function WindowActions_tsickle_Closure_declarations() {
    /** @type {?} */
    WindowActions.RESIZE;
}
