// @caiu/library/esm2015/lib/store/store.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { NgModule } from '@angular/core';
import { StoreModule as NgrxStoreModule } from '@ngrx/store';
export class StoreModule {
    /**
     * @template T, V
     * @param {?} reducers
     * @param {?=} config
     * @return {?}
     */
    static forRoot(reducers, config) {
        return NgrxStoreModule.forRoot(reducers, config);
    }
    /**
     * @template T, V
     * @param {?} featureName
     * @param {?} reducers
     * @param {?=} config
     * @return {?}
     */
    static forFeatures(featureName, reducers, config) {
        return NgrxStoreModule.forFeature(featureName, reducers, config);
    }
    /**
     * @template T, V
     * @param {?} featureName
     * @param {?} reducer
     * @param {?=} config
     * @return {?}
     */
    static forFeature(featureName, reducer, config) {
        return NgrxStoreModule.forFeature(featureName, reducer, config);
    }
}
StoreModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    NgrxStoreModule,
                ],
                declarations: [],
                exports: [
                    NgrxStoreModule,
                ]
            },] },
];
