// @caiu/library/esm5/lib/components/file-upload/file-upload.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { build } from '../../shared/utils';
var FileUpload = /** @class */ (function () {
    function FileUpload() {
        this.lastModified = 0;
        this.lastModifiedDate = new Date();
        this.name = '';
        this.size = 0;
        this.type = '';
        this.webkitRelativePath = '';
        this.order = 0;
        this.readyState = 'EMPTY';
        this.src = '';
    }
    /**
     * @param {?} file
     * @return {?}
     */
    FileUpload.BuildFromFile = /**
     * @param {?} file
     * @return {?}
     */
    function (file) {
        return build(FileUpload, {
            name: file.fileName,
            size: file.fileSize,
            type: file.mimeType,
            src: file.src,
            order: file.order
        });
    };
    /**
     * @param {?} reader
     * @return {?}
     */
    FileUpload.GetReadyState = /**
     * @param {?} reader
     * @return {?}
     */
    function (reader) {
        switch (reader.readyState) {
            case 0:
                return 'EMPTY';
            case 1:
                return 'LOADING';
            case 2:
                return 'DONE';
            default:
                return 'EMPTY';
        }
    };
    Object.defineProperty(FileUpload.prototype, "extension", {
        get: /**
         * @return {?}
         */
        function () {
            return this.name.split('.').pop();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FileUpload.prototype, "isImage", {
        get: /**
         * @return {?}
         */
        function () {
            return this.type.split('/')[0] === 'image';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FileUpload.prototype, "loading", {
        get: /**
         * @return {?}
         */
        function () {
            return this.readyState === 'LOADING';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FileUpload.prototype, "file", {
        get: /**
         * @return {?}
         */
        function () {
            return File.BuildFromFileUpload(this);
        },
        enumerable: true,
        configurable: true
    });
    return FileUpload;
}());
export { FileUpload };
function FileUpload_tsickle_Closure_declarations() {
    /** @type {?} */
    FileUpload.prototype.lastModified;
    /** @type {?} */
    FileUpload.prototype.lastModifiedDate;
    /** @type {?} */
    FileUpload.prototype.name;
    /** @type {?} */
    FileUpload.prototype.size;
    /** @type {?} */
    FileUpload.prototype.type;
    /** @type {?} */
    FileUpload.prototype.webkitRelativePath;
    /** @type {?} */
    FileUpload.prototype.order;
    /** @type {?} */
    FileUpload.prototype.readyState;
    /** @type {?} */
    FileUpload.prototype.src;
}
/**
 * @param {?} reader
 * @return {?}
 */
export function getReadyState(reader) {
    switch (reader.readyState) {
        case 0:
            return 'EMPTY';
        case 1:
            return 'LOADING';
        case 2:
            return 'DONE';
        default:
            return 'EMPTY';
    }
}
var File = /** @class */ (function () {
    function File() {
        this.id = 0;
        this.fileBinary = [];
        this.fileExtension = '';
        this.fileName = '';
        this.fileSize = 0;
        this.mimeType = '';
        this.order = 0;
    }
    /**
     * @param {?} upload
     * @return {?}
     */
    File.BuildFromFileUpload = /**
     * @param {?} upload
     * @return {?}
     */
    function (upload) {
        return build(File, {
            fileBinary: upload.src.replace(File.GetSrcPrefix(upload.type), ''),
            fileExtension: upload.extension,
            fileName: upload.name,
            fileSize: upload.size,
            mimeType: upload.type,
            order: upload.order
        });
    };
    /**
     * @param {?} img
     * @param {?=} defaultSrc
     * @return {?}
     */
    File.GetImageBinarySrc = /**
     * @param {?} img
     * @param {?=} defaultSrc
     * @return {?}
     */
    function (img, defaultSrc) {
        if (defaultSrc === void 0) { defaultSrc = ''; }
        if (Array.isArray(img)) {
            if (img.length > 0) {
                return "data:" + img[0].mimeType + ";base64," + img[0].fileBinary;
            }
            else {
                return defaultSrc;
            }
        }
        return img && img.mimeType && img.fileBinary ? "data:" + img.mimeType + ";base64," + img.fileBinary : defaultSrc;
    };
    /**
     * @param {?} mimeType
     * @return {?}
     */
    File.GetSrcPrefix = /**
     * @param {?} mimeType
     * @return {?}
     */
    function (mimeType) {
        return "data:" + mimeType + ";base64,";
    };
    Object.defineProperty(File.prototype, "src", {
        // get fileUpload(): FileUpload {
        //     return FileUpload.BuildFromFile(this);
        // }
        get: /**
         * @return {?}
         */
        function () {
            return File.GetImageBinarySrc(this);
        },
        enumerable: true,
        configurable: true
    });
    return File;
}());
export { File };
function File_tsickle_Closure_declarations() {
    /** @type {?} */
    File.prototype.id;
    /** @type {?} */
    File.prototype.fileBinary;
    /** @type {?} */
    File.prototype.fileExtension;
    /** @type {?} */
    File.prototype.fileName;
    /** @type {?} */
    File.prototype.fileSize;
    /** @type {?} */
    File.prototype.mimeType;
    /** @type {?} */
    File.prototype.order;
}
