// @caiu/library/esm5/lib/components/file-upload/upload/upload.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, Input, Output, EventEmitter } from '@angular/core';
import { FileUpload } from '../file-upload.model';
var UploadComponent = /** @class */ (function () {
    function UploadComponent() {
        this.upload = new FileUpload();
        this.height = 200;
        this.width = 600;
        this.remove = new EventEmitter();
    }
    Object.defineProperty(UploadComponent.prototype, "graphic", {
        get: /**
         * @return {?}
         */
        function () {
            if (this.upload.loading) {
                return 'cloud_upload';
            }
            else if (this.upload.isImage) {
                return 'image';
            }
            else {
                switch (this.upload.extension) {
                    case 'pdf':
                        return 'picture_as_pdf';
                    default:
                        return 'cloud_done';
                }
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(UploadComponent.prototype, "showImage", {
        get: /**
         * @return {?}
         */
        function () {
            return this.graphic === 'image';
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    UploadComponent.prototype.onRemove = /**
     * @return {?}
     */
    function () {
        this.remove.emit(this.upload);
    };
    UploadComponent.decorators = [
        { type: Component, args: [{
                    selector: 'iu-upload',
                    template: "<div class=\"upload\" [style.height]=\"img?.style?.height\">\n  <div class=\"graphic\" #img>\n    <img *ngIf=\"showImage\" [src]=\"upload.src\" [style.max-height.px]=\"height\" [style.max-width.px]=\"width\">\n    <mat-icon *ngIf=\"!showImage\" class=\"icon-graphic\" [style.max-height.px]=\"height\" [style.max-width.px]=\"width\">{{ graphic }}</mat-icon>\n    <mat-icon *ngIf=\"!upload.loading\" class=\"remove-upload\" (click)=\"onRemove()\">cancel</mat-icon>\n  </div>\n  <mat-spinner mode=\"indeterminate\" *ngIf=\"upload.loading\"></mat-spinner>\n  <div class=\"file-info\" *ngIf=\"!upload.loading\">\n    <span class=\"file-name\">{{ upload.name }}</span>\n  </div>\n</div>\n",
                    styles: [".upload{max-height:420px;max-width:600px;margin:10px 20px;position:relative;display:flex}.upload img{max-height:400px}.graphic{padding:0!important;float:left;margin:10px;position:relative}.icon-graphic{font-size:72px;height:72px;width:72px}.mat-progress-bar{position:absolute;left:0;top:73px}.file-info{float:right}.file-name{margin:18px 6px;display:inline-block;line-height:36px;font-size:20px;height:36px;font-weight:700;color:#000}.remove-upload{color:red;cursor:pointer;position:absolute;right:-20px;top:-20px;font-size:40px;height:40px;width:40px}"]
                },] },
    ];
    /** @nocollapse */
    UploadComponent.ctorParameters = function () { return []; };
    UploadComponent.propDecorators = {
        upload: [{ type: Input }],
        height: [{ type: Input }],
        width: [{ type: Input }],
        remove: [{ type: Output }]
    };
    return UploadComponent;
}());
export { UploadComponent };
function UploadComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    UploadComponent.prototype.upload;
    /** @type {?} */
    UploadComponent.prototype.height;
    /** @type {?} */
    UploadComponent.prototype.width;
    /** @type {?} */
    UploadComponent.prototype.remove;
}
