// @caiu/library/esm5/lib/storage/storage.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Injectable, Inject } from '@angular/core';
import { Store } from '@ngrx/store';
import { StorageActions } from './storage.actions';
import { ActionStore } from './storage.models';
import { STORE_KEY } from './storage.module';
import { build, inArray, filterState } from '../shared/utils';
import * as i0 from "@angular/core";
import * as i1 from "@ngrx/store";
import * as i2 from "./storage.module";
var StorageService = /** @class */ (function () {
    function StorageService(store, storeKey) {
        this.store = store;
        this.storeKey = storeKey;
        this.localStorageActions = [];
        this.sessionStorageActions = [];
        this.localStorageMapper = function (s) { return s; };
        this.sessionStorageMapper = function (s) { return s; };
    }
    Object.defineProperty(StorageService.prototype, "localStore", {
        get: /**
         * @return {?}
         */
        function () {
            return localStorage.getItem(this.storeKey);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StorageService.prototype, "sessionStore", {
        get: /**
         * @return {?}
         */
        function () {
            return sessionStorage.getItem(this.storeKey);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StorageService.prototype, "storage", {
        get: /**
         * @return {?}
         */
        function () {
            var /** @type {?} */ localStore = this.localStore;
            var /** @type {?} */ sessionStore = this.sessionStore;
            return build(Storage, Object.assign({}, localStore, sessionStore), {
                localStore: localStore,
                sessionStore: sessionStore
            });
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Initialize the use of local storage and/or session storage
     * @param localStorageMapper Mapper function for mapping state to local store
     * @param sessionStorageMapper Mapper function for mapping state to session store
     * @param localStorageActions Actions after which to save to local storage
     * @param sessionStorageActions Actions after which to save to session storage
     */
    /**
     * Initialize the use of local storage and/or session storage
     * @param {?=} localStorageMapper Mapper function for mapping state to local store
     * @param {?=} sessionStorageMapper Mapper function for mapping state to session store
     * @param {?=} localStorageActions Actions after which to save to local storage
     * @param {?=} sessionStorageActions Actions after which to save to session storage
     * @return {?}
     */
    StorageService.prototype.init = /**
     * Initialize the use of local storage and/or session storage
     * @param {?=} localStorageMapper Mapper function for mapping state to local store
     * @param {?=} sessionStorageMapper Mapper function for mapping state to session store
     * @param {?=} localStorageActions Actions after which to save to local storage
     * @param {?=} sessionStorageActions Actions after which to save to session storage
     * @return {?}
     */
    function (localStorageMapper, sessionStorageMapper, localStorageActions, sessionStorageActions) {
        if (localStorageMapper === void 0) { localStorageMapper = function (s) { return s; }; }
        if (sessionStorageMapper === void 0) { sessionStorageMapper = function (s) { return s; }; }
        if (localStorageActions === void 0) { localStorageActions = []; }
        if (sessionStorageActions === void 0) { sessionStorageActions = []; }
        var /** @type {?} */ localStore = this.initLocalStore(localStorageMapper, localStorageActions);
        var /** @type {?} */ sessionStore = this.initSessionStore(sessionStorageMapper, sessionStorageActions);
        this.store.dispatch(StorageActions.initStore(localStore, sessionStore));
    };
    /**
     * Initialize the use of local storage
     * @param localStorageMapper Mapper function for mapping state to local store
     * @param localStorageActions Actions after which to save to local storage
     */
    /**
     * Initialize the use of local storage
     * @param {?=} localStorageMapper Mapper function for mapping state to local store
     * @param {?=} localStorageActions Actions after which to save to local storage
     * @return {?}
     */
    StorageService.prototype.initLocalStore = /**
     * Initialize the use of local storage
     * @param {?=} localStorageMapper Mapper function for mapping state to local store
     * @param {?=} localStorageActions Actions after which to save to local storage
     * @return {?}
     */
    function (localStorageMapper, localStorageActions) {
        if (localStorageMapper === void 0) { localStorageMapper = function (s) { return s; }; }
        if (localStorageActions === void 0) { localStorageActions = []; }
        this.localStorageMapper = localStorageMapper;
        this.localStorageActions = localStorageActions;
        var /** @type {?} */ ls = this.localStore;
        var /** @type {?} */ localStore = ls ? JSON.parse(ls) : {};
        return localStore;
    };
    /**
     * Initialize the use of session storage
     * @param sessionStorageMapper Mapper function for mapping state to session store
     * @param sessionStorageActions Actions after which to save to session storage
     */
    /**
     * Initialize the use of session storage
     * @param {?=} sessionStorageMapper Mapper function for mapping state to session store
     * @param {?=} sessionStorageActions Actions after which to save to session storage
     * @return {?}
     */
    StorageService.prototype.initSessionStore = /**
     * Initialize the use of session storage
     * @param {?=} sessionStorageMapper Mapper function for mapping state to session store
     * @param {?=} sessionStorageActions Actions after which to save to session storage
     * @return {?}
     */
    function (sessionStorageMapper, sessionStorageActions) {
        if (sessionStorageMapper === void 0) { sessionStorageMapper = function (s) { return s; }; }
        if (sessionStorageActions === void 0) { sessionStorageActions = []; }
        this.sessionStorageMapper = sessionStorageMapper;
        this.sessionStorageActions = sessionStorageActions;
        var /** @type {?} */ ss = this.sessionStore;
        var /** @type {?} */ sessionStore = ss ? JSON.parse(ss) : {};
        return sessionStorage;
    };
    /**
     * @param {?} actionType
     * @return {?}
     */
    StorageService.prototype.inLocalStorage = /**
     * @param {?} actionType
     * @return {?}
     */
    function (actionType) {
        return inArray(this.localStorageActions, actionType);
    };
    /**
     * @param {?} actionType
     * @return {?}
     */
    StorageService.prototype.inSessionStorage = /**
     * @param {?} actionType
     * @return {?}
     */
    function (actionType) {
        return inArray(this.sessionStorageActions, actionType);
    };
    /**
     * Store state to local storage.
     */
    /**
     * Store state to local storage.
     * @param {?} state
     * @param {?} action
     * @return {?}
     */
    StorageService.prototype.storeLocal = /**
     * Store state to local storage.
     * @param {?} state
     * @param {?} action
     * @return {?}
     */
    function (state, action) {
        try {
            // const filteredState = new ActionStore(action, this.filterState(Object.assign({}, state)));
            var /** @type {?} */ filteredState = Object.assign(filterState(new ActionStore(action)), filterState(Object.assign({}, this.localStorageMapper(state))));
            var /** @type {?} */ serializedState = JSON.stringify(filteredState);
            localStorage.removeItem(this.storeKey);
            localStorage.setItem(this.storeKey, serializedState);
        }
        catch (/** @type {?} */ err) {
            console.error("ERROR SAVING LOCAL STATE!\nAction:\t" + action.actionType + "\n", err);
        }
    };
    /**
     * Store state to session storage.
     */
    /**
     * Store state to session storage.
     * @param {?} state
     * @param {?} action
     * @return {?}
     */
    StorageService.prototype.storeSession = /**
     * Store state to session storage.
     * @param {?} state
     * @param {?} action
     * @return {?}
     */
    function (state, action) {
        try {
            // const filteredState = new ActionStore(action, this.filterState(Object.assign({}, state)));
            // const filteredState = this.filterState(Object.assign({}, state));
            var /** @type {?} */ filteredState = Object.assign(filterState(new ActionStore(action)), filterState(Object.assign({}, this.sessionStorageMapper(state))));
            var /** @type {?} */ serializedState = JSON.stringify(filteredState);
            localStorage.removeItem(this.storeKey);
            sessionStorage.setItem(this.storeKey, serializedState);
        }
        catch (/** @type {?} */ err) {
            console.error("ERROR SAVING SESSION STATE!\nAction:\t" + action.actionType + "\n", err);
        }
    };
    StorageService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] },
    ];
    /** @nocollapse */
    StorageService.ctorParameters = function () { return [
        { type: Store },
        { type: String, decorators: [{ type: Inject, args: [STORE_KEY,] }] }
    ]; };
    /** @nocollapse */ StorageService.ngInjectableDef = i0.defineInjectable({ factory: function StorageService_Factory() { return new StorageService(i0.inject(i1.Store), i0.inject(i2.STORE_KEY)); }, token: StorageService, providedIn: "root" });
    return StorageService;
}());
export { StorageService };
function StorageService_tsickle_Closure_declarations() {
    /** @type {?} */
    StorageService.prototype.localStorageActions;
    /** @type {?} */
    StorageService.prototype.sessionStorageActions;
    /** @type {?} */
    StorageService.prototype.localStorageMapper;
    /** @type {?} */
    StorageService.prototype.sessionStorageMapper;
    /** @type {?} */
    StorageService.prototype.store;
    /** @type {?} */
    StorageService.prototype.storeKey;
}
