// @caiu/library/esm2015/lib/components/grid/simple-grid/simple-grid.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { Grid } from '../grid.model';
import { GridColumn } from '../grid-column/grid-column.model';
import { Metadata } from '../../../shared/models';
import { toArray } from '../../../shared/utils';
export class SimpleGridComponent {
    constructor() {
        this.data = [];
        this.export = false;
        this.filterable = false;
        this.fileName = 'export';
        this.height = 600;
        this.pageSize = 30;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set metadata(value) {
        this._metadata = value;
    }
    /**
     * @return {?}
     */
    get metadata() {
        return this._metadata ? this._metadata : (toArray(this.data).length > 0 ? toArray(this.data)[0].metadata : null);
    }
    /**
     * @return {?}
     */
    get columns() {
        return this.metadata ? this.columnKeys.map(key => new GridColumn(key, this.metadata[key].columnLabel)) : [];
    }
    /**
     * @return {?}
     */
    get columnKeys() {
        return this.metadata ? Object.keys(this.metadata).filter(key => this.metadata[key].columnLabel ? true : false) : [];
    }
    /**
     * @return {?}
     */
    get columnLabels() {
        return this.metadata ? this.columnKeys.map(key => this.metadata[key].columnLabel) : [];
    }
    /**
     * @return {?}
     */
    get gridData() {
        return Grid.Build(toArray(this.data));
    }
}
SimpleGridComponent.decorators = [
    { type: Component, args: [{
                selector: 'iu-simple-grid',
                template: `<iu-grid [model]="gridData" [export]="export" [filterable]="filterable" [fileName]="fileName" [pageSize]="pageSize" [height]="height">
  <iu-grid-column *ngFor="let column of columns" [model]="column"></iu-grid-column>
</iu-grid>
`,
                styles: [``]
            },] },
];
/** @nocollapse */
SimpleGridComponent.ctorParameters = () => [];
SimpleGridComponent.propDecorators = {
    data: [{ type: Input }],
    export: [{ type: Input }],
    filterable: [{ type: Input }],
    fileName: [{ type: Input }],
    height: [{ type: Input }],
    pageSize: [{ type: Input }],
    metadata: [{ type: Input }]
};
function SimpleGridComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    SimpleGridComponent.prototype.data;
    /** @type {?} */
    SimpleGridComponent.prototype.export;
    /** @type {?} */
    SimpleGridComponent.prototype.filterable;
    /** @type {?} */
    SimpleGridComponent.prototype.fileName;
    /** @type {?} */
    SimpleGridComponent.prototype.height;
    /** @type {?} */
    SimpleGridComponent.prototype.pageSize;
    /** @type {?} */
    SimpleGridComponent.prototype._metadata;
}
