// @caiu/library/esm5/lib/components/wallpaper/wallpaper.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, ChangeDetectionStrategy, Input, ElementRef } from '@angular/core';
import { toPx } from '../../shared/utils';
var WallpaperComponent = /** @class */ (function () {
    function WallpaperComponent(elementRef) {
        this.elementRef = elementRef;
        this.images = [];
        this.windowHeight = 0;
        this.windowWidth = 0;
    }
    Object.defineProperty(WallpaperComponent.prototype, "canvasHeight", {
        get: /**
         * @return {?}
         */
        function () {
            return this.windowWidth * 2;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WallpaperComponent.prototype, "canvasWidth", {
        get: /**
         * @return {?}
         */
        function () {
            return this.windowWidth - 100;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WallpaperComponent.prototype, "elementHeight", {
        get: /**
         * @return {?}
         */
        function () {
            return this.elementRef && this.elementRef.nativeElement && this.elementRef.nativeElement.style ?
                this.elementRef.nativeElement.style.height : 0;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WallpaperComponent.prototype, "elementHeightPx", {
        get: /**
         * @return {?}
         */
        function () {
            return toPx(this.elementHeight);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WallpaperComponent.prototype, "elementWidth", {
        get: /**
         * @return {?}
         */
        function () {
            return this.elementRef && this.elementRef.nativeElement && this.elementRef.nativeElement.style ?
                this.elementRef.nativeElement.style.width : 0;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WallpaperComponent.prototype, "elementWidthPx", {
        get: /**
         * @return {?}
         */
        function () {
            return toPx(this.elementWidth);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WallpaperComponent.prototype, "maxColumns", {
        get: /**
         * @return {?}
         */
        function () {
            return 3;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WallpaperComponent.prototype, "maxRows", {
        get: /**
         * @return {?}
         */
        function () {
            return 3;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WallpaperComponent.prototype, "totalColumns", {
        get: /**
         * @return {?}
         */
        function () {
            return 12;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WallpaperComponent.prototype, "totalRows", {
        get: /**
         * @return {?}
         */
        function () {
            return 24;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    WallpaperComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    WallpaperComponent.decorators = [
        { type: Component, args: [{
                    selector: 'iu-wallpaper',
                    template: "<div class=\"wallpaper\" [style.height.px]=\"windowHeight\" [style.width.px]=\"windowWidth\">\n  <iu-collage [images]=\"images\" [canvasHeight]=\"canvasHeight\" [canvasWidth]=\"canvasWidth\" [totalRows]=\"totalRows\" [totalColumns]=\"totalColumns\"\n    [maxRows]=\"maxRows\" [maxColumns]=\"maxColumns\"></iu-collage>\n</div>",
                    styles: [".wallpaper{height:100%;width:100%}"],
                    changeDetection: ChangeDetectionStrategy.OnPush
                },] },
    ];
    /** @nocollapse */
    WallpaperComponent.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    WallpaperComponent.propDecorators = {
        images: [{ type: Input }],
        windowHeight: [{ type: Input }],
        windowWidth: [{ type: Input }]
    };
    return WallpaperComponent;
}());
export { WallpaperComponent };
function WallpaperComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    WallpaperComponent.prototype.images;
    /** @type {?} */
    WallpaperComponent.prototype.windowHeight;
    /** @type {?} */
    WallpaperComponent.prototype.windowWidth;
    /** @type {?} */
    WallpaperComponent.prototype.elementRef;
}
