// @caiu/library/esm5/lib/components/datepicker/datepicker.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { ChangeDetectionStrategy, Component, EventEmitter, forwardRef, Input, Output, ViewChild, } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { MatDatepicker } from '@angular/material';
import { DateHelper } from '../../shared/date';
export var /** @type {?} */ DATEPICKER_ACCESSOR = {
    provide: NG_VALUE_ACCESSOR,
    useExisting: forwardRef(function () { return DatepickerComponent; }),
    multi: true
};
var DatepickerComponent = /** @class */ (function () {
    function DatepickerComponent() {
        this.placeholder = 'Choose a date';
        this.startView = 'month';
        this.selectedChanged = new EventEmitter();
    }
    Object.defineProperty(DatepickerComponent.prototype, "value", {
        get: /**
         * @return {?}
         */
        function () {
            return this._value;
        },
        set: /**
         * @param {?} val
         * @return {?}
         */
        function (val) {
            this._value = val;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} fn
     * @return {?}
     */
    DatepickerComponent.prototype.registerOnChange = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.onModelChange = fn;
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    DatepickerComponent.prototype.registerOnTouched = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.onTouch = fn;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    DatepickerComponent.prototype.writeValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.value = value;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    DatepickerComponent.prototype.onChange = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.value = value;
        if (this.onModelChange) {
            this.onModelChange(value);
        }
    };
    /**
     * @param {?} input
     * @return {?}
     */
    DatepickerComponent.prototype.onBlur = /**
     * @param {?} input
     * @return {?}
     */
    function (input) {
        var /** @type {?} */ date = new Date(input.value);
        if (DateHelper.IsValidDate(date)) {
            this.changeSelected(date);
        }
        else {
            input.value = DateHelper.FormatDate(this.value);
        }
    };
    Object.defineProperty(DatepickerComponent.prototype, "id", {
        get: /**
         * @return {?}
         */
        function () {
            return this.datepicker.id;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DatepickerComponent.prototype, "opened", {
        get: /**
         * @return {?}
         */
        function () {
            return this.datepicker.opened;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    DatepickerComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.startAt = this.value;
    };
    /**
     * @param {?} date
     * @return {?}
     */
    DatepickerComponent.prototype.changeSelected = /**
     * @param {?} date
     * @return {?}
     */
    function (date) {
        this.onChange(date);
        this.selectedChanged.emit(date);
    };
    /**
     * @return {?}
     */
    DatepickerComponent.prototype.close = /**
     * @return {?}
     */
    function () {
        this.datepicker.close();
    };
    /**
     * @return {?}
     */
    DatepickerComponent.prototype.open = /**
     * @return {?}
     */
    function () {
        this.datepicker.open();
    };
    DatepickerComponent.decorators = [
        { type: Component, args: [{
                    selector: 'iu-datepicker',
                    template: "<mat-form-field>\n  <input matInput [placeholder]=\"placeholder\" [matDatepicker]=\"picker\" (blur)=\"onBlur(pickerInput)\" (change)=\"onChange(value)\"\n    [min]=\"min\" [max]=\"max\" [value]=\"value\" #pickerInput>\n  <mat-datepicker-toggle matSuffix [for]=\"picker\"></mat-datepicker-toggle>\n  <mat-datepicker #picker [startAt]=\"startAt\" [startView]=\"startView\" [touchUi]=\"touchUi\" (selectedChanged)=\"changeSelected($event)\"></mat-datepicker>\n</mat-form-field>",
                    styles: [".mat-input-container{float:left}"],
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    providers: [DATEPICKER_ACCESSOR]
                },] },
    ];
    /** @nocollapse */
    DatepickerComponent.ctorParameters = function () { return []; };
    DatepickerComponent.propDecorators = {
        min: [{ type: Input }],
        max: [{ type: Input }],
        placeholder: [{ type: Input }],
        startAt: [{ type: Input }],
        startView: [{ type: Input }],
        touchUi: [{ type: Input }],
        selectedChanged: [{ type: Output }],
        datepicker: [{ type: ViewChild, args: ['picker',] }],
        value: [{ type: Input }]
    };
    return DatepickerComponent;
}());
export { DatepickerComponent };
function DatepickerComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    DatepickerComponent.prototype.min;
    /** @type {?} */
    DatepickerComponent.prototype.max;
    /** @type {?} */
    DatepickerComponent.prototype.placeholder;
    /** @type {?} */
    DatepickerComponent.prototype.startAt;
    /** @type {?} */
    DatepickerComponent.prototype.startView;
    /** @type {?} */
    DatepickerComponent.prototype.touchUi;
    /** @type {?} */
    DatepickerComponent.prototype.selectedChanged;
    /** @type {?} */
    DatepickerComponent.prototype.datepicker;
    /** @type {?} */
    DatepickerComponent.prototype.onModelChange;
    /** @type {?} */
    DatepickerComponent.prototype.onTouch;
    /** @type {?} */
    DatepickerComponent.prototype._value;
    /** @type {?} */
    DatepickerComponent.prototype.focused;
}
