// @caiu/library/esm2015/lib/errors/errors.actions.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
export class ErrorActions {
    /**
     * @param {?} e
     * @param {?} outlet
     * @return {?}
     */
    static addAtOutlet(e, outlet) {
        let /** @type {?} */ payload = {
            key: outlet.key,
            error: outlet.handler ? outlet.handler(e) : e
        };
        return ErrorActions.add(payload);
    }
    /**
     * @param {?} key
     * @param {?} e
     * @return {?}
     */
    static addError(key, e) {
        let /** @type {?} */ payload = {
            key: key,
            error: e
        };
        return ErrorActions.add(payload);
    }
    /**
     * @param {?} payload
     * @return {?}
     */
    static add(payload) {
        return {
            type: ErrorActions.ADD,
            payload: payload
        };
    }
    /**
     * @param {?} key
     * @return {?}
     */
    static remove(key) {
        return {
            type: ErrorActions.REMOVE,
            payload: key
        };
    }
}
ErrorActions.ADD = '[Error] Add Error';
ErrorActions.REMOVE = '[Error] Remove Error';
ErrorActions.HANDLE_ERROR = '[Error Handler] Handle Error';
ErrorActions.THROW_ERROR = '[Error Handler] Throw Error';
function ErrorActions_tsickle_Closure_declarations() {
    /** @type {?} */
    ErrorActions.ADD;
    /** @type {?} */
    ErrorActions.REMOVE;
    /** @type {?} */
    ErrorActions.HANDLE_ERROR;
    /** @type {?} */
    ErrorActions.THROW_ERROR;
}
