// @caiu/library/esm2015/lib/components/tile/tile.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, Input, ChangeDetectionStrategy } from '@angular/core';
import { Tile } from './tile.model';
export class TileComponent {
    constructor() {
        this.tile = new Tile();
    }
    /**
     * @return {?}
     */
    get backgroundColor() {
        return '';
    }
    /**
     * @return {?}
     */
    get height() {
        return this.tile.height;
    }
    /**
     * @return {?}
     */
    get imageHeight() {
        return this.tile.imageHeight;
    }
    /**
     * @return {?}
     */
    get imageSrc() {
        return this.tile.image.src;
    }
    /**
     * @return {?}
     */
    get imageWidth() {
        return this.tile.imageWidth;
    }
    /**
     * @return {?}
     */
    get position() {
        return 'absolute';
    }
    /**
     * @return {?}
     */
    get positionLeft() {
        return this.tile.positionLeft;
    }
    /**
     * @return {?}
     */
    get positionTop() {
        return this.tile.positionTop;
    }
    /**
     * @return {?}
     */
    get width() {
        return this.tile.width;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
}
TileComponent.decorators = [
    { type: Component, args: [{
                selector: 'iu-tile',
                template: `<div fxLayout fxLayoutAlign="center center" class="tile" [style.position]="position" [style.top.px]="positionTop" [style.left.px]="positionLeft"
  [style.background-color]="backgroundColor" [style.height.px]="height" [style.width.px]="width">
  <img [src]="imageSrc" [style.height.px]="imageHeight" [style.width.px]="imageWidth">
</div>
`,
                styles: [``],
                changeDetection: ChangeDetectionStrategy.OnPush
            },] },
];
/** @nocollapse */
TileComponent.ctorParameters = () => [];
TileComponent.propDecorators = {
    tile: [{ type: Input }]
};
function TileComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    TileComponent.prototype.tile;
}
