// @caiu/library/esm2015/lib/components/tile/tile.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Dimensions, Image, Coordinates } from '../../shared/models';
export class Tile {
    constructor() {
        this.id = 0;
        this.cellHeight = 0;
        this.cellWidth = 0;
        this.color = '#fff';
        this.name = '';
        this.src = '';
        this.text = '';
        this._coordinates = new Coordinates();
        this._dimensions = new Dimensions();
        this._image = new Image();
    }
    /**
     * @return {?}
     */
    get coordinates() {
        return this._coordinates || new Coordinates();
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set coordinates(value) {
        this._coordinates = value;
    }
    /**
     * @return {?}
     */
    get dimensions() {
        return this._dimensions || new Dimensions();
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set dimensions(value) {
        this._dimensions = value;
    }
    /**
     * @return {?}
     */
    get image() {
        return this._image || new Image();
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set image(value) {
        this._image = value;
    }
    /**
     * @return {?}
     */
    get height() {
        return this.rows * this.cellHeight;
    }
    /**
     * @return {?}
     */
    get width() {
        return this.columns * this.cellWidth;
    }
    /**
     * @return {?}
     */
    get imageSrc() {
        return this.image.src;
    }
    /**
     * @return {?}
     */
    get imageHeight() {
        return this.image.height * this.scale;
    }
    /**
     * @return {?}
     */
    get imageWidth() {
        return this.image.width * this.scale;
    }
    /**
     * @return {?}
     */
    get rows() {
        return this.dimensions.rows;
    }
    /**
     * @return {?}
     */
    get columns() {
        return this.dimensions.columns;
    }
    /**
     * @return {?}
     */
    get positionTop() {
        return this.startRow * this.cellHeight;
    }
    /**
     * @return {?}
     */
    get positionLeft() {
        return this.startColumn * this.cellWidth;
    }
    /**
     * @return {?}
     */
    get approxRatio() {
        return this.rows * this.cellHeight / this.columns * this.cellWidth;
    }
    /**
     * @return {?}
     */
    get ratio() {
        return this.dimensions.ratio;
    }
    /**
     * @return {?}
     */
    get scale() {
        return this.heightScale >= this.widthScale ? this.heightScale : this.widthScale;
    }
    /**
     * @return {?}
     */
    get heightScale() {
        return (this.rows * this.cellHeight) / this.image.height;
    }
    /**
     * @return {?}
     */
    get widthScale() {
        return (this.columns * this.cellWidth) / this.image.width;
    }
    /**
     * @return {?}
     */
    get startRow() {
        return this.coordinates.row;
    }
    /**
     * @return {?}
     */
    get startColumn() {
        return this.coordinates.column;
    }
}
function Tile_tsickle_Closure_declarations() {
    /** @type {?} */
    Tile.prototype.id;
    /** @type {?} */
    Tile.prototype.cellHeight;
    /** @type {?} */
    Tile.prototype.cellWidth;
    /** @type {?} */
    Tile.prototype.color;
    /** @type {?} */
    Tile.prototype.name;
    /** @type {?} */
    Tile.prototype.src;
    /** @type {?} */
    Tile.prototype.text;
    /** @type {?} */
    Tile.prototype._coordinates;
    /** @type {?} */
    Tile.prototype._dimensions;
    /** @type {?} */
    Tile.prototype._image;
}
