// @caiu/library/esm2015/lib/components/grid/grid-cell/grid-cell.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, Input, ViewContainerRef, TemplateRef } from '@angular/core';
export class GridCellComponent {
    /**
     * @param {?} viewContainerRef
     */
    constructor(viewContainerRef) {
        this.viewContainerRef = viewContainerRef;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.viewContainerRef.createEmbeddedView(this.templateRef, this.dataItem);
    }
}
GridCellComponent.decorators = [
    { type: Component, args: [{
                selector: 'iu-grid-cell',
                template: `<div></div>`,
                styles: [``]
            },] },
];
/** @nocollapse */
GridCellComponent.ctorParameters = () => [
    { type: ViewContainerRef }
];
GridCellComponent.propDecorators = {
    dataItem: [{ type: Input }],
    templateRef: [{ type: Input }]
};
function GridCellComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    GridCellComponent.prototype.dataItem;
    /** @type {?} */
    GridCellComponent.prototype.templateRef;
    /** @type {?} */
    GridCellComponent.prototype.viewContainerRef;
}
