// @caiu/library/esm5/lib/components/tile/tile.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, Input, ChangeDetectionStrategy } from '@angular/core';
import { Tile } from './tile.model';
var TileComponent = /** @class */ (function () {
    function TileComponent() {
        this.tile = new Tile();
    }
    Object.defineProperty(TileComponent.prototype, "backgroundColor", {
        get: /**
         * @return {?}
         */
        function () {
            return '';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TileComponent.prototype, "height", {
        get: /**
         * @return {?}
         */
        function () {
            return this.tile.height;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TileComponent.prototype, "imageHeight", {
        get: /**
         * @return {?}
         */
        function () {
            return this.tile.imageHeight;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TileComponent.prototype, "imageSrc", {
        get: /**
         * @return {?}
         */
        function () {
            return this.tile.image.src;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TileComponent.prototype, "imageWidth", {
        get: /**
         * @return {?}
         */
        function () {
            return this.tile.imageWidth;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TileComponent.prototype, "position", {
        get: /**
         * @return {?}
         */
        function () {
            return 'absolute';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TileComponent.prototype, "positionLeft", {
        get: /**
         * @return {?}
         */
        function () {
            return this.tile.positionLeft;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TileComponent.prototype, "positionTop", {
        get: /**
         * @return {?}
         */
        function () {
            return this.tile.positionTop;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TileComponent.prototype, "width", {
        get: /**
         * @return {?}
         */
        function () {
            return this.tile.width;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    TileComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    TileComponent.decorators = [
        { type: Component, args: [{
                    selector: 'iu-tile',
                    template: "<div fxLayout fxLayoutAlign=\"center center\" class=\"tile\" [style.position]=\"position\" [style.top.px]=\"positionTop\" [style.left.px]=\"positionLeft\"\n  [style.background-color]=\"backgroundColor\" [style.height.px]=\"height\" [style.width.px]=\"width\">\n  <img [src]=\"imageSrc\" [style.height.px]=\"imageHeight\" [style.width.px]=\"imageWidth\">\n</div>\n",
                    styles: [""],
                    changeDetection: ChangeDetectionStrategy.OnPush
                },] },
    ];
    /** @nocollapse */
    TileComponent.ctorParameters = function () { return []; };
    TileComponent.propDecorators = {
        tile: [{ type: Input }]
    };
    return TileComponent;
}());
export { TileComponent };
function TileComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    TileComponent.prototype.tile;
}
