// @caiu/library/esm5/lib/components/tile/tile.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Dimensions, Image, Coordinates } from '../../shared/models';
var Tile = /** @class */ (function () {
    function Tile() {
        this.id = 0;
        this.cellHeight = 0;
        this.cellWidth = 0;
        this.color = '#fff';
        this.name = '';
        this.src = '';
        this.text = '';
        this._coordinates = new Coordinates();
        this._dimensions = new Dimensions();
        this._image = new Image();
    }
    Object.defineProperty(Tile.prototype, "coordinates", {
        get: /**
         * @return {?}
         */
        function () {
            return this._coordinates || new Coordinates();
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._coordinates = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Tile.prototype, "dimensions", {
        get: /**
         * @return {?}
         */
        function () {
            return this._dimensions || new Dimensions();
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._dimensions = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Tile.prototype, "image", {
        get: /**
         * @return {?}
         */
        function () {
            return this._image || new Image();
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._image = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Tile.prototype, "height", {
        get: /**
         * @return {?}
         */
        function () {
            return this.rows * this.cellHeight;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Tile.prototype, "width", {
        get: /**
         * @return {?}
         */
        function () {
            return this.columns * this.cellWidth;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Tile.prototype, "imageSrc", {
        get: /**
         * @return {?}
         */
        function () {
            return this.image.src;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Tile.prototype, "imageHeight", {
        get: /**
         * @return {?}
         */
        function () {
            return this.image.height * this.scale;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Tile.prototype, "imageWidth", {
        get: /**
         * @return {?}
         */
        function () {
            return this.image.width * this.scale;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Tile.prototype, "rows", {
        get: /**
         * @return {?}
         */
        function () {
            return this.dimensions.rows;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Tile.prototype, "columns", {
        get: /**
         * @return {?}
         */
        function () {
            return this.dimensions.columns;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Tile.prototype, "positionTop", {
        get: /**
         * @return {?}
         */
        function () {
            return this.startRow * this.cellHeight;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Tile.prototype, "positionLeft", {
        get: /**
         * @return {?}
         */
        function () {
            return this.startColumn * this.cellWidth;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Tile.prototype, "approxRatio", {
        get: /**
         * @return {?}
         */
        function () {
            return this.rows * this.cellHeight / this.columns * this.cellWidth;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Tile.prototype, "ratio", {
        get: /**
         * @return {?}
         */
        function () {
            return this.dimensions.ratio;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Tile.prototype, "scale", {
        get: /**
         * @return {?}
         */
        function () {
            return this.heightScale >= this.widthScale ? this.heightScale : this.widthScale;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Tile.prototype, "heightScale", {
        get: /**
         * @return {?}
         */
        function () {
            return (this.rows * this.cellHeight) / this.image.height;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Tile.prototype, "widthScale", {
        get: /**
         * @return {?}
         */
        function () {
            return (this.columns * this.cellWidth) / this.image.width;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Tile.prototype, "startRow", {
        get: /**
         * @return {?}
         */
        function () {
            return this.coordinates.row;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Tile.prototype, "startColumn", {
        get: /**
         * @return {?}
         */
        function () {
            return this.coordinates.column;
        },
        enumerable: true,
        configurable: true
    });
    return Tile;
}());
export { Tile };
function Tile_tsickle_Closure_declarations() {
    /** @type {?} */
    Tile.prototype.id;
    /** @type {?} */
    Tile.prototype.cellHeight;
    /** @type {?} */
    Tile.prototype.cellWidth;
    /** @type {?} */
    Tile.prototype.color;
    /** @type {?} */
    Tile.prototype.name;
    /** @type {?} */
    Tile.prototype.src;
    /** @type {?} */
    Tile.prototype.text;
    /** @type {?} */
    Tile.prototype._coordinates;
    /** @type {?} */
    Tile.prototype._dimensions;
    /** @type {?} */
    Tile.prototype._image;
}
