// @caiu/library/esm5/lib/components/file-upload/uploads/uploads.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, Input, ChangeDetectionStrategy, Output, EventEmitter } from '@angular/core';
var UploadsComponent = /** @class */ (function () {
    function UploadsComponent() {
        this.uploads = [];
        this.ordered = true;
        this.moveUp = new EventEmitter();
        this.moveDown = new EventEmitter();
        this.remove = new EventEmitter();
    }
    /**
     * @param {?} f
     * @return {?}
     */
    UploadsComponent.prototype.onMoveUp = /**
     * @param {?} f
     * @return {?}
     */
    function (f) {
        this.moveUp.emit(f);
    };
    /**
     * @param {?} f
     * @return {?}
     */
    UploadsComponent.prototype.onMoveDown = /**
     * @param {?} f
     * @return {?}
     */
    function (f) {
        this.moveDown.emit(f);
    };
    /**
     * @param {?} f
     * @return {?}
     */
    UploadsComponent.prototype.onRemove = /**
     * @param {?} f
     * @return {?}
     */
    function (f) {
        this.remove.emit(f);
    };
    UploadsComponent.decorators = [
        { type: Component, args: [{
                    selector: 'iu-uploads',
                    template: "<div class=\"uploads\" [style.height]=\"height\" [style.width]=\"width\">\n  <iu-file-preview *ngFor=\"let upload of uploads; let first = first; let last = last;\" [file]=\"upload\" [first]=\"first\" [last]=\"last\"\n    [ordered]=\"ordered\" (moveUp)=\"onMoveUp($event)\" (moveDown)=\"onMoveDown($event)\" (remove)=\"onRemove($event)\"></iu-file-preview>\n</div>\n",
                    styles: [".uploads{max-height:420px;max-width:600px;overflow-y:scroll;margin:10px 20px}.uploads::-webkit-scrollbar-track{-webkit-box-shadow:inset 0 0 6px rgba(0,0,0,.3);border-radius:10px;background-color:#424242}.uploads::-webkit-scrollbar{width:12px;background-color:#424242}.uploads::-webkit-scrollbar-thumb{border-radius:10px;-webkit-box-shadow:inset 0 0 6px rgba(0,0,0,.3)}.uploads .dragging{background-color:#424242}"],
                    changeDetection: ChangeDetectionStrategy.OnPush
                },] },
    ];
    /** @nocollapse */
    UploadsComponent.ctorParameters = function () { return []; };
    UploadsComponent.propDecorators = {
        uploads: [{ type: Input }],
        height: [{ type: Input }],
        width: [{ type: Input }],
        ordered: [{ type: Input }],
        moveUp: [{ type: Output }],
        moveDown: [{ type: Output }],
        remove: [{ type: Output }]
    };
    return UploadsComponent;
}());
export { UploadsComponent };
function UploadsComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    UploadsComponent.prototype.uploads;
    /** @type {?} */
    UploadsComponent.prototype.height;
    /** @type {?} */
    UploadsComponent.prototype.width;
    /** @type {?} */
    UploadsComponent.prototype.ordered;
    /** @type {?} */
    UploadsComponent.prototype.moveUp;
    /** @type {?} */
    UploadsComponent.prototype.moveDown;
    /** @type {?} */
    UploadsComponent.prototype.remove;
}
