// @caiu/library/esm5/lib/components/grid/grid-cell/grid-cell.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, Input, ViewContainerRef, TemplateRef } from '@angular/core';
var GridCellComponent = /** @class */ (function () {
    function GridCellComponent(viewContainerRef) {
        this.viewContainerRef = viewContainerRef;
    }
    /**
     * @return {?}
     */
    GridCellComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.viewContainerRef.createEmbeddedView(this.templateRef, this.dataItem);
    };
    GridCellComponent.decorators = [
        { type: Component, args: [{
                    selector: 'iu-grid-cell',
                    template: "<div></div>",
                    styles: [""]
                },] },
    ];
    /** @nocollapse */
    GridCellComponent.ctorParameters = function () { return [
        { type: ViewContainerRef }
    ]; };
    GridCellComponent.propDecorators = {
        dataItem: [{ type: Input }],
        templateRef: [{ type: Input }]
    };
    return GridCellComponent;
}());
export { GridCellComponent };
function GridCellComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    GridCellComponent.prototype.dataItem;
    /** @type {?} */
    GridCellComponent.prototype.templateRef;
    /** @type {?} */
    GridCellComponent.prototype.viewContainerRef;
}
