// @caiu/library/esm5/lib/components/grid/grid-column/grid-column.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, ElementRef, Input, TemplateRef, ViewChild, ViewContainerRef } from '@angular/core';
import { GridColumn } from './grid-column.model';
var GridColumnComponent = /** @class */ (function () {
    function GridColumnComponent(view) {
        this.view = view;
        this.editable = true;
        this.field = '';
        this.filter = 'text';
        this.filterable = true;
        this.hidden = false;
        this.locked = false;
        this.sortable = true;
    }
    Object.defineProperty(GridColumnComponent.prototype, "colField", {
        get: /**
         * @return {?}
         */
        function () {
            return this.model.metadata.name;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GridColumnComponent.prototype, "colTitle", {
        get: /**
         * @return {?}
         */
        function () {
            return this.model.title;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GridColumnComponent.prototype, "colWidth", {
        get: /**
         * @return {?}
         */
        function () {
            return this.model.width;
        },
        enumerable: true,
        configurable: true
    });
    GridColumnComponent.decorators = [
        { type: Component, args: [{
                    selector: 'iu-grid-column',
                    template: "<kendo-grid-column [field]=\"colField\" [title]=\"colTitle\" [width]=\"colWidth\" #kendoGridColumn>\n  <span #content>\n    <ng-content></ng-content>\n  </span>\n</kendo-grid-column>\n",
                    styles: [""]
                },] },
    ];
    /** @nocollapse */
    GridColumnComponent.ctorParameters = function () { return [
        { type: ViewContainerRef }
    ]; };
    GridColumnComponent.propDecorators = {
        model: [{ type: Input }],
        templateRef: [{ type: Input }],
        class: [{ type: Input }],
        editable: [{ type: Input }],
        editor: [{ type: Input }],
        field: [{ type: Input }],
        filter: [{ type: Input }],
        filterable: [{ type: Input }],
        footerClass: [{ type: Input }],
        footerStyle: [{ type: Input }],
        format: [{ type: Input }],
        headerClass: [{ type: Input }],
        headerStyle: [{ type: Input }],
        hidden: [{ type: Input }],
        locked: [{ type: Input }],
        media: [{ type: Input }],
        sortable: [{ type: Input }],
        style: [{ type: Input }],
        title: [{ type: Input }],
        width: [{ type: Input }],
        content: [{ type: ViewChild, args: ['content',] }]
    };
    return GridColumnComponent;
}());
export { GridColumnComponent };
function GridColumnComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    GridColumnComponent.prototype.model;
    /** @type {?} */
    GridColumnComponent.prototype.templateRef;
    /** @type {?} */
    GridColumnComponent.prototype.class;
    /** @type {?} */
    GridColumnComponent.prototype.editable;
    /** @type {?} */
    GridColumnComponent.prototype.editor;
    /** @type {?} */
    GridColumnComponent.prototype.field;
    /** @type {?} */
    GridColumnComponent.prototype.filter;
    /** @type {?} */
    GridColumnComponent.prototype.filterable;
    /** @type {?} */
    GridColumnComponent.prototype.footerClass;
    /** @type {?} */
    GridColumnComponent.prototype.footerStyle;
    /** @type {?} */
    GridColumnComponent.prototype.format;
    /** @type {?} */
    GridColumnComponent.prototype.headerClass;
    /** @type {?} */
    GridColumnComponent.prototype.headerStyle;
    /** @type {?} */
    GridColumnComponent.prototype.hidden;
    /** @type {?} */
    GridColumnComponent.prototype.locked;
    /** @type {?} */
    GridColumnComponent.prototype.media;
    /** @type {?} */
    GridColumnComponent.prototype.sortable;
    /** @type {?} */
    GridColumnComponent.prototype.style;
    /** @type {?} */
    GridColumnComponent.prototype.title;
    /** @type {?} */
    GridColumnComponent.prototype.width;
    /** @type {?} */
    GridColumnComponent.prototype.content;
    /** @type {?} */
    GridColumnComponent.prototype.view;
}
