// @caiu/library/esm5/lib/components/grid/grid-column/grid-column.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { StyleModel } from '../../../shared/models';
/**
 * @template T
 */
var /**
 * @template T
 */
GridColumn = /** @class */ (function () {
    function GridColumn(name, label) {
        this.name = name;
        this.label = label;
        this.styleModel = new StyleModel();
        this.editable = true;
        this.editor = 'text';
        this.filter = 'text';
        this.filterable = true;
        this.hidden = false;
        this.locked = false;
        this.sortable = true;
        this.metadata = { name: name, label: label };
        this.title = label;
    }
    Object.defineProperty(GridColumn.prototype, "class", {
        get: /**
         * @return {?}
         */
        function () {
            return this._class;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._class = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GridColumn.prototype, "field", {
        get: /**
         * @return {?}
         */
        function () {
            return this._field ? this._field : this.metadata.name;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._field = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GridColumn.prototype, "title", {
        get: /**
         * @return {?}
         */
        function () {
            return this._title ? this._title : this.metadata.label;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._title = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GridColumn.prototype, "width", {
        get: /**
         * @return {?}
         */
        function () {
            return this._width;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._width = value;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} col
     * @return {?}
     */
    GridColumn.prototype.update = /**
     * @param {?} col
     * @return {?}
     */
    function (col) {
        this.class = col.class;
        this.editable = col.editable;
        this.editor = col.editor;
        this.field = col.field;
        this.filter = col.filter;
        this.filterable = col.filterable;
        this.footerClass = col.footerClass;
        this.footerStyle = col.footerStyle;
        this.format = col.format;
        this.headerClass = col.headerClass;
        this.headerStyle = col.headerStyle;
        this.hidden = col.hidden;
        this.locked = col.locked;
        this.media = col.media;
        this.sortable = col.sortable;
        this.style = col.style;
        this.title = col.title;
        this.width = col.width;
    };
    return GridColumn;
}());
/**
 * @template T
 */
export { GridColumn };
function GridColumn_tsickle_Closure_declarations() {
    /** @type {?} */
    GridColumn.prototype._class;
    /** @type {?} */
    GridColumn.prototype._field;
    /** @type {?} */
    GridColumn.prototype._title;
    /** @type {?} */
    GridColumn.prototype._width;
    /** @type {?} */
    GridColumn.prototype.styleModel;
    /** @type {?} */
    GridColumn.prototype.metadata;
    /** @type {?} */
    GridColumn.prototype.editable;
    /** @type {?} */
    GridColumn.prototype.editor;
    /** @type {?} */
    GridColumn.prototype.filter;
    /** @type {?} */
    GridColumn.prototype.filterable;
    /** @type {?} */
    GridColumn.prototype.footerClass;
    /** @type {?} */
    GridColumn.prototype.footerStyle;
    /** @type {?} */
    GridColumn.prototype.format;
    /** @type {?} */
    GridColumn.prototype.headerClass;
    /** @type {?} */
    GridColumn.prototype.headerStyle;
    /** @type {?} */
    GridColumn.prototype.hidden;
    /** @type {?} */
    GridColumn.prototype.locked;
    /** @type {?} */
    GridColumn.prototype.media;
    /** @type {?} */
    GridColumn.prototype.sortable;
    /** @type {?} */
    GridColumn.prototype.style;
    /** @type {?} */
    GridColumn.prototype.name;
    /** @type {?} */
    GridColumn.prototype.label;
}
