// @caiu/library/esm2015/lib/components/collage/collage.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, ChangeDetectionStrategy, Input } from '@angular/core';
import { Collage } from './collage.model';
export class CollageComponent {
    constructor() {
        this.images = [];
        this.canvasHeight = 0;
        this.canvasWidth = 0;
        this.totalRows = 0;
        this.totalColumns = 0;
        this.maxRows = 0;
        this.maxColumns = 0;
    }
    /**
     * @return {?}
     */
    get collage() {
        return this.canvasHeight > 0 && this.canvasWidth > 0 ?
            Collage.Build(this.images, this.canvasHeight, this.canvasWidth, this.totalRows, this.totalColumns, this.maxRows, this.maxColumns)
            : new Collage();
    }
    /**
     * @return {?}
     */
    get tiles() {
        return this.collage.tiles;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
}
CollageComponent.decorators = [
    { type: Component, args: [{
                selector: 'iu-collage',
                template: `<iu-tile *ngFor="let tile of tiles" [tile]="tile"></iu-tile>`,
                styles: [``],
                changeDetection: ChangeDetectionStrategy.OnPush
            },] },
];
/** @nocollapse */
CollageComponent.ctorParameters = () => [];
CollageComponent.propDecorators = {
    images: [{ type: Input }],
    canvasHeight: [{ type: Input }],
    canvasWidth: [{ type: Input }],
    totalRows: [{ type: Input }],
    totalColumns: [{ type: Input }],
    maxRows: [{ type: Input }],
    maxColumns: [{ type: Input }]
};
function CollageComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    CollageComponent.prototype.images;
    /** @type {?} */
    CollageComponent.prototype.canvasHeight;
    /** @type {?} */
    CollageComponent.prototype.canvasWidth;
    /** @type {?} */
    CollageComponent.prototype.totalRows;
    /** @type {?} */
    CollageComponent.prototype.totalColumns;
    /** @type {?} */
    CollageComponent.prototype.maxRows;
    /** @type {?} */
    CollageComponent.prototype.maxColumns;
}
