// @caiu/library/esm2015/lib/router/actions.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
export class RouterActions {
    /**
     * @param {?} routeName
     * @param {?=} parent
     * @return {?}
     */
    static activate(routeName, parent) {
        const /** @type {?} */ payload = {
            routeName,
            parent
        };
        return {
            type: RouterActions.ACTIVATE,
            payload
        };
    }
    /**
     * @param {?} r
     * @return {?}
     */
    static navigate(r) {
        return {
            type: RouterActions.NAVIGATE,
            payload: r
        };
    }
    /**
     * @return {?}
     */
    static navigateSuccess() {
        return {
            type: RouterActions.NAVIGATE_SUCCESS
        };
    }
    /**
     * @return {?}
     */
    static navigateError() {
        return {
            type: RouterActions.NAVIGATE_ERROR
        };
    }
    /**
     * @param {?} e
     * @return {?}
     */
    static navCancel(e) {
        return {
            type: RouterActions.NAVIGATION_CANCEL,
            payload: e
        };
    }
    /**
     * @param {?} e
     * @return {?}
     */
    static navEnd(e) {
        return {
            type: RouterActions.NAVIGATION_END,
            payload: e
        };
    }
    /**
     * @param {?} e
     * @return {?}
     */
    static navError(e) {
        return {
            type: RouterActions.NAVIGATION_ERROR,
            payload: e
        };
    }
    /**
     * @param {?} e
     * @return {?}
     */
    static navStart(e) {
        return {
            type: RouterActions.NAVIGATION_START,
            payload: e
        };
    }
    /**
     * @param {?} e
     * @return {?}
     */
    static routesRecognized(e) {
        return {
            type: RouterActions.ROUTES_RECOGNIZED,
            payload: e
        };
    }
}
RouterActions.ACTIVATE = '[Router] Activate Route';
RouterActions.NAVIGATE = '[Router] Navigate';
RouterActions.NAVIGATE_SUCCESS = '[Router] Navigate Success';
RouterActions.NAVIGATE_ERROR = '[Router] Navigate Error';
RouterActions.NAVIGATION_CANCEL = '[Router] Cancel Navigation';
RouterActions.NAVIGATION_END = '[Router] End Navigation';
RouterActions.NAVIGATION_ERROR = '[Router] Error Navigation';
RouterActions.NAVIGATION_START = '[Router] Start Navigation';
RouterActions.ROUTES_RECOGNIZED = '[Router] Routes Recognized';
function RouterActions_tsickle_Closure_declarations() {
    /** @type {?} */
    RouterActions.ACTIVATE;
    /** @type {?} */
    RouterActions.NAVIGATE;
    /** @type {?} */
    RouterActions.NAVIGATE_SUCCESS;
    /** @type {?} */
    RouterActions.NAVIGATE_ERROR;
    /** @type {?} */
    RouterActions.NAVIGATION_CANCEL;
    /** @type {?} */
    RouterActions.NAVIGATION_END;
    /** @type {?} */
    RouterActions.NAVIGATION_ERROR;
    /** @type {?} */
    RouterActions.NAVIGATION_START;
    /** @type {?} */
    RouterActions.ROUTES_RECOGNIZED;
}
