// @caiu/library/esm2015/lib/router/router.effects.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { Effect, Actions, ofType } from '@ngrx/effects';
import { Observable, of } from 'rxjs';
import { map, switchMap, catchError } from 'rxjs/operators';
import { RouterActions } from './actions';
import { RouterService } from './router.service';
export class RouterEffects {
    /**
     * @param {?} actions$
     * @param {?} router
     */
    constructor(actions$, router) {
        this.actions$ = actions$;
        this.router = router;
        /**
         * All imperative navigation should be performed using this effects
         * so that the state is aware of the navigation.
         */
        this.navigate = this.actions$.pipe(ofType(RouterActions.NAVIGATE), switchMap((action) => this.navigate$(action.payload)));
    }
    /**
     * @param {?} r
     * @return {?}
     */
    navigate$(r) {
        const /** @type {?} */ nav$ = typeof (r) === 'string' ? this.router.navigateByUrl(r) : this.router.navigate(r);
        return nav$.pipe(map(() => RouterActions.navigateSuccess()), catchError((e) => {
            return of(RouterActions.navigateError());
        }));
    }
}
RouterEffects.decorators = [
    { type: Injectable },
];
/** @nocollapse */
RouterEffects.ctorParameters = () => [
    { type: Actions },
    { type: RouterService }
];
tslib_1.__decorate([
    Effect(),
    tslib_1.__metadata("design:type", Observable)
], RouterEffects.prototype, "navigate", void 0);
function RouterEffects_tsickle_Closure_declarations() {
    /** @type {?} */
    RouterEffects.prototype.actionTypes;
    /**
     * All imperative navigation should be performed using this effects
     * so that the state is aware of the navigation.
     * @type {?}
     */
    RouterEffects.prototype.navigate;
    /** @type {?} */
    RouterEffects.prototype.actions$;
    /** @type {?} */
    RouterEffects.prototype.router;
}
