// @caiu/library/esm2015/lib/router/router.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Injectable, EventEmitter } from '@angular/core';
import { Router, ActivatedRoute, NavigationCancel, NavigationEnd, NavigationError, NavigationStart, RoutesRecognized } from '@angular/router';
import { Store } from '@ngrx/store';
import { from } from 'rxjs';
import { RouterActions } from './actions';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
import * as i2 from "@ngrx/store";
export class RouterService {
    /**
     * @param {?} _router
     * @param {?} _route
     * @param {?} store
     */
    constructor(_router, _route, store) {
        this._router = _router;
        this._route = _route;
        this.store = store;
        this.navigationCancel = new EventEmitter();
        this.navigationEnd = new EventEmitter();
        this.navigationError = new EventEmitter();
        this.navigationStart = new EventEmitter();
        this.routesRecognized = new EventEmitter();
        _router.events.subscribe((e) => {
            if (e instanceof NavigationCancel) {
                this.onNavigationCancel(e);
            }
            else if (e instanceof NavigationEnd) {
                this.onNavigationEnd(e);
            }
            else if (e instanceof NavigationError) {
                this.onNavigationError(e);
            }
            else if (e instanceof NavigationStart) {
                this.onNavigationStart(e);
            }
            else if (e instanceof RoutesRecognized) {
                this.onRoutesRecognized(e);
            }
        });
    }
    /**
     * @param {?} e
     * @return {?}
     */
    onNavigationCancel(e) {
        this.store.dispatch(RouterActions.navCancel(e));
    }
    /**
     * @param {?} e
     * @return {?}
     */
    onNavigationEnd(e) {
        this.store.dispatch(RouterActions.navEnd(e));
    }
    /**
     * @param {?} e
     * @return {?}
     */
    onNavigationError(e) {
        this.store.dispatch(RouterActions.navError(e));
    }
    /**
     * @param {?} e
     * @return {?}
     */
    onNavigationStart(e) {
        this.store.dispatch(RouterActions.navStart(e));
    }
    /**
     * @param {?} e
     * @return {?}
     */
    onRoutesRecognized(e) {
        this.store.dispatch(RouterActions.routesRecognized(e));
    }
    /**
     * @return {?}
     */
    get activatedRoute() {
        return this._route;
    }
    /**
     * @return {?}
     */
    get routeData$() {
        return this.activatedRoute.data;
    }
    /**
     * @return {?}
     */
    get routeParams$() {
        return this.activatedRoute.params;
    }
    /**
     * @return {?}
     */
    get router() {
        return this._router;
    }
    /**
     * @return {?}
     */
    get url$() {
        return this.store.select(s => s.router.url);
    }
    /**
     * @param {?} commands
     * @param {?=} extras
     * @return {?}
     */
    navigate(commands, extras) {
        const /** @type {?} */ promise = this.router.navigate(commands, extras);
        return from(promise);
    }
    /**
     * @param {?} url
     * @param {?=} extras
     * @return {?}
     */
    navigateByUrl(url, extras) {
        const /** @type {?} */ promise = this.router.navigateByUrl(url, extras);
        return from(promise);
    }
}
RouterService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] },
];
/** @nocollapse */
RouterService.ctorParameters = () => [
    { type: Router },
    { type: ActivatedRoute },
    { type: Store }
];
/** @nocollapse */ RouterService.ngInjectableDef = i0.defineInjectable({ factory: function RouterService_Factory() { return new RouterService(i0.inject(i1.Router), i0.inject(i1.ActivatedRoute), i0.inject(i2.Store)); }, token: RouterService, providedIn: "root" });
function RouterService_tsickle_Closure_declarations() {
    /** @type {?} */
    RouterService.prototype.navigationCancel;
    /** @type {?} */
    RouterService.prototype.navigationEnd;
    /** @type {?} */
    RouterService.prototype.navigationError;
    /** @type {?} */
    RouterService.prototype.navigationStart;
    /** @type {?} */
    RouterService.prototype.routesRecognized;
    /** @type {?} */
    RouterService.prototype._router;
    /** @type {?} */
    RouterService.prototype._route;
    /** @type {?} */
    RouterService.prototype.store;
}
