// @caiu/library/esm2015/lib/router/selectors.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { of, combineLatest } from 'rxjs';
import { map, distinctUntilChanged } from 'rxjs/operators';
import { Breadcrumbs } from './models';
import { QueryModel } from '../shared/models';
import { toArray, toInt } from '../shared/utils';
/**
 * @param {?} store
 * @return {?}
 */
export function routeSelector(store) {
    return store.select('route');
}
/**
 * @param {?} store
 * @return {?}
 */
export function breadcrumbsSelector(store) {
    return routeSelector(store).pipe(map(r => {
        const /** @type {?} */ outlets = r.activatedOutlets;
        return new Breadcrumbs(Object.keys(outlets).map(key => outlets[key]));
    }));
}
/**
 * @param {?} store
 * @return {?}
 */
export function routeNameSelector(store) {
    return routeSelector(store).pipe(map(r => r.routeName), distinctUntilChanged());
}
/**
 * @param {?} store
 * @return {?}
 */
export function routeParamsSelector(store) {
    return routeSelector(store).pipe(map(r => r.params), distinctUntilChanged());
}
/**
 * @param {?} store
 * @param {?} key
 * @return {?}
 */
export function lazyRouteParamsSelector(store, key) {
    return routeParamsSelector(store).pipe(map(params => params[key]), distinctUntilChanged());
}
/**
 * @param {?} store
 * @param {?} key
 * @param {?=} defaultValue
 * @return {?}
 */
export function routeParamSelector(store, key, defaultValue = null) {
    return combineLatest(of(defaultValue), lazyRouteParamsSelector(store, key), (x, y) => {
        return y || x;
    });
}
/**
 * @param {?} store
 * @param {?} key
 * @return {?}
 */
export function routeParamArraySelector(store, key) {
    return routeParamSelector(store, key, []).pipe(map(x => toArray(x)));
}
/**
 * @param {?} store
 * @param {?} key
 * @return {?}
 */
export function routeParamIdSelector(store, key) {
    return routeParamIntSelector(store, key);
}
/**
 * @param {?} store
 * @param {?} key
 * @return {?}
 */
export function routeParamIntSelector(store, key) {
    return routeParamSelector(store, key, 0).pipe(map(x => toInt(x)));
}
/**
 * @param {?} store
 * @return {?}
 */
export function urlSelector(store) {
    return routeSelector(store).pipe(map(r => r.url), distinctUntilChanged());
}
/**
 * @param {?} store
 * @return {?}
 */
export function querySelector(store) {
    const /** @type {?} */ skip$ = routeParamIntSelector(store, 'skip');
    const /** @type {?} */ take$ = routeParamIntSelector(store, 'take');
    const /** @type {?} */ term$ = routeParamSelector(store, 'term');
    return combineLatest(skip$, take$, term$, (skip, take, term) => (Object.assign(new QueryModel(), {
        skip,
        take,
        term
    })));
}
