// @caiu/library/esm5/lib/components/editor/editor-window/editor-window.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, Inject, ViewEncapsulation } from '@angular/core';
import { MAT_DIALOG_DATA, MatDialogRef } from '@angular/material';
import { DialogAction, DialogModel } from '../../dialog/dialog.model';
import { build } from '../../../shared/utils';
var EditorWindowComponent = /** @class */ (function () {
    function EditorWindowComponent(data, dialogRef) {
        this.data = data;
        this.dialogRef = dialogRef;
        this.actionWidth = '100%';
    }
    Object.defineProperty(EditorWindowComponent.prototype, "actions", {
        get: /**
         * @return {?}
         */
        function () {
            return [
                build(DialogAction, { value: this.data, label: 'Save and Close' }),
            ];
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(EditorWindowComponent.prototype, "dialog", {
        get: /**
         * @return {?}
         */
        function () {
            return build(DialogModel, {
                title: this.title,
                actions: this.actions
            });
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(EditorWindowComponent.prototype, "title", {
        get: /**
         * @return {?}
         */
        function () {
            return 'Editor';
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    EditorWindowComponent.prototype.close = /**
     * @return {?}
     */
    function () {
        this.dialogRef.close(this.data);
    };
    /**
     * @param {?} data
     * @return {?}
     */
    EditorWindowComponent.prototype.update = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        this.data = data;
    };
    EditorWindowComponent.decorators = [
        { type: Component, args: [{
                    selector: 'iu-editor-window',
                    template: "<span id=\"editor-window\">\n  <mat-toolbar color=\"primary\" mat-dialog-title>\n    <span class=\"dialog-title\">\n      {{ title }}\n      <button mat-button [mat-dialog-close]=\"null\">\n        <mat-icon>close</mat-icon>\n      </button>\n    </span>\n  </mat-toolbar>\n  <mat-dialog-content>\n    <div id=\"expanded-editor\">\n      <iu-editor [value]=\"data\" [expanded]=\"true\" [height]=\"380\" (changes)=\"update($event)\"></iu-editor>\n    </div>\n  </mat-dialog-content>\n  <mat-dialog-actions>\n    <button mat-raised-button (click)=\"close()\" [style.width]=\"actionWidth\" color=\"accent\">Save and Close</button>\n  </mat-dialog-actions>\n</span>\n",
                    styles: ["#editor-window #expanded-editor{margin:0;padding:0;height:60vh}#editor-window .mat-toolbar{margin:-24px -24px 0;width:calc(100% + 48px);min-height:32px!important}#editor-window .mat-toolbar .mat-toolbar-row{height:48px}#editor-window .mat-toolbar .dialog-title{position:relative;height:calc(100% - 8px);width:100%;margin-top:8px}#editor-window .mat-toolbar .dialog-title button{cursor:pointer;position:absolute;right:0;top:-5px;border-radius:100%;height:42px;width:42px;min-width:36px}#editor-window .mat-toolbar .dialog-title button:hover,#editor-window .mat-toolbar .dialog-title button:hover mat-icon{font-weight:700}#editor-window .mat-toolbar .dialog-title mat-icon{position:absolute;top:9px;right:8px;z-index:1}#editor-window .mat-toolbar-row{height:48px}#editor-window .mat-dialog-actions{margin-left:-24px;margin-right:-24px;color:#fff;padding:0}#editor-window .mat-dialog-actions button{padding:12px 0}#editor-window .mat-dialog-actions .mat-accent{color:#fff!important}"],
                    encapsulation: ViewEncapsulation.None
                },] },
    ];
    /** @nocollapse */
    EditorWindowComponent.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [MAT_DIALOG_DATA,] }] },
        { type: MatDialogRef }
    ]; };
    return EditorWindowComponent;
}());
export { EditorWindowComponent };
function EditorWindowComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    EditorWindowComponent.prototype.actionWidth;
    /** @type {?} */
    EditorWindowComponent.prototype.data;
    /** @type {?} */
    EditorWindowComponent.prototype.dialogRef;
}
