// @caiu/library/esm5/lib/shared/base-entity.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Metadata } from './models';
import { build } from './utils';
var BaseEntity = /** @class */ (function () {
    function BaseEntity() {
        this.createdBy = null;
        this.createdById = 0;
        this.createdDate = new Date();
        this.formId = 0;
        this.lastModifiedBy = null;
        this.lastModifiedById = 0;
        this.lastModifiedDate = new Date();
        this.matches = [];
        this.ignore = [
            '_metadata',
            'metadata',
            'ignore',
            'createdBy',
            'createdById',
            'createdDate',
            'formId',
            'lastModifiedBy',
            'lastModifiedById',
            'lastModifiedDate',
            'lastModifiedOn',
            'lastUpdated',
            'matches',
        ];
        this._metadata = new Metadata();
    }
    Object.defineProperty(BaseEntity.prototype, "metadata", {
        get: /**
         * @return {?}
         */
        function () {
            return build(Metadata, this._metadata, {
                ignore: this.ignore
            });
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._metadata = value;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    BaseEntity.getGetters = /**
     * @return {?}
     */
    function () {
        var _this = this;
        return Object.keys(this.prototype).filter(function (name) {
            return typeof Object.getOwnPropertyDescriptor(_this.prototype, name)['get'] === 'function';
        });
    };
    /**
     * @return {?}
     */
    BaseEntity.getSetters = /**
     * @return {?}
     */
    function () {
        var _this = this;
        return Object.keys(this.prototype).filter(function (name) {
            return typeof Object.getOwnPropertyDescriptor(_this.prototype, name)['set'] === 'function';
        });
    };
    /**
     * @return {?}
     */
    BaseEntity.prototype.getGetters = /**
     * @return {?}
     */
    function () {
        var _this = this;
        return Object.keys(this.constructor.prototype).filter(function (name) {
            return typeof Object.getOwnPropertyDescriptor(_this.constructor.prototype, name)['get'] === 'function';
        });
    };
    /**
     * @return {?}
     */
    BaseEntity.prototype.getSetters = /**
     * @return {?}
     */
    function () {
        var _this = this;
        return Object.keys(this.constructor.prototype).filter(function (name) {
            return typeof Object.getOwnPropertyDescriptor(_this.constructor.prototype, name)['set'] === 'function';
        });
    };
    /**
     * @param {?} value
     * @return {?}
     */
    BaseEntity.prototype.update = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        return Object.assign(new BaseEntity(), value);
    };
    return BaseEntity;
}());
export { BaseEntity };
function BaseEntity_tsickle_Closure_declarations() {
    /** @type {?} */
    BaseEntity.prototype.createdBy;
    /** @type {?} */
    BaseEntity.prototype.createdById;
    /** @type {?} */
    BaseEntity.prototype.createdDate;
    /** @type {?} */
    BaseEntity.prototype.formId;
    /** @type {?} */
    BaseEntity.prototype.lastModifiedBy;
    /** @type {?} */
    BaseEntity.prototype.lastModifiedById;
    /** @type {?} */
    BaseEntity.prototype.lastModifiedDate;
    /** @type {?} */
    BaseEntity.prototype.matches;
    /** @type {?} */
    BaseEntity.prototype.ignore;
    /** @type {?} */
    BaseEntity.prototype._metadata;
}
