// @caiu/library/esm5/lib/shared/collection.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import * as tslib_1 from "tslib";
import { arrayUnion, build, inArray } from './utils';
/**
 * @template T
 */
var /**
 * @template T
 */
Collection = /** @class */ (function () {
    function Collection(ctor) {
        this.activeId = 0;
        this.metadata = {};
        this._items = {};
        if (ctor) {
            this.ctor = ctor;
        }
    }
    /**
     * @template T
     * @param {?} state
     * @param {?} key
     * @param {?} item
     * @return {?}
     */
    Collection.AddItem = /**
     * @template T
     * @param {?} state
     * @param {?} key
     * @param {?} item
     * @return {?}
     */
    function (state, key, item) {
        var /** @type {?} */ newState = {};
        Object.keys(state)
            .forEach(function (k) {
            newState[k] = state[k];
        });
        newState[key] = item;
        return newState;
    };
    /**
     * @template T
     * @param {?} items
     * @param {?=} findByKey
     * @return {?}
     */
    Collection.BuildCollection = /**
     * @template T
     * @param {?} items
     * @param {?=} findByKey
     * @return {?}
     */
    function (items, findByKey) {
        if (findByKey === void 0) { findByKey = 'id'; }
        var /** @type {?} */ dictionary = Collection.BuildItems(items, findByKey);
        return Object.assign(new Collection(), { items: dictionary });
    };
    /**
     * @template T
     * @param {?} value
     * @param {?=} key
     * @return {?}
     */
    Collection.BuildDictionaryFromArray = /**
     * @template T
     * @param {?} value
     * @param {?=} key
     * @return {?}
     */
    function (value, key) {
        if (key === void 0) { key = 'id'; }
        var /** @type {?} */ obj = {};
        value.forEach(function (item) {
            obj[item[key]] = item;
        });
        return obj;
    };
    /**
     * @template T
     * @param {?} items
     * @param {?=} findByKey
     * @return {?}
     */
    Collection.BuildFromArray = /**
     * @template T
     * @param {?} items
     * @param {?=} findByKey
     * @return {?}
     */
    function (items, findByKey) {
        if (findByKey === void 0) { findByKey = 'id'; }
        var /** @type {?} */ dictionary = Collection.BuildItems(items, findByKey);
        return Object.assign(new Collection(), { items: dictionary });
    };
    /**
     * @template T
     * @param {?} items
     * @param {?=} findByKey
     * @return {?}
     */
    Collection.BuildItems = /**
     * @template T
     * @param {?} items
     * @param {?=} findByKey
     * @return {?}
     */
    function (items, findByKey) {
        if (findByKey === void 0) { findByKey = 'id'; }
        var /** @type {?} */ dictionary = {};
        items.forEach(function (item) {
            var /** @type {?} */ key = item[findByKey];
            dictionary[key] = item;
        });
        return dictionary;
    };
    /**
     * @template T
     * @param {?} item
     * @param {?} path
     * @return {?}
     */
    Collection.Filter = /**
     * @template T
     * @param {?} item
     * @param {?} path
     * @return {?}
     */
    function (item, path) {
        return Array.isArray(item['matches']) && inArray(item['matches'], path);
    };
    /**
     * @template T
     * @param {?} item
     * @param {?} paths
     * @return {?}
     */
    Collection.FilterAnd = /**
     * @template T
     * @param {?} item
     * @param {?} paths
     * @return {?}
     */
    function (item, paths) {
        return Array.isArray(item['matches']) && paths.every(function (path) { return inArray(item['matches'], path); });
    };
    /**
     * @template T
     * @param {?} item
     * @param {?} paths
     * @return {?}
     */
    Collection.FilterOr = /**
     * @template T
     * @param {?} item
     * @param {?} paths
     * @return {?}
     */
    function (item, paths) {
        return Array.isArray(item['matches']) && paths.some(function (path) { return inArray(item['matches'], path); });
    };
    /**
     * @template T
     * @param {?} state
     * @param {?} key
     * @return {?}
     */
    Collection.RemoveItem = /**
     * @template T
     * @param {?} state
     * @param {?} key
     * @return {?}
     */
    function (state, key) {
        var /** @type {?} */ newState = {};
        Object.keys(state)
            .filter(function (k) { return k !== key; })
            .forEach(function (k) {
            newState[k] = state[k];
        });
        return newState;
    };
    Object.defineProperty(Collection.prototype, "active", {
        get: /**
         * @return {?}
         */
        function () {
            var /** @type {?} */ item = this.items[this.activeId] || /** @type {?} */ ({});
            return this.ctor ? build(this.ctor, item) : item;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Collection.prototype, "ctor", {
        get: /**
         * @return {?}
         */
        function () {
            return this._ctor;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._ctor = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Collection.prototype, "count", {
        get: /**
         * @return {?}
         */
        function () {
            return this.toArray().length;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Collection.prototype, "defaultItem", {
        get: /**
         * @return {?}
         */
        function () {
            return this.ctor ? new this.ctor() : /** @type {?} */ ({});
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Collection.prototype, "empty", {
        get: /**
         * @return {?}
         */
        function () {
            return this.toArray().length === 0;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Collection.prototype, "instance", {
        get: /**
         * @return {?}
         */
        function () {
            var /** @type {?} */ activeId = this.activeId;
            var /** @type {?} */ ctor = this.ctor;
            var /** @type {?} */ lastUpdated = this.lastUpdated;
            var /** @type {?} */ metadata = this.metadata;
            // const collection = this.constructor();
            var /** @type {?} */ collection = new Collection();
            collection.activeId = activeId;
            collection.ctor = ctor;
            collection.lastUpdated = lastUpdated;
            collection.metadata = metadata;
            return collection;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Collection.prototype, "items", {
        get: /**
         * @return {?}
         */
        function () {
            return this._items;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._items = value;
            this.lastUpdated = new Date();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Collection.prototype, "keys", {
        get: /**
         * @return {?}
         */
        function () {
            return Object.keys(this.items);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Collection.prototype, "asArray", {
        get: /**
         * @return {?}
         */
        function () {
            return this.toArray();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Collection.prototype, "itemsArray", {
        get: /**
         * @return {?}
         */
        function () {
            return this.asArray;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    Collection.prototype.toArray = /**
     * @return {?}
     */
    function () {
        var _this = this;
        return Object.keys(this.items).map(function (key) { return (_this.items[key]); });
    };
    /**
     * @param {?} id
     * @param {?=} key
     * @return {?}
     */
    Collection.prototype.activate = /**
     * @param {?} id
     * @param {?=} key
     * @return {?}
     */
    function (id, key) {
        if (key === void 0) { key = 'id'; }
        var /** @type {?} */ items = this.items[id] ? this.copyItems() : Object.assign(this.copyItems(), (_a = {}, _a[id] = (_b = {}, _b[key] = id, _b), _a));
        return Object.assign(new Collection(), this, { activeId: id, items: items });
        var _a, _b;
    };
    /**
     * @param {?} item
     * @param {?=} key
     * @return {?}
     */
    Collection.prototype.addItem = /**
     * @param {?} item
     * @param {?=} key
     * @return {?}
     */
    function (item, key) {
        var /** @type {?} */ items = this.copyItems();
        return Object.assign(this.constructor(), this, {
            items: Object.assign(items, (_a = {}, _a[key] = item, _a))
        });
        var _a;
    };
    /**
     * @param {?} items
     * @param {?=} key
     * @return {?}
     */
    Collection.prototype.addItems = /**
     * @param {?} items
     * @param {?=} key
     * @return {?}
     */
    function (items, key) {
        var _this = this;
        if (key === void 0) { key = 'id'; }
        var /** @type {?} */ newItems = {};
        this.toArray().forEach(function (item) {
            newItems[item[key]] = item;
        });
        items.forEach(function (item) {
            var /** @type {?} */ existingItem = newItems[item[key]] ? newItems[item[key]] : /** @type {?} */ ({});
            var /** @type {?} */ existingMatches = existingItem['matches'] || [];
            var /** @type {?} */ newMatches = item['matches'] || [];
            var /** @type {?} */ matches = arrayUnion(existingMatches, newMatches);
            var /** @type {?} */ newItem = _this.buildItem(item, { matches: matches });
            newItems[item[key]] = _this.buildItem(existingItem, newItem);
        });
        this.items = newItems;
        return Object.assign(new Collection(), this);
    };
    /**
     * @param {?} items
     * @param {?=} ctor
     * @return {?}
     */
    Collection.prototype.build = /**
     * @param {?} items
     * @param {?=} ctor
     * @return {?}
     */
    function (items, ctor) {
        return ctor ? build(ctor, { items: items })
            : /** @type {?} */ (Object.assign(this.instance, { items: items }));
    };
    /**
     * @param {?} existingItem
     * @param {?} newItem
     * @param {?=} ctor
     * @return {?}
     */
    Collection.prototype.buildItem = /**
     * @param {?} existingItem
     * @param {?} newItem
     * @param {?=} ctor
     * @return {?}
     */
    function (existingItem, newItem, ctor) {
        var /** @type {?} */ existing = existingItem || {};
        var /** @type {?} */ item = ctor ? build(ctor, existing, newItem)
            : this.ctor ? build(this.ctor, existing, newItem)
                : Object.assign({}, existing, newItem);
        return item;
    };
    /**
     * @param {?} items
     * @param {?=} ctor
     * @return {?}
     */
    Collection.prototype.buildItems = /**
     * @param {?} items
     * @param {?=} ctor
     * @return {?}
     */
    function (items, ctor) {
        var /** @type {?} */ factory = ctor ? ctor : this.ctor;
        return items.map(function (x) { return factory ? build(factory, x) : Object.assign(/** @type {?} */ ({}), x); });
    };
    /**
     * @return {?}
     */
    Collection.prototype.copyItems = /**
     * @return {?}
     */
    function () {
        var _this = this;
        return Object.keys(this.items)
            .reduce(function (acc, key) {
            acc[key] = _this.get(key);
            return acc;
        }, {});
    };
    /**
     * @param {?} key
     * @return {?}
     */
    Collection.prototype.delete = /**
     * @param {?} key
     * @return {?}
     */
    function (key) {
        return this.removeAt(key);
    };
    /**
     * @param {?} f
     * @return {?}
     */
    Collection.prototype.filter = /**
     * @param {?} f
     * @return {?}
     */
    function (f) {
        var /** @type {?} */ items = Collection.BuildDictionaryFromArray(this.toArray().filter(f));
        return Object.assign(new Collection(), this, { items: items });
    };
    /**
     * @param {?} f
     * @return {?}
     */
    Collection.prototype.filterBy = /**
     * @param {?} f
     * @return {?}
     */
    function (f) {
        return this.filterItems(f);
    };
    /**
     * @param {?} f
     * @return {?}
     */
    Collection.prototype.filterItems = /**
     * @param {?} f
     * @return {?}
     */
    function (f) {
        return this.toArray().filter(f);
    };
    /**
     * @param {?} f
     * @return {?}
     */
    Collection.prototype.findBy = /**
     * @param {?} f
     * @return {?}
     */
    function (f) {
        return this.asArray.find(f) || this.defaultItem;
    };
    /**
     * @param {?} id
     * @return {?}
     */
    Collection.prototype.get = /**
     * @param {?} id
     * @return {?}
     */
    function (id) {
        var /** @type {?} */ existing = this.items[id];
        return this.ctor ? build(this.ctor, existing)
            : Object.assign(/** @type {?} */ ({}), existing);
    };
    /**
     * @param {?} f
     * @return {?}
     */
    Collection.prototype.map = /**
     * @param {?} f
     * @return {?}
     */
    function (f) {
        var /** @type {?} */ items = Collection.BuildDictionaryFromArray(this.toArray().map(f));
        return Object.assign(new Collection(), this, { items: items });
    };
    /**
     * @param {?} path
     * @return {?}
     */
    Collection.prototype.query = /**
     * @param {?} path
     * @return {?}
     */
    function (path) {
        return this.toArray().filter(function (item) { return Collection.Filter(item, path); });
    };
    /**
     * @param {?} paths
     * @return {?}
     */
    Collection.prototype.queryAnd = /**
     * @param {?} paths
     * @return {?}
     */
    function (paths) {
        return this.toArray().filter(function (item) { return Collection.FilterAnd(item, paths); });
    };
    /**
     * @param {?} paths
     * @return {?}
     */
    Collection.prototype.queryOr = /**
     * @param {?} paths
     * @return {?}
     */
    function (paths) {
        return this.toArray().filter(function (item) { return Collection.FilterOr(item, paths); });
    };
    /**
     * @param {?} key
     * @return {?}
     */
    Collection.prototype.removeAt = /**
     * @param {?} key
     * @return {?}
     */
    function (key) {
        var _this = this;
        var /** @type {?} */ keys = Object.keys(this.items);
        var /** @type {?} */ items = Collection.BuildDictionaryFromArray(keys.filter(function (x) { return x.toString() !== key.toString(); }).map(function (y) { return _this.items[y]; }));
        return Object.assign(new Collection(), this, { items: items });
    };
    /**
     * @param {?} key
     * @return {?}
     */
    Collection.prototype.removeItem = /**
     * @param {?} key
     * @return {?}
     */
    function (key) {
        var _this = this;
        var /** @type {?} */ collection = this.constructor();
        var /** @type {?} */ keys = Object.keys(this.items).filter(function (k) { return k !== key; });
        keys.forEach(function (k) {
            collection.items[k] = /** @type {?} */ (Object.assign({}, _this.items[k]));
        });
        return collection;
    };
    /**
     * @param {?} filter
     * @param {?=} key
     * @return {?}
     */
    Collection.prototype.removeItems = /**
     * @param {?} filter
     * @param {?=} key
     * @return {?}
     */
    function (filter, key) {
        if (key === void 0) { key = 'id'; }
        var /** @type {?} */ keys = this.toArray().filter(filter).map(function (x) { return x[key]; });
        this.removeKeys(keys);
    };
    /**
     * @param {?} keys
     * @return {?}
     */
    Collection.prototype.removeKeys = /**
     * @param {?} keys
     * @return {?}
     */
    function (keys) {
        var _this = this;
        keys.forEach(function (key) {
            _this.removeItem(key);
        });
    };
    /**
     * @param {?} itemsArray
     * @return {?}
     */
    Collection.prototype.replace = /**
     * @param {?} itemsArray
     * @return {?}
     */
    function (itemsArray) {
        var /** @type {?} */ items = Collection.BuildDictionaryFromArray(this.buildItems(itemsArray));
        return Object.assign(new Collection(), this, { items: items });
    };
    /**
     * @param {?} items
     * @param {?=} filter
     * @return {?}
     */
    Collection.prototype.replaceItems = /**
     * @param {?} items
     * @param {?=} filter
     * @return {?}
     */
    function (items, filter) {
        if (filter) {
            var /** @type {?} */ f = function (x) { return !filter(x); };
            return this.replaceItems(tslib_1.__spread(this.filterItems(f), items));
        }
        return Object.assign(new Collection(), this, { items: items });
    };
    /**
     * @param {?} value
     * @return {?}
     */
    Collection.prototype.setValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        return Object.assign(new Collection(), this, value);
    };
    /**
     * @param {?} value
     * @param {?=} ctor
     * @param {?=} key
     * @return {?}
     */
    Collection.prototype.update = /**
     * @param {?} value
     * @param {?=} ctor
     * @param {?=} key
     * @return {?}
     */
    function (value, ctor, key) {
        if (key === void 0) { key = 'id'; }
        if (Array.isArray(value)) {
            return this.updateItems(value, ctor, key);
        }
        return this.updateItem(value, value[key], ctor);
    };
    /**
     * @param {?} item
     * @param {?} key
     * @param {?=} ctor
     * @return {?}
     */
    Collection.prototype.updateItem = /**
     * @param {?} item
     * @param {?} key
     * @param {?=} ctor
     * @return {?}
     */
    function (item, key, ctor) {
        var /** @type {?} */ items = this.copyItems();
        var /** @type {?} */ updatedItem = this.buildItem(items[key], item, ctor);
        return this.build(Object.assign(items, (_a = {}, _a[key] = updatedItem, _a)));
        var _a;
    };
    /**
     * @param {?} items
     * @param {?=} ctor
     * @param {?=} key
     * @return {?}
     */
    Collection.prototype.updateItems = /**
     * @param {?} items
     * @param {?=} ctor
     * @param {?=} key
     * @return {?}
     */
    function (items, ctor, key) {
        var _this = this;
        if (key === void 0) { key = 'id'; }
        var /** @type {?} */ existingItems = this.copyItems();
        var /** @type {?} */ updatedItems = items.reduce(function (acc, item) {
            var /** @type {?} */ existingItem = acc[item[key]];
            acc[item[key]] = _this.buildItem(existingItem, item, ctor);
            return acc;
        }, existingItems);
        return this.build(updatedItems);
    };
    return Collection;
}());
/**
 * @template T
 */
export { Collection };
function Collection_tsickle_Closure_declarations() {
    /** @type {?} */
    Collection.prototype.activeId;
    /** @type {?} */
    Collection.prototype.lastUpdated;
    /** @type {?} */
    Collection.prototype.metadata;
    /** @type {?} */
    Collection.prototype._ctor;
    /** @type {?} */
    Collection.prototype._items;
}
