// @caiu/library/esm5/lib/shared/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import * as tslib_1 from "tslib";
import { getValue, truthy } from './utils';
var DumbComponent = /** @class */ (function () {
    function DumbComponent() {
        this.requestState = 'DEFAULT';
        this.subscriptions = [];
    }
    Object.defineProperty(DumbComponent.prototype, "inErrorState", {
        get: /**
         * @return {?}
         */
        function () {
            return this.requestState === 'ERROR';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DumbComponent.prototype, "inSuccessState", {
        get: /**
         * @return {?}
         */
        function () {
            return this.requestState === 'SUCCESS';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DumbComponent.prototype, "message", {
        get: /**
         * @return {?}
         */
        function () {
            return this.inErrorState ? "An error has occurred. Please try again later."
                : this.inSuccessState ? "Saved successfully!" : '';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DumbComponent.prototype, "showMessage", {
        get: /**
         * @return {?}
         */
        function () {
            return this.inErrorState || this.inSuccessState;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    DumbComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.removeSubscriptions();
    };
    /**
     * @param {?} subscription
     * @return {?}
     */
    DumbComponent.prototype.addSubscription = /**
     * @param {?} subscription
     * @return {?}
     */
    function (subscription) {
        this.subscriptions.push(subscription);
    };
    /**
     * @param {?} result
     * @return {?}
     */
    DumbComponent.prototype.closeDialog = /**
     * @param {?} result
     * @return {?}
     */
    function (result) {
        this.dialogRef.unsubscribe();
    };
    /**
     * @param {?} component
     * @param {?=} config
     * @return {?}
     */
    DumbComponent.prototype.openDialog = /**
     * @param {?} component
     * @param {?=} config
     * @return {?}
     */
    function (component, config) {
        var _this = this;
        if (config === void 0) { config = {}; }
        var /** @type {?} */ dialogRef = this.dialog.open(component, config);
        this.dialogRef = dialogRef.afterClosed().subscribe(function (result) {
            _this.closeDialog(result);
        });
    };
    /**
     * @return {?}
     */
    DumbComponent.prototype.removeSubscriptions = /**
     * @return {?}
     */
    function () {
        this.subscriptions.forEach(function (s) {
            s.unsubscribe();
        });
    };
    /**
     * @param {?} subscriptions
     * @return {?}
     */
    DumbComponent.prototype.subscribe = /**
     * @param {?} subscriptions
     * @return {?}
     */
    function (subscriptions) {
        var _this = this;
        subscriptions.forEach(function (s) {
            _this.addSubscription(s);
        });
    };
    /**
     * @param {?=} duration
     * @return {?}
     */
    DumbComponent.prototype.flashErrorMessage = /**
     * @param {?=} duration
     * @return {?}
     */
    function (duration) {
        var _this = this;
        if (duration === void 0) { duration = 5000; }
        this.toErrorState();
        setTimeout(function () {
            _this.resetRequestState();
        }, duration);
    };
    /**
     * @param {?=} duration
     * @return {?}
     */
    DumbComponent.prototype.flashSuccessMessage = /**
     * @param {?=} duration
     * @return {?}
     */
    function (duration) {
        var _this = this;
        if (duration === void 0) { duration = 5000; }
        this.toSucessState();
        setTimeout(function () {
            _this.resetRequestState();
        }, duration);
    };
    /**
     * @param {?} e
     * @return {?}
     */
    DumbComponent.prototype.onError = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        this.flashErrorMessage();
    };
    /**
     * @param {?} e
     * @return {?}
     */
    DumbComponent.prototype.onSuccess = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        this.flashSuccessMessage();
    };
    /**
     * @return {?}
     */
    DumbComponent.prototype.resetRequestState = /**
     * @return {?}
     */
    function () {
        this.requestState = 'DEFAULT';
    };
    /**
     * @param {?} value
     * @return {?}
     */
    DumbComponent.prototype.setValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        if (this.form && this.form.setValue) {
            this.form.setValue(getValue(value));
        }
    };
    /**
     * @return {?}
     */
    DumbComponent.prototype.toErrorState = /**
     * @return {?}
     */
    function () {
        this.requestState = 'ERROR';
    };
    /**
     * @return {?}
     */
    DumbComponent.prototype.toSucessState = /**
     * @return {?}
     */
    function () {
        this.requestState = 'SUCCESS';
    };
    return DumbComponent;
}());
export { DumbComponent };
function DumbComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    DumbComponent.prototype.dialog;
    /** @type {?} */
    DumbComponent.prototype.dialogRef;
    /** @type {?} */
    DumbComponent.prototype.form;
    /** @type {?} */
    DumbComponent.prototype.requestState;
    /** @type {?} */
    DumbComponent.prototype.subscriptions;
}
var FormComponent = /** @class */ (function (_super) {
    tslib_1.__extends(FormComponent, _super);
    function FormComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(FormComponent.prototype, "editing", {
        get: /**
         * @return {?}
         */
        function () {
            return truthy(this.id);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FormComponent.prototype, "id", {
        get: /**
         * @return {?}
         */
        function () {
            return this.model.id;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FormComponent.prototype, "isValid", {
        get: /**
         * @return {?}
         */
        function () {
            return this.form.valid;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    FormComponent.prototype.markAsSubmitted = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.form.markAsTouched();
        Object.keys(this.form.controls).forEach(function (key) {
            _this.form.controls[key].markAsTouched();
        });
    };
    return FormComponent;
}(DumbComponent));
export { FormComponent };
function FormComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    FormComponent.prototype.form;
    /** @type {?} */
    FormComponent.prototype.model;
}
var SmartComponent = /** @class */ (function (_super) {
    tslib_1.__extends(SmartComponent, _super);
    function SmartComponent(store) {
        var _this = _super.call(this) || this;
        _this.store = store;
        return _this;
    }
    /**
     * @param {?} action
     * @return {?}
     */
    SmartComponent.prototype.dispatch = /**
     * @param {?} action
     * @return {?}
     */
    function (action) {
        this.store.dispatch(action);
    };
    /**
     * @param {?} action
     * @param {?=} onSuccess
     * @param {?=} onError
     * @return {?}
     */
    SmartComponent.prototype.dispatchAndSubscribe = /**
     * @param {?} action
     * @param {?=} onSuccess
     * @param {?=} onError
     * @return {?}
     */
    function (action, onSuccess, onError) {
        var _this = this;
        var /** @type {?} */ f1 = onSuccess ? onSuccess : function (e) {
            _this.flashSuccessMessage();
        };
        var /** @type {?} */ f2 = onError ? onError : function (e) {
            _this.flashErrorMessage();
        };
        if (this.events && this.events.dispatch) {
            this.addSubscription(this.events.dispatch(action).subscribe(f1, f2));
        }
        else {
            this.store.dispatch(action);
        }
    };
    return SmartComponent;
}(DumbComponent));
export { SmartComponent };
function SmartComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    SmartComponent.prototype.events;
    /** @type {?} */
    SmartComponent.prototype.store;
}
