// @caiu/library/esm5/lib/shared/config.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import * as tslib_1 from "tslib";
var Config = /** @class */ (function () {
    function Config() {
        this.dev = false;
        this.test = false;
        this.staging = false;
        this.production = false;
        this.apiBaseUrl = '';
    }
    return Config;
}());
export { Config };
function Config_tsickle_Closure_declarations() {
    /** @type {?} */
    Config.prototype.dev;
    /** @type {?} */
    Config.prototype.test;
    /** @type {?} */
    Config.prototype.staging;
    /** @type {?} */
    Config.prototype.production;
    /** @type {?} */
    Config.prototype.apiBaseUrl;
}
var ViewConfig = /** @class */ (function (_super) {
    tslib_1.__extends(ViewConfig, _super);
    function ViewConfig() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.viewMode = 'Default';
        return _this;
    }
    return ViewConfig;
}(Config));
export { ViewConfig };
function ViewConfig_tsickle_Closure_declarations() {
    /** @type {?} */
    ViewConfig.prototype.viewMode;
}
