// @caiu/library/esm5/lib/shared/question.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Collection } from './collection';
import { DateHelper } from './date';
import { build } from './utils';
var Question = /** @class */ (function () {
    function Question() {
        this.question = '';
        this.title = '';
    }
    return Question;
}());
export { Question };
function Question_tsickle_Closure_declarations() {
    /** @type {?} */
    Question.prototype.answer;
    /** @type {?} */
    Question.prototype.lookup;
    /** @type {?} */
    Question.prototype.key;
    /** @type {?} */
    Question.prototype.question;
    /** @type {?} */
    Question.prototype.title;
    /** @type {?} */
    Question.prototype.type;
}
var Questions = /** @class */ (function (_super) {
    tslib_1.__extends(Questions, _super);
    function Questions() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @param {?} questions
     * @param {?} answers
     * @return {?}
     */
    Questions.AnswerQuestions = /**
     * @param {?} questions
     * @param {?} answers
     * @return {?}
     */
    function (questions, answers) {
        return questions.map(function (question) { return Questions.AnswerQuestion(question, answers[question.key]); });
    };
    /**
     * @param {?} question
     * @param {?} answerData
     * @return {?}
     */
    Questions.AnswerQuestion = /**
     * @param {?} question
     * @param {?} answerData
     * @return {?}
     */
    function (question, answerData) {
        var /** @type {?} */ answer = answerData;
        switch (question.type) {
            case 'date':
                answer = answerData;
                break;
            case 'monthyear':
                answer = DateHelper.ToMonthYearName(answerData);
                break;
            case 'yesno':
                answer = answerData ? 'Yes' : 'No';
                break;
        }
        return build(Question, question, { answer: answer });
    };
    /**
     * @param {?} answers
     * @return {?}
     */
    Questions.prototype.answerQuestions = /**
     * @param {?} answers
     * @return {?}
     */
    function (answers) {
        return Questions.AnswerQuestions(this.toArray(), answers);
    };
    return Questions;
}(Collection));
export { Questions };
/**
 * @abstract
 */
var /**
 * @abstract
 */
QuestionsModel = /** @class */ (function () {
    function QuestionsModel() {
        this.ignore = ['isAdd', 'questions'];
    }
    Object.defineProperty(QuestionsModel.prototype, "answers", {
        get: /**
         * @return {?}
         */
        function () {
            return Questions.AnswerQuestions(this.questions, this);
        },
        enumerable: true,
        configurable: true
    });
    return QuestionsModel;
}());
/**
 * @abstract
 */
export { QuestionsModel };
function QuestionsModel_tsickle_Closure_declarations() {
    /** @type {?} */
    QuestionsModel.prototype.ignore;
    /** @type {?} */
    QuestionsModel.prototype.questions;
}
