// @caiu/library/esm5/lib/shared/token.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var Token = /** @class */ (function () {
    function Token() {
        this.access_token = '';
        this.expiresIn = 0;
    }
    Object.defineProperty(Token.prototype, "authenticated", {
        get: /**
         * @return {?}
         */
        function () {
            return !this.expired;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Token.prototype, "expires_in", {
        get: /**
         * @return {?}
         */
        function () {
            return this.expiresIn;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.expiresIn = value;
            this.expiresInDate = new Date();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Token.prototype, "expirationDate", {
        get: /**
         * @return {?}
         */
        function () {
            var /** @type {?} */ t = new Date(this.expiresInDate);
            t.setSeconds(t.getSeconds() + this.expiresIn);
            return t;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Token.prototype, "expired", {
        get: /**
         * @return {?}
         */
        function () {
            var /** @type {?} */ now = new Date();
            return this.access_token && this.expiresIn && this.expiresInDate ? now > this.expirationDate : true;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    Token.prototype.toString = /**
     * @return {?}
     */
    function () {
        return this.expired ? '' : this.access_token;
    };
    return Token;
}());
export { Token };
function Token_tsickle_Closure_declarations() {
    /** @type {?} */
    Token.prototype.access_token;
    /** @type {?} */
    Token.prototype.expiresIn;
    /** @type {?} */
    Token.prototype.expiresInDate;
}
