// @caiu/library/esm5/lib/shared/user.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Token } from './token';
import { build } from './utils';
var User = /** @class */ (function () {
    function User() {
        this.email = '';
        this.firstName = '';
        this.lastName = '';
        this.fullName = '';
        this.password = '';
        this.confirmPassword = '';
        this.userName = '';
    }
    return User;
}());
export { User };
function User_tsickle_Closure_declarations() {
    /** @type {?} */
    User.prototype.email;
    /** @type {?} */
    User.prototype.firstName;
    /** @type {?} */
    User.prototype.lastName;
    /** @type {?} */
    User.prototype.fullName;
    /** @type {?} */
    User.prototype.password;
    /** @type {?} */
    User.prototype.confirmPassword;
    /** @type {?} */
    User.prototype.userName;
}
var CurrentUser = /** @class */ (function (_super) {
    tslib_1.__extends(CurrentUser, _super);
    function CurrentUser() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.token = new Token();
        return _this;
    }
    /**
     * @param {?} data
     * @return {?}
     */
    CurrentUser.Build = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        var /** @type {?} */ token = data && data.token ? build(Token, data.token) : new Token();
        return build(CurrentUser, data, { token: token });
    };
    Object.defineProperty(CurrentUser.prototype, "authenticated", {
        get: /**
         * @return {?}
         */
        function () {
            return this.token && this.token.authenticated;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CurrentUser.prototype, "userRoles", {
        get: /**
         * @return {?}
         */
        function () {
            return new UserRoles(this);
        },
        enumerable: true,
        configurable: true
    });
    return CurrentUser;
}(User));
export { CurrentUser };
function CurrentUser_tsickle_Closure_declarations() {
    /** @type {?} */
    CurrentUser.prototype.token;
}
var UserRoles = /** @class */ (function () {
    function UserRoles(user) {
        if (user === void 0) { user = new CurrentUser(); }
        this.user = user;
    }
    return UserRoles;
}());
export { UserRoles };
function UserRoles_tsickle_Closure_declarations() {
    /** @type {?} */
    UserRoles.prototype.user;
}
var Login = /** @class */ (function () {
    function Login() {
        this.grant_type = 'password';
        this.userName = '';
        this.password = '';
    }
    return Login;
}());
export { Login };
function Login_tsickle_Closure_declarations() {
    /** @type {?} */
    Login.prototype.grant_type;
    /** @type {?} */
    Login.prototype.userName;
    /** @type {?} */
    Login.prototype.password;
}
var ResetPassword = /** @class */ (function () {
    function ResetPassword() {
        this.passwordResetCode = '';
        this.password = '';
        this.confirmPassword = '';
    }
    return ResetPassword;
}());
export { ResetPassword };
function ResetPassword_tsickle_Closure_declarations() {
    /** @type {?} */
    ResetPassword.prototype.passwordResetCode;
    /** @type {?} */
    ResetPassword.prototype.password;
    /** @type {?} */
    ResetPassword.prototype.confirmPassword;
}
