// @caiu/library/esm5/lib/shared/window.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var Window = /** @class */ (function () {
    function Window() {
        this.windowHeight = 0;
        this.windowWidth = 0;
    }
    Object.defineProperty(Window.prototype, "windowWidthSmall", {
        get: /**
         * @return {?}
         */
        function () {
            return this.windowWidth < 1200;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Window.prototype, "windowWidthMedium", {
        get: /**
         * @return {?}
         */
        function () {
            return !this.windowWidthSmall && this.windowWidth < 1300;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Window.prototype, "windowWidthLarge", {
        get: /**
         * @return {?}
         */
        function () {
            return this.windowWidth >= 1300;
        },
        enumerable: true,
        configurable: true
    });
    return Window;
}());
export { Window };
function Window_tsickle_Closure_declarations() {
    /** @type {?} */
    Window.prototype.windowHeight;
    /** @type {?} */
    Window.prototype.windowWidth;
}
var WindowResize = /** @class */ (function () {
    function WindowResize() {
        this.windowHeight = 0;
        this.windowWidth = 0;
    }
    return WindowResize;
}());
export { WindowResize };
function WindowResize_tsickle_Closure_declarations() {
    /** @type {?} */
    WindowResize.prototype.windowHeight;
    /** @type {?} */
    WindowResize.prototype.windowWidth;
}
