// @caiu/library/esm5/lib/components/collage/collage.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, ChangeDetectionStrategy, Input } from '@angular/core';
import { Collage } from './collage.model';
var CollageComponent = /** @class */ (function () {
    function CollageComponent() {
        this.images = [];
        this.canvasHeight = 0;
        this.canvasWidth = 0;
        this.totalRows = 0;
        this.totalColumns = 0;
        this.maxRows = 0;
        this.maxColumns = 0;
    }
    Object.defineProperty(CollageComponent.prototype, "collage", {
        get: /**
         * @return {?}
         */
        function () {
            return this.canvasHeight > 0 && this.canvasWidth > 0 ?
                Collage.Build(this.images, this.canvasHeight, this.canvasWidth, this.totalRows, this.totalColumns, this.maxRows, this.maxColumns)
                : new Collage();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CollageComponent.prototype, "tiles", {
        get: /**
         * @return {?}
         */
        function () {
            return this.collage.tiles;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    CollageComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    CollageComponent.decorators = [
        { type: Component, args: [{
                    selector: 'iu-collage',
                    template: "<iu-tile *ngFor=\"let tile of tiles\" [tile]=\"tile\"></iu-tile>",
                    styles: [""],
                    changeDetection: ChangeDetectionStrategy.OnPush
                },] },
    ];
    /** @nocollapse */
    CollageComponent.ctorParameters = function () { return []; };
    CollageComponent.propDecorators = {
        images: [{ type: Input }],
        canvasHeight: [{ type: Input }],
        canvasWidth: [{ type: Input }],
        totalRows: [{ type: Input }],
        totalColumns: [{ type: Input }],
        maxRows: [{ type: Input }],
        maxColumns: [{ type: Input }]
    };
    return CollageComponent;
}());
export { CollageComponent };
function CollageComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    CollageComponent.prototype.images;
    /** @type {?} */
    CollageComponent.prototype.canvasHeight;
    /** @type {?} */
    CollageComponent.prototype.canvasWidth;
    /** @type {?} */
    CollageComponent.prototype.totalRows;
    /** @type {?} */
    CollageComponent.prototype.totalColumns;
    /** @type {?} */
    CollageComponent.prototype.maxRows;
    /** @type {?} */
    CollageComponent.prototype.maxColumns;
}
