// @caiu/library/esm2015/lib/storage/storage.effects.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { Effect, ofType, Actions } from '@ngrx/effects';
import { Store } from '@ngrx/store';
import { Observable } from 'rxjs';
import { map, filter, withLatestFrom } from 'rxjs/operators';
import { StorageActions } from './storage.actions';
import { StorageService } from './storage.service';
import { toPayload } from '../shared/utils';
export class StorageEffects {
    /**
     * @param {?} actions$
     * @param {?} service
     * @param {?} state$
     */
    constructor(actions$, service, state$) {
        this.actions$ = actions$;
        this.service = service;
        this.state$ = state$;
        /**
         * Listens for dispatched actions, updates local storage in response.
         */
        this.onSyncLocalStorage = this.actions$.pipe(filter((action) => this.service.inLocalStorage(action.type)), map(action => ({
            type: StorageActions.UPDATE_LOCAL_STORAGE,
            payload: action
        })));
        /**
         * Listens for dispatched actions, updates session storage in response.
         */
        this.onSyncSessionStorage = this.actions$.pipe(filter((action) => this.service.inSessionStorage(action.type)), map(action => ({
            type: StorageActions.UPDATE_SESSION_STORAGE,
            payload: action
        })));
        /**
         * Stores state to local storage after each action.
         */
        this.onStoreLocal = this.actions$.pipe(ofType(StorageActions.UPDATE_LOCAL_STORAGE), map(toPayload), withLatestFrom(this.state$, (action, state) => {
            this.service.storeLocal(state, action);
        }));
        /**
         * Stores state to local storage after each action.
         */
        this.onStoreSession = this.actions$.pipe(ofType(StorageActions.UPDATE_SESSION_STORAGE), map(toPayload), withLatestFrom(this.state$, (action, state) => {
            this.service.storeSession(state, action);
        }));
    }
}
StorageEffects.decorators = [
    { type: Injectable },
];
/** @nocollapse */
StorageEffects.ctorParameters = () => [
    { type: Actions },
    { type: StorageService },
    { type: Store }
];
tslib_1.__decorate([
    Effect(),
    tslib_1.__metadata("design:type", Observable)
], StorageEffects.prototype, "onSyncLocalStorage", void 0);
tslib_1.__decorate([
    Effect(),
    tslib_1.__metadata("design:type", Observable)
], StorageEffects.prototype, "onSyncSessionStorage", void 0);
tslib_1.__decorate([
    Effect({ dispatch: false }),
    tslib_1.__metadata("design:type", Observable)
], StorageEffects.prototype, "onStoreLocal", void 0);
tslib_1.__decorate([
    Effect({ dispatch: false }),
    tslib_1.__metadata("design:type", Observable)
], StorageEffects.prototype, "onStoreSession", void 0);
function StorageEffects_tsickle_Closure_declarations() {
    /** @type {?} */
    StorageEffects.prototype.actionTypes;
    /**
     * Listens for dispatched actions, updates local storage in response.
     * @type {?}
     */
    StorageEffects.prototype.onSyncLocalStorage;
    /**
     * Listens for dispatched actions, updates session storage in response.
     * @type {?}
     */
    StorageEffects.prototype.onSyncSessionStorage;
    /**
     * Stores state to local storage after each action.
     * @type {?}
     */
    StorageEffects.prototype.onStoreLocal;
    /**
     * Stores state to local storage after each action.
     * @type {?}
     */
    StorageEffects.prototype.onStoreSession;
    /** @type {?} */
    StorageEffects.prototype.actions$;
    /** @type {?} */
    StorageEffects.prototype.service;
    /** @type {?} */
    StorageEffects.prototype.state$;
}
