// @caiu/library/esm2015/lib/components/file-upload/file-upload.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { build } from '../../shared/utils';
export class FileUpload {
    constructor() {
        this.lastModified = 0;
        this.lastModifiedDate = new Date();
        this.name = '';
        this.size = 0;
        this.type = '';
        this.webkitRelativePath = '';
        this.order = 0;
        this.readyState = 'EMPTY';
        this.src = '';
    }
    /**
     * @param {?} file
     * @return {?}
     */
    static BuildFromFile(file) {
        return build(FileUpload, {
            name: file.fileName,
            size: file.fileSize,
            type: file.mimeType,
            src: file.src,
            order: file.order
        });
    }
    /**
     * @param {?} reader
     * @return {?}
     */
    static GetReadyState(reader) {
        switch (reader.readyState) {
            case 0:
                return 'EMPTY';
            case 1:
                return 'LOADING';
            case 2:
                return 'DONE';
            default:
                return 'EMPTY';
        }
    }
    /**
     * @return {?}
     */
    get extension() {
        return this.name.split('.').pop();
    }
    /**
     * @return {?}
     */
    get isImage() {
        return this.type.split('/')[0] === 'image';
    }
    /**
     * @return {?}
     */
    get loading() {
        return this.readyState === 'LOADING';
    }
    /**
     * @return {?}
     */
    get file() {
        return File.BuildFromFileUpload(this);
    }
}
function FileUpload_tsickle_Closure_declarations() {
    /** @type {?} */
    FileUpload.prototype.lastModified;
    /** @type {?} */
    FileUpload.prototype.lastModifiedDate;
    /** @type {?} */
    FileUpload.prototype.name;
    /** @type {?} */
    FileUpload.prototype.size;
    /** @type {?} */
    FileUpload.prototype.type;
    /** @type {?} */
    FileUpload.prototype.webkitRelativePath;
    /** @type {?} */
    FileUpload.prototype.order;
    /** @type {?} */
    FileUpload.prototype.readyState;
    /** @type {?} */
    FileUpload.prototype.src;
}
/**
 * @param {?} reader
 * @return {?}
 */
export function getReadyState(reader) {
    switch (reader.readyState) {
        case 0:
            return 'EMPTY';
        case 1:
            return 'LOADING';
        case 2:
            return 'DONE';
        default:
            return 'EMPTY';
    }
}
export class File {
    constructor() {
        this.id = 0;
        this.fileBinary = [];
        this.fileExtension = '';
        this.fileName = '';
        this.fileSize = 0;
        this.mimeType = '';
        this.order = 0;
    }
    /**
     * @param {?} upload
     * @return {?}
     */
    static BuildFromFileUpload(upload) {
        return build(File, {
            fileBinary: upload.src.replace(File.GetSrcPrefix(upload.type), ''),
            fileExtension: upload.extension,
            fileName: upload.name,
            fileSize: upload.size,
            mimeType: upload.type,
            order: upload.order
        });
    }
    /**
     * @param {?} img
     * @param {?=} defaultSrc
     * @return {?}
     */
    static GetImageBinarySrc(img, defaultSrc = '') {
        if (Array.isArray(img)) {
            if (img.length > 0) {
                return `data:${img[0].mimeType};base64,${img[0].fileBinary}`;
            }
            else {
                return defaultSrc;
            }
        }
        return img && img.mimeType && img.fileBinary ? `data:${img.mimeType};base64,${img.fileBinary}` : defaultSrc;
    }
    /**
     * @param {?} mimeType
     * @return {?}
     */
    static GetSrcPrefix(mimeType) {
        return `data:${mimeType};base64,`;
    }
    /**
     * @return {?}
     */
    get src() {
        return File.GetImageBinarySrc(this);
    }
}
function File_tsickle_Closure_declarations() {
    /** @type {?} */
    File.prototype.id;
    /** @type {?} */
    File.prototype.fileBinary;
    /** @type {?} */
    File.prototype.fileExtension;
    /** @type {?} */
    File.prototype.fileName;
    /** @type {?} */
    File.prototype.fileSize;
    /** @type {?} */
    File.prototype.mimeType;
    /** @type {?} */
    File.prototype.order;
}
