// @caiu/library/esm2015/lib/components/wallpaper/wallpaper.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, ChangeDetectionStrategy, Input, ElementRef } from '@angular/core';
import { toPx } from '../../shared/utils';
export class WallpaperComponent {
    /**
     * @param {?} elementRef
     */
    constructor(elementRef) {
        this.elementRef = elementRef;
        this.images = [];
        this.windowHeight = 0;
        this.windowWidth = 0;
    }
    /**
     * @return {?}
     */
    get canvasHeight() {
        return this.windowWidth * 2;
    }
    /**
     * @return {?}
     */
    get canvasWidth() {
        return this.windowWidth - 100;
    }
    /**
     * @return {?}
     */
    get elementHeight() {
        return this.elementRef && this.elementRef.nativeElement && this.elementRef.nativeElement.style ?
            this.elementRef.nativeElement.style.height : 0;
    }
    /**
     * @return {?}
     */
    get elementHeightPx() {
        return toPx(this.elementHeight);
    }
    /**
     * @return {?}
     */
    get elementWidth() {
        return this.elementRef && this.elementRef.nativeElement && this.elementRef.nativeElement.style ?
            this.elementRef.nativeElement.style.width : 0;
    }
    /**
     * @return {?}
     */
    get elementWidthPx() {
        return toPx(this.elementWidth);
    }
    /**
     * @return {?}
     */
    get maxColumns() {
        return 3;
    }
    /**
     * @return {?}
     */
    get maxRows() {
        return 3;
    }
    /**
     * @return {?}
     */
    get totalColumns() {
        return 12;
    }
    /**
     * @return {?}
     */
    get totalRows() {
        return 24;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
}
WallpaperComponent.decorators = [
    { type: Component, args: [{
                selector: 'iu-wallpaper',
                template: `<div class="wallpaper" [style.height.px]="windowHeight" [style.width.px]="windowWidth">
  <iu-collage [images]="images" [canvasHeight]="canvasHeight" [canvasWidth]="canvasWidth" [totalRows]="totalRows" [totalColumns]="totalColumns"
    [maxRows]="maxRows" [maxColumns]="maxColumns"></iu-collage>
</div>`,
                styles: [`.wallpaper{height:100%;width:100%}`],
                changeDetection: ChangeDetectionStrategy.OnPush
            },] },
];
/** @nocollapse */
WallpaperComponent.ctorParameters = () => [
    { type: ElementRef }
];
WallpaperComponent.propDecorators = {
    images: [{ type: Input }],
    windowHeight: [{ type: Input }],
    windowWidth: [{ type: Input }]
};
function WallpaperComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    WallpaperComponent.prototype.images;
    /** @type {?} */
    WallpaperComponent.prototype.windowHeight;
    /** @type {?} */
    WallpaperComponent.prototype.windowWidth;
    /** @type {?} */
    WallpaperComponent.prototype.elementRef;
}
