// @caiu/library/esm5/lib/components/grid/simple-grid/simple-grid.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { Grid } from '../grid.model';
import { GridColumn } from '../grid-column/grid-column.model';
import { Metadata } from '../../../shared/models';
import { toArray } from '../../../shared/utils';
var SimpleGridComponent = /** @class */ (function () {
    function SimpleGridComponent() {
        this.data = [];
        this.export = false;
        this.filterable = false;
        this.fileName = 'export';
        this.height = 600;
        this.pageSize = 30;
    }
    Object.defineProperty(SimpleGridComponent.prototype, "metadata", {
        get: /**
         * @return {?}
         */
        function () {
            return this._metadata ? this._metadata : (toArray(this.data).length > 0 ? toArray(this.data)[0].metadata : null);
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._metadata = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SimpleGridComponent.prototype, "columns", {
        get: /**
         * @return {?}
         */
        function () {
            var _this = this;
            return this.metadata ? this.columnKeys.map(function (key) { return new GridColumn(key, _this.metadata[key].columnLabel); }) : [];
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SimpleGridComponent.prototype, "columnKeys", {
        get: /**
         * @return {?}
         */
        function () {
            var _this = this;
            return this.metadata ? Object.keys(this.metadata).filter(function (key) { return _this.metadata[key].columnLabel ? true : false; }) : [];
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SimpleGridComponent.prototype, "columnLabels", {
        get: /**
         * @return {?}
         */
        function () {
            var _this = this;
            return this.metadata ? this.columnKeys.map(function (key) { return _this.metadata[key].columnLabel; }) : [];
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SimpleGridComponent.prototype, "gridData", {
        get: /**
         * @return {?}
         */
        function () {
            return Grid.Build(toArray(this.data));
        },
        enumerable: true,
        configurable: true
    });
    SimpleGridComponent.decorators = [
        { type: Component, args: [{
                    selector: 'iu-simple-grid',
                    template: "<iu-grid [model]=\"gridData\" [export]=\"export\" [filterable]=\"filterable\" [fileName]=\"fileName\" [pageSize]=\"pageSize\" [height]=\"height\">\n  <iu-grid-column *ngFor=\"let column of columns\" [model]=\"column\"></iu-grid-column>\n</iu-grid>\n",
                    styles: [""]
                },] },
    ];
    /** @nocollapse */
    SimpleGridComponent.ctorParameters = function () { return []; };
    SimpleGridComponent.propDecorators = {
        data: [{ type: Input }],
        export: [{ type: Input }],
        filterable: [{ type: Input }],
        fileName: [{ type: Input }],
        height: [{ type: Input }],
        pageSize: [{ type: Input }],
        metadata: [{ type: Input }]
    };
    return SimpleGridComponent;
}());
export { SimpleGridComponent };
function SimpleGridComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    SimpleGridComponent.prototype.data;
    /** @type {?} */
    SimpleGridComponent.prototype.export;
    /** @type {?} */
    SimpleGridComponent.prototype.filterable;
    /** @type {?} */
    SimpleGridComponent.prototype.fileName;
    /** @type {?} */
    SimpleGridComponent.prototype.height;
    /** @type {?} */
    SimpleGridComponent.prototype.pageSize;
    /** @type {?} */
    SimpleGridComponent.prototype._metadata;
}
