// @caiu/library/esm2015/lib/components/datepicker/daterange/daterange.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { ChangeDetectionStrategy, Component, forwardRef, Input, ViewChild, Output, EventEmitter, ChangeDetectorRef } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { DatepickerComponent } from '../datepicker.component';
import { DateRange } from '../../../shared/date';
import { build } from '../../../shared/utils';
export const /** @type {?} */ DATERANGE_ACCESSOR = {
    provide: NG_VALUE_ACCESSOR,
    useExisting: forwardRef(() => DaterangeComponent),
    multi: true
};
export class DaterangeComponent {
    /**
     * @param {?} changeDetectorRef
     */
    constructor(changeDetectorRef) {
        this.changeDetectorRef = changeDetectorRef;
        this.inline = false;
        this.startLabel = 'Start Date';
        this.endLabel = 'End Date';
        this.startDateChanged = new EventEmitter();
        this.endDateChanged = new EventEmitter();
        this._value = new DateRange();
        this.focused = new DateRange();
    }
    /**
     * @return {?}
     */
    get value() {
        return this._value;
    }
    /**
     * @param {?} val
     * @return {?}
     */
    set value(val) {
        this._value = val;
        if (this && this.changeDetectorRef && !this.changeDetectorRef['destroyed']) {
            this.changeDetectorRef.detectChanges();
        }
    }
    /**
     * @return {?}
     */
    get startDateId() {
        return this.startDatepicker.id;
    }
    /**
     * @return {?}
     */
    get startDateOpened() {
        return this.startDatepicker.opened;
    }
    /**
     * @return {?}
     */
    get endDateId() {
        return this.endDatepicker.id;
    }
    /**
     * @return {?}
     */
    get endDateOpened() {
        return this.endDatepicker.opened;
    }
    /**
     * @return {?}
     */
    get startDateValue() {
        return this.value.startDate;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set startDateValue(value) {
        this.changeStartDate(value);
    }
    /**
     * @return {?}
     */
    get endDateValue() {
        return this.value.endDate;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set endDateValue(value) {
        this.changeEndDate(value);
    }
    /**
     * @param {?} date
     * @return {?}
     */
    changeStartDate(date) {
        this.onChange(build(DateRange, {
            endDate: this.value.endDate,
            startDate: date
        }));
        this.startDateChanged.emit(date);
    }
    /**
     * @return {?}
     */
    closeStartDate() {
        this.startDatepicker.close();
    }
    /**
     * @return {?}
     */
    openStateDate() {
        this.startDatepicker.open();
    }
    /**
     * @param {?} date
     * @return {?}
     */
    changeEndDate(date) {
        this.onChange(build(DateRange, {
            endDate: date,
            startDate: this.value.startDate
        }));
        this.endDateChanged.emit(date);
    }
    /**
     * @return {?}
     */
    closeEndDate() {
        this.endDatepicker.close();
    }
    /**
     * @return {?}
     */
    openEndDate() {
        this.endDatepicker.open();
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnChange(fn) {
        this.onModelChange = fn;
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnTouched(fn) {
        this.onTouch = fn;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    writeValue(value) {
        this.value = value;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    onChange(value) {
        this.value = value;
        if (this.onModelChange) {
            this.onModelChange(value);
        }
    }
    /**
     * @param {?} value
     * @return {?}
     */
    onBlur(value) {
    }
    /**
     * @param {?} value
     * @return {?}
     */
    onFocus(value) {
        this.focused = value;
        if (this.onTouch) {
            this.onTouch();
        }
    }
}
DaterangeComponent.decorators = [
    { type: Component, args: [{
                selector: 'iu-daterange',
                template: `<div class="daterange">
  <div class="start-date" [class.date]="!inline" [class.inline-date]="inline">
    <label class="lbl-date iu-primary">{{ startLabel }}:</label>
    <iu-datepicker #startDate (selectedChanged)="changeStartDate($event)" [value]="startDateValue"></iu-datepicker>
  </div>
  <div class="end-date" [class.date]="!inline" [class.inline-date]="inline">
    <label class="lbl-date iu-primary">{{ endLabel }}:</label>
    <iu-datepicker #endDate (selectedChanged)="changeEndDate($event)" [value]="endDateValue"></iu-datepicker>
  </div>
</div>
`,
                styles: [`.daterange{display:inline-block;line-height:18px;padding:0 10px;border-radius:3px}.daterange .date{display:flex;padding:6px}.daterange .inline-date{display:inline-flex;padding:0}.daterange .inline-date.end-date{margin-left:32px}.daterange .lbl-date{width:76px;display:inline-block;font-size:14px;padding-top:22px}`],
                changeDetection: ChangeDetectionStrategy.OnPush,
                providers: [DATERANGE_ACCESSOR]
            },] },
];
/** @nocollapse */
DaterangeComponent.ctorParameters = () => [
    { type: ChangeDetectorRef }
];
DaterangeComponent.propDecorators = {
    inline: [{ type: Input }],
    min: [{ type: Input }],
    max: [{ type: Input }],
    startAt: [{ type: Input }],
    startLabel: [{ type: Input }],
    endLabel: [{ type: Input }],
    startView: [{ type: Input }],
    touchUi: [{ type: Input }],
    startDateChanged: [{ type: Output }],
    endDateChanged: [{ type: Output }],
    startDatepicker: [{ type: ViewChild, args: ['startDate',] }],
    endDatepicker: [{ type: ViewChild, args: ['endDate',] }],
    value: [{ type: Input }]
};
function DaterangeComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    DaterangeComponent.prototype.inline;
    /** @type {?} */
    DaterangeComponent.prototype.min;
    /** @type {?} */
    DaterangeComponent.prototype.max;
    /** @type {?} */
    DaterangeComponent.prototype.startAt;
    /** @type {?} */
    DaterangeComponent.prototype.startLabel;
    /** @type {?} */
    DaterangeComponent.prototype.endLabel;
    /** @type {?} */
    DaterangeComponent.prototype.startView;
    /** @type {?} */
    DaterangeComponent.prototype.touchUi;
    /** @type {?} */
    DaterangeComponent.prototype.startDateChanged;
    /** @type {?} */
    DaterangeComponent.prototype.endDateChanged;
    /** @type {?} */
    DaterangeComponent.prototype.startDatepicker;
    /** @type {?} */
    DaterangeComponent.prototype.endDatepicker;
    /** @type {?} */
    DaterangeComponent.prototype.onModelChange;
    /** @type {?} */
    DaterangeComponent.prototype.onTouch;
    /** @type {?} */
    DaterangeComponent.prototype._value;
    /** @type {?} */
    DaterangeComponent.prototype.focused;
    /** @type {?} */
    DaterangeComponent.prototype.changeDetectorRef;
}
