// @caiu/library/esm5/lib/events/events.models.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Collection } from '../shared/collection';
import { build } from '../shared/utils';
var Event = /** @class */ (function () {
    function Event() {
        this.emittedCount = 0;
        this.key = '';
        this.onAction = '';
        this.value = null;
    }
    return Event;
}());
export { Event };
function Event_tsickle_Closure_declarations() {
    /** @type {?} */
    Event.prototype.emittedCount;
    /** @type {?} */
    Event.prototype.handler;
    /** @type {?} */
    Event.prototype.key;
    /** @type {?} */
    Event.prototype.onAction;
    /** @type {?} */
    Event.prototype.value;
}
var Events = /** @class */ (function (_super) {
    tslib_1.__extends(Events, _super);
    function Events() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(Events.prototype, "actions", {
        get: /**
         * @return {?}
         */
        function () {
            return this.toArray().reduce(function (acc, event) {
                var /** @type {?} */ addValues = !event.onAction ?
                    [] : typeof event.onAction === 'string' ?
                    acc.indexOf(event.onAction) === -1 ? [event.onAction] : []
                    : event.onAction.filter(function (x) { return acc.indexOf(x) === -1; });
                return tslib_1.__spread(acc, addValues);
            }, []);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Events.prototype, "eventsByAction", {
        get: /**
         * @return {?}
         */
        function () {
            return this.toArray().reduce(function (acc, event) {
                if (typeof event.onAction === 'string') {
                    var /** @type {?} */ val = acc[event.onAction] || [];
                    acc[event.onAction] = tslib_1.__spread(val, [event]);
                }
                else {
                    event.onAction.forEach(function (actionType) {
                        var /** @type {?} */ val = acc[actionType] || [];
                        acc[actionType] = tslib_1.__spread(val, [event]);
                    });
                }
                return acc;
            }, {});
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} payload
     * @return {?}
     */
    Events.prototype.addEvent = /**
     * @param {?} payload
     * @return {?}
     */
    function (payload) {
        return build(Events, this.addItem(payload, payload.key));
    };
    /**
     * @param {?} actionType
     * @return {?}
     */
    Events.prototype.getActionEvents = /**
     * @param {?} actionType
     * @return {?}
     */
    function (actionType) {
        return this.eventsByAction[actionType];
    };
    /**
     * @param {?} payload
     * @return {?}
     */
    Events.prototype.removeEvent = /**
     * @param {?} payload
     * @return {?}
     */
    function (payload) {
        return build(Events, this.removeItem(payload));
    };
    /**
     * @param {?} payload
     * @return {?}
     */
    Events.prototype.replaceEvents = /**
     * @param {?} payload
     * @return {?}
     */
    function (payload) {
        var /** @type {?} */ events = this.copyItems();
        payload.forEach(function (e) {
            events[e.key] = e;
        });
        return build(Events, this, { items: events });
    };
    /**
     * @param {?} payload
     * @return {?}
     */
    Events.prototype.triggerEvent = /**
     * @param {?} payload
     * @return {?}
     */
    function (payload) {
        var /** @type {?} */ events = this.getActionEvents(payload.type)
            .map(function (e) { return build(Event, e, { value: e.handler ? e.handler(payload) : payload }); });
        return this.replaceEvents(events);
    };
    /**
     * @param {?} payload
     * @return {?}
     */
    Events.prototype.updateValue = /**
     * @param {?} payload
     * @return {?}
     */
    function (payload) {
        var /** @type {?} */ event = build(Event, this.items[payload.key], { value: payload.value });
        return build(Events, this.updateItem(event, payload.key, Event));
    };
    return Events;
}(Collection));
export { Events };
var UpdateValuePayload = /** @class */ (function () {
    function UpdateValuePayload() {
    }
    return UpdateValuePayload;
}());
export { UpdateValuePayload };
function UpdateValuePayload_tsickle_Closure_declarations() {
    /** @type {?} */
    UpdateValuePayload.prototype.key;
    /** @type {?} */
    UpdateValuePayload.prototype.value;
}
