// @caiu/library/esm5/lib/events/events.effects.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { Actions, Effect } from '@ngrx/effects';
import { Store } from '@ngrx/store';
import { Observable } from 'rxjs';
import { filter, map } from 'rxjs/operators';
import { EventActions } from './events.actions';
import { Events } from './events.models';
var EventEffects = /** @class */ (function () {
    function EventEffects(actions$, store) {
        var _this = this;
        this.actions$ = actions$;
        this.store = store;
        this.events = new Events();
        /**
         * Trigger event if action is a registered trigger.
         */
        this.onTriggerEvent = this.actions$.pipe(filter(function (action) { return _this.events.actions.indexOf(action.type) !== -1; }), map(this.trigger));
        this.store.select(function (s) { return s['events']; })
            .subscribe(function (events) {
            _this.events = events;
        });
    }
    /**
     * @param {?} action
     * @return {?}
     */
    EventEffects.prototype.trigger = /**
     * @param {?} action
     * @return {?}
     */
    function (action) {
        return EventActions.triggerEvent(action);
    };
    EventEffects.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    EventEffects.ctorParameters = function () { return [
        { type: Actions },
        { type: Store }
    ]; };
    tslib_1.__decorate([
        Effect(),
        tslib_1.__metadata("design:type", Observable)
    ], EventEffects.prototype, "onTriggerEvent", void 0);
    return EventEffects;
}());
export { EventEffects };
function EventEffects_tsickle_Closure_declarations() {
    /** @type {?} */
    EventEffects.prototype.events;
    /**
     * Trigger event if action is a registered trigger.
     * @type {?}
     */
    EventEffects.prototype.onTriggerEvent;
    /** @type {?} */
    EventEffects.prototype.actions$;
    /** @type {?} */
    EventEffects.prototype.store;
}
