// @caiu/library/esm5/lib/events/events.actions.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Event, UpdateValuePayload } from './events.models';
import { build, guid } from '../shared/utils';
var EventActions = /** @class */ (function () {
    function EventActions() {
    }
    /**
     * @param {?} onAction
     * @param {?=} handler
     * @param {?=} value
     * @return {?}
     */
    EventActions.addEvent = /**
     * @param {?} onAction
     * @param {?=} handler
     * @param {?=} value
     * @return {?}
     */
    function (onAction, handler, value) {
        if (value === void 0) { value = null; }
        var /** @type {?} */ e = build(Event, {
            onAction: onAction,
            handler: handler,
            value: value,
            key: guid()
        });
        return {
            type: EventActions.ADD_EVENT,
            payload: e,
            key: e.key
        };
    };
    /**
     * @param {?} key
     * @return {?}
     */
    EventActions.removeEvent = /**
     * @param {?} key
     * @return {?}
     */
    function (key) {
        return {
            type: EventActions.REMOVE_EVENT,
            payload: key
        };
    };
    /**
     * @param {?} action
     * @return {?}
     */
    EventActions.triggerEvent = /**
     * @param {?} action
     * @return {?}
     */
    function (action) {
        return {
            type: EventActions.TRIGGER_EVENT,
            payload: action
        };
    };
    /**
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    EventActions.updateValue = /**
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    function (key, value) {
        var /** @type {?} */ payload = build(UpdateValuePayload, {
            key: key,
            value: value
        });
        return {
            type: EventActions.UPDATE_VALUE,
            payload: payload
        };
    };
    EventActions.ADD_EVENT = '[Events] Add Event';
    EventActions.REMOVE_EVENT = '[Events] Remove Event';
    EventActions.TRIGGER_EVENT = '[Events] Trigger Event';
    EventActions.UPDATE_VALUE = '[Events] Update Value';
    return EventActions;
}());
export { EventActions };
function EventActions_tsickle_Closure_declarations() {
    /** @type {?} */
    EventActions.ADD_EVENT;
    /** @type {?} */
    EventActions.REMOVE_EVENT;
    /** @type {?} */
    EventActions.TRIGGER_EVENT;
    /** @type {?} */
    EventActions.UPDATE_VALUE;
}
