// @caiu/library/esm5/lib/events/events.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Store } from '@ngrx/store';
import { of } from 'rxjs';
import { combineLatest, skip, take } from 'rxjs/operators';
import { EventActions } from './events.actions';
import { eventValueSelector } from './events.selectors';
import * as i0 from "@angular/core";
import * as i1 from "@ngrx/store";
var EventsService = /** @class */ (function () {
    function EventsService(store) {
        this.store = store;
        this.eventIndex = 0;
        this.events = [];
    }
    Object.defineProperty(EventsService.prototype, "eventKey", {
        get: /**
         * @return {?}
         */
        function () {
            this.eventIndex++;
            return "event-" + this.eventIndex;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} actionType
     * @param {?=} handler
     * @return {?}
     */
    EventsService.prototype.addEvent = /**
     * @param {?} actionType
     * @param {?=} handler
     * @return {?}
     */
    function (actionType, handler) {
        var /** @type {?} */ f = handler ? handler : function (action) { return true; };
        var /** @type {?} */ actionWithKey = EventActions.addEvent(actionType, f, false);
        this.store.dispatch(actionWithKey);
        return actionWithKey.key;
    };
    /**
     * @param {?} action
     * @param {?=} handler
     * @return {?}
     */
    EventsService.prototype.dispatch = /**
     * @param {?} action
     * @param {?=} handler
     * @return {?}
     */
    function (action, handler) {
        return this.dispatchHttpAction(action, handler);
    };
    /**
     * @param {?} action
     * @param {?=} handler
     * @return {?}
     */
    EventsService.prototype.dispatchHttpAction = /**
     * @param {?} action
     * @param {?=} handler
     * @return {?}
     */
    function (action, handler) {
        var /** @type {?} */ successAction = action.payload.onSuccess;
        var /** @type {?} */ errorAction = typeof (action.payload.onError) === 'string' ? action.payload.onError : null;
        var /** @type {?} */ success = this.takeOne(successAction, handler);
        var /** @type {?} */ obs = errorAction ? success.pipe(combineLatest(this.takeOne(errorAction, handler), function (x, y) { return true; })) : success;
        this.store.dispatch(action);
        return obs;
    };
    /**
     * @param {?} eventKey
     * @return {?}
     */
    EventsService.prototype.removeEvent = /**
     * @param {?} eventKey
     * @return {?}
     */
    function (eventKey) {
        this.store.dispatch(EventActions.removeEvent(eventKey));
    };
    /**
     * @param {?} action
     * @param {?} handler
     * @return {?}
     */
    EventsService.prototype.subscribe = /**
     * @param {?} action
     * @param {?} handler
     * @return {?}
     */
    function (action, handler) {
        return this.take(1, action).subscribe(handler);
    };
    /**
     * @param {?=} n
     * @param {?=} actionType
     * @param {?=} handler
     * @return {?}
     */
    EventsService.prototype.take = /**
     * @param {?=} n
     * @param {?=} actionType
     * @param {?=} handler
     * @return {?}
     */
    function (n, actionType, handler) {
        if (n === void 0) { n = 1; }
        var /** @type {?} */ key = this.addEvent(actionType, handler);
        return eventValueSelector(this.store, key).pipe(skip(1), take(n));
    };
    /**
     * @param {?=} n
     * @param {?=} actions
     * @param {?=} handler
     * @return {?}
     */
    EventsService.prototype.takeFrom = /**
     * @param {?=} n
     * @param {?=} actions
     * @param {?=} handler
     * @return {?}
     */
    function (n, actions, handler) {
        var _this = this;
        if (n === void 0) { n = 1; }
        var /** @type {?} */ keys = actions.map(function (action) { return _this.addEvent(action, handler); });
        return keys.map(function (key) { return eventValueSelector(_this.store, key); })
            .reduce(function (acc, obs) {
            return acc.pipe(combineLatest(obs, function (x, y) { return true; }));
        }, of(false)).pipe(skip(1), take(n));
    };
    /**
     * @param {?} actionType
     * @param {?=} handler
     * @return {?}
     */
    EventsService.prototype.takeOne = /**
     * @param {?} actionType
     * @param {?=} handler
     * @return {?}
     */
    function (actionType, handler) {
        return this.take(1, actionType, handler);
    };
    EventsService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] },
    ];
    /** @nocollapse */
    EventsService.ctorParameters = function () { return [
        { type: Store }
    ]; };
    /** @nocollapse */ EventsService.ngInjectableDef = i0.defineInjectable({ factory: function EventsService_Factory() { return new EventsService(i0.inject(i1.Store)); }, token: EventsService, providedIn: "root" });
    return EventsService;
}());
export { EventsService };
function EventsService_tsickle_Closure_declarations() {
    /** @type {?} */
    EventsService.prototype.eventIndex;
    /** @type {?} */
    EventsService.prototype.events;
    /** @type {?} */
    EventsService.prototype.store;
}
