// @caiu/library/esm5/lib/components/dialog/dialog.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var DialogModel = /** @class */ (function () {
    function DialogModel() {
        this.title = ' ';
        this.actions = [];
        /**
         * Whether the dialog has a backdrop.
         */
        this.hasBackdrop = true;
        /**
         * Whether the user can use escape or clicking outside to close a modal.
         */
        this.disableClose = false;
    }
    Object.defineProperty(DialogModel.prototype, "config", {
        get: /**
         * @return {?}
         */
        function () {
            return {
                viewContainerRef: this.viewContainerRef,
                role: this.role,
                panelClass: this.panelClass,
                hasBackdrop: this.hasBackdrop,
                backdropClass: this.backdropClass,
                disableClose: this.disableClose,
                width: this.width,
                height: this.height,
                position: this.position,
                data: this.data,
                direction: this.direction
            };
        },
        enumerable: true,
        configurable: true
    });
    return DialogModel;
}());
export { DialogModel };
function DialogModel_tsickle_Closure_declarations() {
    /** @type {?} */
    DialogModel.prototype.title;
    /** @type {?} */
    DialogModel.prototype.actions;
    /**
     * Height of the dialog.
     * @type {?}
     */
    DialogModel.prototype.height;
    /**
     * Width of the dialog.
     * @type {?}
     */
    DialogModel.prototype.width;
    /**
     * Where the attached component should live in Angular's *logical* component tree.
     * This affects what is available for injection and the change detection order for the
     * component instantiated inside of the dialog. This does not affect where the dialog
     * content will be rendered.
     * @type {?}
     */
    DialogModel.prototype.viewContainerRef;
    /**
     * The ARIA role of the dialog element.
     * @type {?}
     */
    DialogModel.prototype.role;
    /**
     * Custom class for the overlay pane.
     * @type {?}
     */
    DialogModel.prototype.panelClass;
    /**
     * Whether the dialog has a backdrop.
     * @type {?}
     */
    DialogModel.prototype.hasBackdrop;
    /**
     * Custom class for the backdrop,
     * @type {?}
     */
    DialogModel.prototype.backdropClass;
    /**
     * Whether the user can use escape or clicking outside to close a modal.
     * @type {?}
     */
    DialogModel.prototype.disableClose;
    /**
     * Position overrides.
     * @type {?}
     */
    DialogModel.prototype.position;
    /**
     * Data being injected into the child component.
     * @type {?}
     */
    DialogModel.prototype.data;
    /**
     * Layout direction for the dialog's content.
     * @type {?}
     */
    DialogModel.prototype.direction;
}
var DialogAction = /** @class */ (function () {
    function DialogAction() {
        this.value = null;
        this.label = '';
        this.color = 'accent';
    }
    return DialogAction;
}());
export { DialogAction };
function DialogAction_tsickle_Closure_declarations() {
    /** @type {?} */
    DialogAction.prototype.value;
    /** @type {?} */
    DialogAction.prototype.label;
    /** @type {?} */
    DialogAction.prototype.color;
}
