// @caiu/library/esm5/lib/components/dialog/dialog.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, Input, ViewEncapsulation } from '@angular/core';
import { DialogModel } from './dialog.model';
var DialogComponent = /** @class */ (function () {
    function DialogComponent() {
        this.model = new DialogModel();
        this.showActions = true;
        this.showCloseAction = false;
        this.showCloseIcon = true;
        this.defaultActionColor = 'accent';
    }
    Object.defineProperty(DialogComponent.prototype, "actions", {
        get: /**
         * @return {?}
         */
        function () {
            return this.model.actions;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DialogComponent.prototype, "actionWidth", {
        get: /**
         * @return {?}
         */
        function () {
            var /** @type {?} */ actionsCount = Math.max(this.actions.length, 1);
            return 100 / actionsCount + "%";
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DialogComponent.prototype, "title", {
        get: /**
         * @return {?}
         */
        function () {
            return this.model.title;
        },
        enumerable: true,
        configurable: true
    });
    DialogComponent.decorators = [
        { type: Component, args: [{
                    selector: 'iu-dialog',
                    template: "<mat-toolbar color=\"primary\" mat-dialog-title>\n  <span class=\"dialog-title\">\n    {{ title }}\n    <button *ngIf=\"showCloseIcon\" mat-button [mat-dialog-close]=\"null\">\n      <mat-icon>close</mat-icon>\n    </button>\n  </span>\n</mat-toolbar>\n<mat-dialog-content class=\"scrollable\">\n  <ng-content></ng-content>\n</mat-dialog-content>\n<mat-dialog-actions *ngIf=\"showActions && actions.length > 0\">\n  <button *ngFor=\"let action of actions\" mat-raised-button [mat-dialog-close]=\"action.value\" [style.width]=\"actionWidth\" [style.color]=\"action.color\">{{ action.label }}</button>\n</mat-dialog-actions>\n<mat-dialog-actions *ngIf=\"showCloseAction && actions.length === 0\">\n  <button mat-raised-button mat-dialog-close=\"close\" [style.width]=\"actionWidth\" [style.color]=\"defaultActionColor\">Close</button>\n</mat-dialog-actions>\n",
                    styles: ["iu-dialog .mat-toolbar{margin:-24px -24px 0;width:calc(100% + 48px);min-height:32px}iu-dialog .mat-toolbar .dialog-title{position:relative;height:calc(100% - 8px);width:100%;margin-top:8px}iu-dialog .mat-toolbar .dialog-title button{cursor:pointer;position:absolute;right:0;top:-5px;border-radius:100%;height:42px;width:42px;min-width:36px}iu-dialog .mat-toolbar .dialog-title button:hover,iu-dialog .mat-toolbar .dialog-title button:hover mat-icon{font-weight:700}iu-dialog .mat-toolbar .dialog-title mat-icon{position:absolute;top:9px;right:8px;z-index:1}iu-dialog .mat-toolbar-row{height:48px}iu-dialog .mat-dialog-actions{margin-left:-24px;margin-right:-24px;color:#fff;padding:0}iu-dialog .mat-dialog-actions button{padding:12px 0}iu-dialog .mat-dialog-actions .mat-accent{color:#fff!important}"],
                    encapsulation: ViewEncapsulation.None
                },] },
    ];
    /** @nocollapse */
    DialogComponent.ctorParameters = function () { return []; };
    DialogComponent.propDecorators = {
        model: [{ type: Input }],
        showActions: [{ type: Input }],
        showCloseAction: [{ type: Input }],
        showCloseIcon: [{ type: Input }],
        defaultActionColor: [{ type: Input }]
    };
    return DialogComponent;
}());
export { DialogComponent };
function DialogComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    DialogComponent.prototype.model;
    /** @type {?} */
    DialogComponent.prototype.showActions;
    /** @type {?} */
    DialogComponent.prototype.showCloseAction;
    /** @type {?} */
    DialogComponent.prototype.showCloseIcon;
    /** @type {?} */
    DialogComponent.prototype.defaultActionColor;
}
