// @caiu/library/esm2015/lib/components/grid/grid-column/grid-column.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, ElementRef, Input, TemplateRef, ViewChild, ViewContainerRef } from '@angular/core';
import { GridColumn } from './grid-column.model';
export class GridColumnComponent {
    /**
     * @param {?} view
     */
    constructor(view) {
        this.view = view;
        this.editable = true;
        this.field = '';
        this.filter = 'text';
        this.filterable = true;
        this.hidden = false;
        this.locked = false;
        this.sortable = true;
    }
    /**
     * @return {?}
     */
    get colField() {
        return this.model.metadata.name;
    }
    /**
     * @return {?}
     */
    get colTitle() {
        return this.model.title;
    }
    /**
     * @return {?}
     */
    get colWidth() {
        return this.model.width;
    }
}
GridColumnComponent.decorators = [
    { type: Component, args: [{
                selector: 'iu-grid-column',
                template: `<kendo-grid-column [field]="colField" [title]="colTitle" [width]="colWidth" #kendoGridColumn>
  <span #content>
    <ng-content></ng-content>
  </span>
</kendo-grid-column>
`,
                styles: [``]
            },] },
];
/** @nocollapse */
GridColumnComponent.ctorParameters = () => [
    { type: ViewContainerRef }
];
GridColumnComponent.propDecorators = {
    model: [{ type: Input }],
    templateRef: [{ type: Input }],
    class: [{ type: Input }],
    editable: [{ type: Input }],
    editor: [{ type: Input }],
    field: [{ type: Input }],
    filter: [{ type: Input }],
    filterable: [{ type: Input }],
    footerClass: [{ type: Input }],
    footerStyle: [{ type: Input }],
    format: [{ type: Input }],
    headerClass: [{ type: Input }],
    headerStyle: [{ type: Input }],
    hidden: [{ type: Input }],
    locked: [{ type: Input }],
    media: [{ type: Input }],
    sortable: [{ type: Input }],
    style: [{ type: Input }],
    title: [{ type: Input }],
    width: [{ type: Input }],
    content: [{ type: ViewChild, args: ['content',] }]
};
function GridColumnComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    GridColumnComponent.prototype.model;
    /** @type {?} */
    GridColumnComponent.prototype.templateRef;
    /** @type {?} */
    GridColumnComponent.prototype.class;
    /** @type {?} */
    GridColumnComponent.prototype.editable;
    /** @type {?} */
    GridColumnComponent.prototype.editor;
    /** @type {?} */
    GridColumnComponent.prototype.field;
    /** @type {?} */
    GridColumnComponent.prototype.filter;
    /** @type {?} */
    GridColumnComponent.prototype.filterable;
    /** @type {?} */
    GridColumnComponent.prototype.footerClass;
    /** @type {?} */
    GridColumnComponent.prototype.footerStyle;
    /** @type {?} */
    GridColumnComponent.prototype.format;
    /** @type {?} */
    GridColumnComponent.prototype.headerClass;
    /** @type {?} */
    GridColumnComponent.prototype.headerStyle;
    /** @type {?} */
    GridColumnComponent.prototype.hidden;
    /** @type {?} */
    GridColumnComponent.prototype.locked;
    /** @type {?} */
    GridColumnComponent.prototype.media;
    /** @type {?} */
    GridColumnComponent.prototype.sortable;
    /** @type {?} */
    GridColumnComponent.prototype.style;
    /** @type {?} */
    GridColumnComponent.prototype.title;
    /** @type {?} */
    GridColumnComponent.prototype.width;
    /** @type {?} */
    GridColumnComponent.prototype.content;
    /** @type {?} */
    GridColumnComponent.prototype.view;
}
