// @caiu/library/esm2015/lib/components/grid/grid-column/grid-column.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { StyleModel } from '../../../shared/models';
/**
 * @template T
 */
export class GridColumn {
    /**
     * @param {?} name
     * @param {?} label
     */
    constructor(name, label) {
        this.name = name;
        this.label = label;
        this.styleModel = new StyleModel();
        this.editable = true;
        this.editor = 'text';
        this.filter = 'text';
        this.filterable = true;
        this.hidden = false;
        this.locked = false;
        this.sortable = true;
        this.metadata = { name, label };
        this.title = label;
    }
    /**
     * @return {?}
     */
    get class() {
        return this._class;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set class(value) {
        this._class = value;
    }
    /**
     * @return {?}
     */
    get field() {
        return this._field ? this._field : this.metadata.name;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set field(value) {
        this._field = value;
    }
    /**
     * @return {?}
     */
    get title() {
        return this._title ? this._title : this.metadata.label;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set title(value) {
        this._title = value;
    }
    /**
     * @return {?}
     */
    get width() {
        return this._width;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set width(value) {
        this._width = value;
    }
    /**
     * @param {?} col
     * @return {?}
     */
    update(col) {
        this.class = col.class;
        this.editable = col.editable;
        this.editor = col.editor;
        this.field = col.field;
        this.filter = col.filter;
        this.filterable = col.filterable;
        this.footerClass = col.footerClass;
        this.footerStyle = col.footerStyle;
        this.format = col.format;
        this.headerClass = col.headerClass;
        this.headerStyle = col.headerStyle;
        this.hidden = col.hidden;
        this.locked = col.locked;
        this.media = col.media;
        this.sortable = col.sortable;
        this.style = col.style;
        this.title = col.title;
        this.width = col.width;
    }
}
function GridColumn_tsickle_Closure_declarations() {
    /** @type {?} */
    GridColumn.prototype._class;
    /** @type {?} */
    GridColumn.prototype._field;
    /** @type {?} */
    GridColumn.prototype._title;
    /** @type {?} */
    GridColumn.prototype._width;
    /** @type {?} */
    GridColumn.prototype.styleModel;
    /** @type {?} */
    GridColumn.prototype.metadata;
    /** @type {?} */
    GridColumn.prototype.editable;
    /** @type {?} */
    GridColumn.prototype.editor;
    /** @type {?} */
    GridColumn.prototype.filter;
    /** @type {?} */
    GridColumn.prototype.filterable;
    /** @type {?} */
    GridColumn.prototype.footerClass;
    /** @type {?} */
    GridColumn.prototype.footerStyle;
    /** @type {?} */
    GridColumn.prototype.format;
    /** @type {?} */
    GridColumn.prototype.headerClass;
    /** @type {?} */
    GridColumn.prototype.headerStyle;
    /** @type {?} */
    GridColumn.prototype.hidden;
    /** @type {?} */
    GridColumn.prototype.locked;
    /** @type {?} */
    GridColumn.prototype.media;
    /** @type {?} */
    GridColumn.prototype.sortable;
    /** @type {?} */
    GridColumn.prototype.style;
    /** @type {?} */
    GridColumn.prototype.name;
    /** @type {?} */
    GridColumn.prototype.label;
}
