// @caiu/library/esm2015/lib/lookup/lookup.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Injectable, Inject } from '@angular/core';
import { Store } from '@ngrx/store';
import { map } from 'rxjs/operators';
import { LookupActions } from './lookup.actions';
import { AddLookupPayload } from './lookup.models';
import { LOOKUP_PATH, LOOKUP_KEYS } from './lookup.module';
import { HttpActions } from '../http/http.actions';
import { build } from '../shared/utils';
import * as i0 from "@angular/core";
import * as i1 from "@ngrx/store";
import * as i2 from "./lookup.module";
export class LookupService {
    /**
     * @param {?} store
     * @param {?=} lookupPath
     * @param {?=} lookupKeys
     */
    constructor(store, lookupPath = 'lookup', lookupKeys = []) {
        this.store = store;
        this._keys = [];
        this.path = 'lookup';
        this.path = lookupPath;
        this.keys = lookupKeys;
    }
    /**
     * @return {?}
     */
    get keys() {
        return this._keys;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set keys(value) {
        this._keys = value;
    }
    /**
     * @param {?} keys
     * @param {?=} values
     * @return {?}
     */
    load(keys, values) {
        if (values) {
            this.loadValues(values);
        }
        this.loadKeys(keys);
    }
    /**
     * @param {?} keys
     * @return {?}
     */
    loadKeys(keys) {
        this.keys = keys;
        this.store.dispatch(HttpActions.get(`${this.path}${this.buildQueryString()}`, LookupActions.ADD_KEYS));
    }
    /**
     * @param {?} values
     * @return {?}
     */
    loadValues(values) {
        this.store.dispatch(LookupActions.addValues(values));
    }
    /**
     * @param {?} key
     * @return {?}
     */
    lookup(key) {
        return this.store.select('lookup').pipe(map(s => s[key]));
    }
    /**
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    addKey(key, value) {
        this.store.dispatch(LookupActions.addKey(build(AddLookupPayload, {
            key: key,
            value: value
        })));
    }
    /**
     * @param {?} key
     * @return {?}
     */
    removeKey(key) {
        this.store.dispatch(LookupActions.removeKey(key));
    }
    /**
     * @return {?}
     */
    buildQueryString() {
        let /** @type {?} */ query = this.keys.length > 0 ? `?keys=${this.keys[0]}` : '';
        this.keys.filter((key, index) => index !== 0)
            .forEach(key => {
            query += `&keys=${key}`;
        });
        return query;
    }
}
LookupService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] },
];
/** @nocollapse */
LookupService.ctorParameters = () => [
    { type: Store },
    { type: undefined, decorators: [{ type: Inject, args: [LOOKUP_PATH,] }] },
    { type: Array, decorators: [{ type: Inject, args: [LOOKUP_KEYS,] }] }
];
/** @nocollapse */ LookupService.ngInjectableDef = i0.defineInjectable({ factory: function LookupService_Factory() { return new LookupService(i0.inject(i1.Store), i0.inject(i2.LOOKUP_PATH), i0.inject(i2.LOOKUP_KEYS)); }, token: LookupService, providedIn: "root" });
function LookupService_tsickle_Closure_declarations() {
    /** @type {?} */
    LookupService.prototype._keys;
    /** @type {?} */
    LookupService.prototype.path;
    /** @type {?} */
    LookupService.prototype.store;
}
