// @caiu/library/esm2015/lib/lookup/lookup.actions.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
export class LookupActions {
    /**
     * @param {?} payload
     * @return {?}
     */
    static addKey(payload) {
        return {
            type: LookupActions.ADD_KEY,
            payload: payload
        };
    }
    /**
     * @param {?} payload
     * @return {?}
     */
    static addValues(payload) {
        return {
            type: LookupActions.ADD_VALUES,
            payload: payload
        };
    }
    /**
     * @param {?} key
     * @return {?}
     */
    static removeKey(key) {
        return {
            type: LookupActions.REMOVE_KEY,
            payload: key
        };
    }
}
LookupActions.ADD_KEYS = '[Lookup] Add keys';
LookupActions.ADD_KEY = '[Lookup] Add key';
LookupActions.ADD_VALUES = '[Lookup] Add values';
LookupActions.REMOVE_KEY = '[Lookup] Remove key';
function LookupActions_tsickle_Closure_declarations() {
    /** @type {?} */
    LookupActions.ADD_KEYS;
    /** @type {?} */
    LookupActions.ADD_KEY;
    /** @type {?} */
    LookupActions.ADD_VALUES;
    /** @type {?} */
    LookupActions.REMOVE_KEY;
}
