// @caiu/library/esm2015/lib/lookup/lookup.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { NgModule, InjectionToken } from '@angular/core';
import { Store } from '@ngrx/store';
import { LookupService } from './lookup.service';
export const /** @type {?} */ LOOKUP_PATH = new InjectionToken('API lookup endpoint');
export const /** @type {?} */ LOOKUP_KEYS = new InjectionToken('API lookup keys');
/**
 * @param {?} store
 * @param {?=} LOOKUP_PATH
 * @param {?=} LOOKUP_KEYS
 * @return {?}
 */
export function lookupServiceFactory(store, LOOKUP_PATH = 'lookup', LOOKUP_KEYS = []) {
    return new LookupService(store, LOOKUP_PATH, LOOKUP_KEYS);
}
/**
 * @param {?=} path
 * @param {?=} keys
 * @return {?}
 */
export function provideLookup(path = '', keys = []) {
    return [
        { provide: LookupService, useFactory: lookupServiceFactory, deps: [Store, LOOKUP_PATH, LOOKUP_KEYS] },
        { provide: LOOKUP_PATH, useValue: path },
        { provide: LOOKUP_KEYS, useValue: keys },
    ];
}
export class LookupModule {
    /**
     * @return {?}
     */
    static forRoot() {
        return {
            ngModule: LookupModule,
            providers: provideLookup('lookup', [])
        };
    }
    /**
     * @param {?=} path
     * @return {?}
     */
    static forRootWithPath(path = '') {
        return {
            ngModule: LookupModule,
            providers: provideLookup(path, [])
        };
    }
    /**
     * @param {?} keys
     * @return {?}
     */
    static forRootWithKeys(keys) {
        return {
            ngModule: LookupModule,
            providers: provideLookup('lookup', keys)
        };
    }
    /**
     * @param {?=} path
     * @param {?=} keys
     * @return {?}
     */
    static forRootWithPathAndKeys(path = '', keys) {
        return {
            ngModule: LookupModule,
            providers: provideLookup(path, keys)
        };
    }
}
LookupModule.decorators = [
    { type: NgModule, args: [{
                imports: [],
                declarations: []
            },] },
];
