// @caiu/library/esm2015/lib/lookup/lookup.models.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { build } from '../shared/utils';
export class Lookup {
    constructor() {
        this.typeName = '';
        this.values = [];
        this._key = '';
    }
    /**
     * @return {?}
     */
    get key() {
        return this._key || this.typeName;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set key(value) {
        this._key = value;
    }
    /**
     * @return {?}
     */
    get name() {
        return this.key;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set name(value) {
        this.key = value;
    }
    /**
     * @return {?}
     */
    get names() {
        return this.values.map(x => x.name);
    }
}
function Lookup_tsickle_Closure_declarations() {
    /** @type {?} */
    Lookup.prototype.expires;
    /** @type {?} */
    Lookup.prototype.typeName;
    /** @type {?} */
    Lookup.prototype.values;
    /** @type {?} */
    Lookup.prototype._key;
}
export class LookupValue {
    constructor() {
        this.id = 0;
        this.active = true;
        this.description = '';
        this.name = '';
        this.sort = 0;
        this.tooltip = '';
        this._label = '';
    }
    /**
     * @return {?}
     */
    get value() {
        return this._value || this.id;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set value(value) {
        this._value = value;
    }
    /**
     * @return {?}
     */
    get label() {
        return this._label || this.description;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set label(value) {
        this._label = value;
    }
}
function LookupValue_tsickle_Closure_declarations() {
    /** @type {?} */
    LookupValue.prototype.id;
    /** @type {?} */
    LookupValue.prototype.active;
    /** @type {?} */
    LookupValue.prototype.description;
    /** @type {?} */
    LookupValue.prototype.name;
    /** @type {?} */
    LookupValue.prototype.sort;
    /** @type {?} */
    LookupValue.prototype.tooltip;
    /** @type {?} */
    LookupValue.prototype._label;
    /** @type {?} */
    LookupValue.prototype._value;
}
export class Lookups {
    /**
     * @param {?} state
     * @param {?} lookups
     * @return {?}
     */
    static AddLookups(state, lookups) {
        return lookups.reduce((acc, lookup) => build(Lookups, acc, { [lookup.key]: build(Lookup, lookup) }), state);
    }
    /**
     * @param {?} state
     * @param {?} lookup
     * @return {?}
     */
    static AddLookup(state, lookup) {
        return build(Lookups, state, {
            [lookup.key]: build(Lookup, lookup)
        });
    }
    /**
     * @param {?} values
     * @param {?} names
     * @return {?}
     */
    static BuildLookupValues(values, names) {
        return names.reduce((acc, name) => {
            const /** @type {?} */ value = values.find(x => x.name === name);
            return value ? [...acc, value] : [...acc];
        }, []);
    }
    /**
     * @param {?} values
     * @param {?} name
     * @return {?}
     */
    static FindLookupIdByName(values, name) {
        const /** @type {?} */ lkp = build(LookupValue, values.find(x => x.name === name));
        return lkp.id;
    }
    /**
     * @param {?} values
     * @param {?} id
     * @return {?}
     */
    static FindLookupNameById(values, id) {
        const /** @type {?} */ lkp = build(LookupValue, values.find(x => x.id === id));
        return lkp.name;
    }
    /**
     * @param {?} state
     * @param {?} removeKey
     * @return {?}
     */
    static RemoveLookup(state, removeKey) {
        const /** @type {?} */ newState = new Lookups();
        const /** @type {?} */ keys = Object.keys(state).filter(key => key !== removeKey);
        keys.forEach(key => {
            newState[key] = Object.assign({}, state[key]);
        });
        return newState;
    }
}
export class AddLookupPayload {
    constructor() {
        this.key = '';
    }
}
function AddLookupPayload_tsickle_Closure_declarations() {
    /** @type {?} */
    AddLookupPayload.prototype.key;
    /** @type {?} */
    AddLookupPayload.prototype.value;
}
