// @caiu/library/esm2015/lib/shared/base-entity.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Metadata } from './models';
import { build } from './utils';
export class BaseEntity {
    constructor() {
        this.createdBy = null;
        this.createdById = 0;
        this.createdDate = new Date();
        this.formId = 0;
        this.lastModifiedBy = null;
        this.lastModifiedById = 0;
        this.lastModifiedDate = new Date();
        this.matches = [];
        this.ignore = [
            '_metadata',
            'metadata',
            'ignore',
            'createdBy',
            'createdById',
            'createdDate',
            'formId',
            'lastModifiedBy',
            'lastModifiedById',
            'lastModifiedDate',
            'lastModifiedOn',
            'lastUpdated',
            'matches',
        ];
        this._metadata = new Metadata();
    }
    /**
     * @return {?}
     */
    get metadata() {
        return build(Metadata, this._metadata, {
            ignore: this.ignore
        });
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set metadata(value) {
        this._metadata = value;
    }
    /**
     * @return {?}
     */
    static getGetters() {
        return Object.keys(this.prototype).filter(name => {
            return typeof Object.getOwnPropertyDescriptor(this.prototype, name)['get'] === 'function';
        });
    }
    /**
     * @return {?}
     */
    static getSetters() {
        return Object.keys(this.prototype).filter(name => {
            return typeof Object.getOwnPropertyDescriptor(this.prototype, name)['set'] === 'function';
        });
    }
    /**
     * @return {?}
     */
    getGetters() {
        return Object.keys(this.constructor.prototype).filter(name => {
            return typeof Object.getOwnPropertyDescriptor(this.constructor.prototype, name)['get'] === 'function';
        });
    }
    /**
     * @return {?}
     */
    getSetters() {
        return Object.keys(this.constructor.prototype).filter(name => {
            return typeof Object.getOwnPropertyDescriptor(this.constructor.prototype, name)['set'] === 'function';
        });
    }
    /**
     * @param {?} value
     * @return {?}
     */
    update(value) {
        return Object.assign(new BaseEntity(), value);
    }
}
function BaseEntity_tsickle_Closure_declarations() {
    /** @type {?} */
    BaseEntity.prototype.createdBy;
    /** @type {?} */
    BaseEntity.prototype.createdById;
    /** @type {?} */
    BaseEntity.prototype.createdDate;
    /** @type {?} */
    BaseEntity.prototype.formId;
    /** @type {?} */
    BaseEntity.prototype.lastModifiedBy;
    /** @type {?} */
    BaseEntity.prototype.lastModifiedById;
    /** @type {?} */
    BaseEntity.prototype.lastModifiedDate;
    /** @type {?} */
    BaseEntity.prototype.matches;
    /** @type {?} */
    BaseEntity.prototype.ignore;
    /** @type {?} */
    BaseEntity.prototype._metadata;
}
