// @caiu/library/esm2015/lib/shared/collection.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { arrayUnion, build, inArray } from './utils';
/**
 * @template T
 */
export class Collection {
    /**
     * @param {?=} ctor
     */
    constructor(ctor) {
        this.activeId = 0;
        this.metadata = {};
        this._items = {};
        if (ctor) {
            this.ctor = ctor;
        }
    }
    /**
     * @template T
     * @param {?} state
     * @param {?} key
     * @param {?} item
     * @return {?}
     */
    static AddItem(state, key, item) {
        const /** @type {?} */ newState = {};
        Object.keys(state)
            .forEach(k => {
            newState[k] = state[k];
        });
        newState[key] = item;
        return newState;
    }
    /**
     * @template T
     * @param {?} items
     * @param {?=} findByKey
     * @return {?}
     */
    static BuildCollection(items, findByKey = 'id') {
        const /** @type {?} */ dictionary = Collection.BuildItems(items, findByKey);
        return Object.assign(new Collection(), { items: dictionary });
    }
    /**
     * @template T
     * @param {?} value
     * @param {?=} key
     * @return {?}
     */
    static BuildDictionaryFromArray(value, key = 'id') {
        const /** @type {?} */ obj = {};
        value.forEach(item => {
            obj[item[key]] = item;
        });
        return obj;
    }
    /**
     * @template T
     * @param {?} items
     * @param {?=} findByKey
     * @return {?}
     */
    static BuildFromArray(items, findByKey = 'id') {
        const /** @type {?} */ dictionary = Collection.BuildItems(items, findByKey);
        return Object.assign(new Collection(), { items: dictionary });
    }
    /**
     * @template T
     * @param {?} items
     * @param {?=} findByKey
     * @return {?}
     */
    static BuildItems(items, findByKey = 'id') {
        const /** @type {?} */ dictionary = {};
        items.forEach(item => {
            const /** @type {?} */ key = item[findByKey];
            dictionary[key] = item;
        });
        return dictionary;
    }
    /**
     * @template T
     * @param {?} item
     * @param {?} path
     * @return {?}
     */
    static Filter(item, path) {
        return Array.isArray(item['matches']) && inArray(item['matches'], path);
    }
    /**
     * @template T
     * @param {?} item
     * @param {?} paths
     * @return {?}
     */
    static FilterAnd(item, paths) {
        return Array.isArray(item['matches']) && paths.every(path => inArray(item['matches'], path));
    }
    /**
     * @template T
     * @param {?} item
     * @param {?} paths
     * @return {?}
     */
    static FilterOr(item, paths) {
        return Array.isArray(item['matches']) && paths.some(path => inArray(item['matches'], path));
    }
    /**
     * @template T
     * @param {?} state
     * @param {?} key
     * @return {?}
     */
    static RemoveItem(state, key) {
        const /** @type {?} */ newState = {};
        Object.keys(state)
            .filter(k => k !== key)
            .forEach(k => {
            newState[k] = state[k];
        });
        return newState;
    }
    /**
     * @return {?}
     */
    get active() {
        const /** @type {?} */ item = this.items[this.activeId] || /** @type {?} */ ({});
        return this.ctor ? build(this.ctor, item) : item;
    }
    /**
     * @return {?}
     */
    get ctor() {
        return this._ctor;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set ctor(value) {
        this._ctor = value;
    }
    /**
     * @return {?}
     */
    get count() {
        return this.toArray().length;
    }
    /**
     * @return {?}
     */
    get defaultItem() {
        return this.ctor ? new this.ctor() : /** @type {?} */ ({});
    }
    /**
     * @return {?}
     */
    get empty() {
        return this.toArray().length === 0;
    }
    /**
     * @return {?}
     */
    get instance() {
        const /** @type {?} */ activeId = this.activeId;
        const /** @type {?} */ ctor = this.ctor;
        const /** @type {?} */ lastUpdated = this.lastUpdated;
        const /** @type {?} */ metadata = this.metadata;
        // const collection = this.constructor();
        const /** @type {?} */ collection = new Collection();
        collection.activeId = activeId;
        collection.ctor = ctor;
        collection.lastUpdated = lastUpdated;
        collection.metadata = metadata;
        return collection;
    }
    /**
     * @return {?}
     */
    get items() {
        return this._items;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set items(value) {
        this._items = value;
        this.lastUpdated = new Date();
    }
    /**
     * @return {?}
     */
    get keys() {
        return Object.keys(this.items);
    }
    /**
     * @return {?}
     */
    get asArray() {
        return this.toArray();
    }
    /**
     * @return {?}
     */
    get itemsArray() {
        return this.asArray;
    }
    /**
     * @return {?}
     */
    toArray() {
        return Object.keys(this.items).map((key) => /** @type {?} */ (this.items[key]));
    }
    /**
     * @param {?} id
     * @param {?=} key
     * @return {?}
     */
    activate(id, key = 'id') {
        const /** @type {?} */ items = this.items[id] ? this.copyItems() : Object.assign(this.copyItems(), { [id]: { [key]: id } });
        return Object.assign(new Collection(), this, { activeId: id, items });
    }
    /**
     * @param {?} item
     * @param {?=} key
     * @return {?}
     */
    addItem(item, key) {
        const /** @type {?} */ items = this.copyItems();
        return Object.assign(this.constructor(), this, {
            items: Object.assign(items, { [key]: item })
        });
    }
    /**
     * @param {?} items
     * @param {?=} key
     * @return {?}
     */
    addItems(items, key = 'id') {
        const /** @type {?} */ newItems = {};
        this.toArray().forEach(item => {
            newItems[item[key]] = item;
        });
        items.forEach(item => {
            const /** @type {?} */ existingItem = newItems[item[key]] ? newItems[item[key]] : /** @type {?} */ ({});
            const /** @type {?} */ existingMatches = existingItem['matches'] || [];
            const /** @type {?} */ newMatches = item['matches'] || [];
            const /** @type {?} */ matches = arrayUnion(existingMatches, newMatches);
            const /** @type {?} */ newItem = this.buildItem(item, { matches: matches });
            newItems[item[key]] = this.buildItem(existingItem, newItem);
        });
        this.items = newItems;
        return Object.assign(new Collection(), this);
    }
    /**
     * @param {?} items
     * @param {?=} ctor
     * @return {?}
     */
    build(items, ctor) {
        return ctor ? build(ctor, { items })
            : /** @type {?} */ (Object.assign(this.instance, { items }));
    }
    /**
     * @param {?} existingItem
     * @param {?} newItem
     * @param {?=} ctor
     * @return {?}
     */
    buildItem(existingItem, newItem, ctor) {
        const /** @type {?} */ existing = existingItem || {};
        const /** @type {?} */ item = ctor ? build(ctor, existing, newItem)
            : this.ctor ? build(this.ctor, existing, newItem)
                : Object.assign({}, existing, newItem);
        return item;
    }
    /**
     * @param {?} items
     * @param {?=} ctor
     * @return {?}
     */
    buildItems(items, ctor) {
        const /** @type {?} */ factory = ctor ? ctor : this.ctor;
        return items.map(x => factory ? build(factory, x) : Object.assign(/** @type {?} */ ({}), x));
    }
    /**
     * @return {?}
     */
    copyItems() {
        return Object.keys(this.items)
            .reduce((acc, key) => {
            acc[key] = this.get(key);
            return acc;
        }, {});
    }
    /**
     * @param {?} key
     * @return {?}
     */
    delete(key) {
        return this.removeAt(key);
    }
    /**
     * @param {?} f
     * @return {?}
     */
    filter(f) {
        const /** @type {?} */ items = Collection.BuildDictionaryFromArray(this.toArray().filter(f));
        return Object.assign(new Collection(), this, { items });
    }
    /**
     * @param {?} f
     * @return {?}
     */
    filterBy(f) {
        return this.filterItems(f);
    }
    /**
     * @param {?} f
     * @return {?}
     */
    filterItems(f) {
        return this.toArray().filter(f);
    }
    /**
     * @param {?} f
     * @return {?}
     */
    findBy(f) {
        return this.asArray.find(f) || this.defaultItem;
    }
    /**
     * @param {?} id
     * @return {?}
     */
    get(id) {
        const /** @type {?} */ existing = this.items[id];
        return this.ctor ? build(this.ctor, existing)
            : Object.assign(/** @type {?} */ ({}), existing);
    }
    /**
     * @param {?} f
     * @return {?}
     */
    map(f) {
        const /** @type {?} */ items = Collection.BuildDictionaryFromArray(this.toArray().map(f));
        return Object.assign(new Collection(), this, { items });
    }
    /**
     * @param {?} path
     * @return {?}
     */
    query(path) {
        return this.toArray().filter(item => Collection.Filter(item, path));
    }
    /**
     * @param {?} paths
     * @return {?}
     */
    queryAnd(paths) {
        return this.toArray().filter(item => Collection.FilterAnd(item, paths));
    }
    /**
     * @param {?} paths
     * @return {?}
     */
    queryOr(paths) {
        return this.toArray().filter(item => Collection.FilterOr(item, paths));
    }
    /**
     * @param {?} key
     * @return {?}
     */
    removeAt(key) {
        const /** @type {?} */ keys = Object.keys(this.items);
        const /** @type {?} */ items = Collection.BuildDictionaryFromArray(keys.filter(x => x.toString() !== key.toString()).map(y => this.items[y]));
        return Object.assign(new Collection(), this, { items });
    }
    /**
     * @param {?} key
     * @return {?}
     */
    removeItem(key) {
        const /** @type {?} */ collection = this.constructor();
        const /** @type {?} */ keys = Object.keys(this.items).filter(k => k !== key);
        keys.forEach((k) => {
            collection.items[k] = /** @type {?} */ (Object.assign({}, this.items[k]));
        });
        return collection;
    }
    /**
     * @param {?} filter
     * @param {?=} key
     * @return {?}
     */
    removeItems(filter, key = 'id') {
        const /** @type {?} */ keys = this.toArray().filter(filter).map(x => x[key]);
        this.removeKeys(keys);
    }
    /**
     * @param {?} keys
     * @return {?}
     */
    removeKeys(keys) {
        keys.forEach((key) => {
            this.removeItem(key);
        });
    }
    /**
     * @param {?} itemsArray
     * @return {?}
     */
    replace(itemsArray) {
        const /** @type {?} */ items = Collection.BuildDictionaryFromArray(this.buildItems(itemsArray));
        return Object.assign(new Collection(), this, { items });
    }
    /**
     * @param {?} items
     * @param {?=} filter
     * @return {?}
     */
    replaceItems(items, filter) {
        if (filter) {
            const /** @type {?} */ f = (x) => !filter(x);
            return this.replaceItems([...this.filterItems(f), ...items]);
        }
        return Object.assign(new Collection(), this, { items });
    }
    /**
     * @param {?} value
     * @return {?}
     */
    setValue(value) {
        return Object.assign(new Collection(), this, value);
    }
    /**
     * @param {?} value
     * @param {?=} ctor
     * @param {?=} key
     * @return {?}
     */
    update(value, ctor, key = 'id') {
        if (Array.isArray(value)) {
            return this.updateItems(value, ctor, key);
        }
        return this.updateItem(value, value[key], ctor);
    }
    /**
     * @param {?} item
     * @param {?} key
     * @param {?=} ctor
     * @return {?}
     */
    updateItem(item, key, ctor) {
        const /** @type {?} */ items = this.copyItems();
        const /** @type {?} */ updatedItem = this.buildItem(items[key], item, ctor);
        return this.build(Object.assign(items, { [key]: updatedItem }));
    }
    /**
     * @param {?} items
     * @param {?=} ctor
     * @param {?=} key
     * @return {?}
     */
    updateItems(items, ctor, key = 'id') {
        const /** @type {?} */ existingItems = this.copyItems();
        const /** @type {?} */ updatedItems = items.reduce((acc, item) => {
            const /** @type {?} */ existingItem = acc[item[key]];
            acc[item[key]] = this.buildItem(existingItem, item, ctor);
            return acc;
        }, existingItems);
        return this.build(updatedItems);
    }
}
function Collection_tsickle_Closure_declarations() {
    /** @type {?} */
    Collection.prototype.activeId;
    /** @type {?} */
    Collection.prototype.lastUpdated;
    /** @type {?} */
    Collection.prototype.metadata;
    /** @type {?} */
    Collection.prototype._ctor;
    /** @type {?} */
    Collection.prototype._items;
}
