// @caiu/library/esm2015/lib/shared/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { getValue, truthy } from './utils';
export class DumbComponent {
    constructor() {
        this.requestState = 'DEFAULT';
        this.subscriptions = [];
    }
    /**
     * @return {?}
     */
    get inErrorState() {
        return this.requestState === 'ERROR';
    }
    /**
     * @return {?}
     */
    get inSuccessState() {
        return this.requestState === 'SUCCESS';
    }
    /**
     * @return {?}
     */
    get message() {
        return this.inErrorState ? `An error has occurred. Please try again later.`
            : this.inSuccessState ? `Saved successfully!` : '';
    }
    /**
     * @return {?}
     */
    get showMessage() {
        return this.inErrorState || this.inSuccessState;
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.removeSubscriptions();
    }
    /**
     * @param {?} subscription
     * @return {?}
     */
    addSubscription(subscription) {
        this.subscriptions.push(subscription);
    }
    /**
     * @param {?} result
     * @return {?}
     */
    closeDialog(result) {
        this.dialogRef.unsubscribe();
    }
    /**
     * @param {?} component
     * @param {?=} config
     * @return {?}
     */
    openDialog(component, config = {}) {
        const /** @type {?} */ dialogRef = this.dialog.open(component, config);
        this.dialogRef = dialogRef.afterClosed().subscribe(result => {
            this.closeDialog(result);
        });
    }
    /**
     * @return {?}
     */
    removeSubscriptions() {
        this.subscriptions.forEach(s => {
            s.unsubscribe();
        });
    }
    /**
     * @param {?} subscriptions
     * @return {?}
     */
    subscribe(subscriptions) {
        subscriptions.forEach(s => {
            this.addSubscription(s);
        });
    }
    /**
     * @param {?=} duration
     * @return {?}
     */
    flashErrorMessage(duration = 5000) {
        this.toErrorState();
        setTimeout(() => {
            this.resetRequestState();
        }, duration);
    }
    /**
     * @param {?=} duration
     * @return {?}
     */
    flashSuccessMessage(duration = 5000) {
        this.toSucessState();
        setTimeout(() => {
            this.resetRequestState();
        }, duration);
    }
    /**
     * @param {?} e
     * @return {?}
     */
    onError(e) {
        this.flashErrorMessage();
    }
    /**
     * @param {?} e
     * @return {?}
     */
    onSuccess(e) {
        this.flashSuccessMessage();
    }
    /**
     * @return {?}
     */
    resetRequestState() {
        this.requestState = 'DEFAULT';
    }
    /**
     * @param {?} value
     * @return {?}
     */
    setValue(value) {
        if (this.form && this.form.setValue) {
            this.form.setValue(getValue(value));
        }
    }
    /**
     * @return {?}
     */
    toErrorState() {
        this.requestState = 'ERROR';
    }
    /**
     * @return {?}
     */
    toSucessState() {
        this.requestState = 'SUCCESS';
    }
}
function DumbComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    DumbComponent.prototype.dialog;
    /** @type {?} */
    DumbComponent.prototype.dialogRef;
    /** @type {?} */
    DumbComponent.prototype.form;
    /** @type {?} */
    DumbComponent.prototype.requestState;
    /** @type {?} */
    DumbComponent.prototype.subscriptions;
}
export class FormComponent extends DumbComponent {
    /**
     * @return {?}
     */
    get editing() {
        return truthy(this.id);
    }
    /**
     * @return {?}
     */
    get id() {
        return this.model.id;
    }
    /**
     * @return {?}
     */
    get isValid() {
        return this.form.valid;
    }
    /**
     * @return {?}
     */
    markAsSubmitted() {
        this.form.markAsTouched();
        Object.keys(this.form.controls).forEach(key => {
            this.form.controls[key].markAsTouched();
        });
    }
}
function FormComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    FormComponent.prototype.form;
    /** @type {?} */
    FormComponent.prototype.model;
}
export class SmartComponent extends DumbComponent {
    /**
     * @param {?} store
     */
    constructor(store) {
        super();
        this.store = store;
    }
    /**
     * @param {?} action
     * @return {?}
     */
    dispatch(action) {
        this.store.dispatch(action);
    }
    /**
     * @param {?} action
     * @param {?=} onSuccess
     * @param {?=} onError
     * @return {?}
     */
    dispatchAndSubscribe(action, onSuccess, onError) {
        const /** @type {?} */ f1 = onSuccess ? onSuccess : e => {
            this.flashSuccessMessage();
        };
        const /** @type {?} */ f2 = onError ? onError : e => {
            this.flashErrorMessage();
        };
        if (this.events && this.events.dispatch) {
            this.addSubscription(this.events.dispatch(action).subscribe(f1, f2));
        }
        else {
            this.store.dispatch(action);
        }
    }
}
function SmartComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    SmartComponent.prototype.events;
    /** @type {?} */
    SmartComponent.prototype.store;
}
