// @caiu/library/esm2015/lib/shared/question.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Collection } from './collection';
import { DateHelper } from './date';
import { build } from './utils';
export class Question {
    constructor() {
        this.question = '';
        this.title = '';
    }
}
function Question_tsickle_Closure_declarations() {
    /** @type {?} */
    Question.prototype.answer;
    /** @type {?} */
    Question.prototype.lookup;
    /** @type {?} */
    Question.prototype.key;
    /** @type {?} */
    Question.prototype.question;
    /** @type {?} */
    Question.prototype.title;
    /** @type {?} */
    Question.prototype.type;
}
export class Questions extends Collection {
    /**
     * @param {?} questions
     * @param {?} answers
     * @return {?}
     */
    static AnswerQuestions(questions, answers) {
        return questions.map(question => Questions.AnswerQuestion(question, answers[question.key]));
    }
    /**
     * @param {?} question
     * @param {?} answerData
     * @return {?}
     */
    static AnswerQuestion(question, answerData) {
        let /** @type {?} */ answer = answerData;
        switch (question.type) {
            case 'date':
                answer = answerData;
                break;
            case 'monthyear':
                answer = DateHelper.ToMonthYearName(answerData);
                break;
            case 'yesno':
                answer = answerData ? 'Yes' : 'No';
                break;
        }
        return build(Question, question, { answer });
    }
    /**
     * @param {?} answers
     * @return {?}
     */
    answerQuestions(answers) {
        return Questions.AnswerQuestions(this.toArray(), answers);
    }
}
/**
 * @abstract
 */
export class QuestionsModel {
    constructor() {
        this.ignore = ['isAdd', 'questions'];
    }
    /**
     * @return {?}
     */
    get answers() {
        return Questions.AnswerQuestions(this.questions, this);
    }
}
function QuestionsModel_tsickle_Closure_declarations() {
    /** @type {?} */
    QuestionsModel.prototype.ignore;
    /** @type {?} */
    QuestionsModel.prototype.questions;
}
