// @caiu/library/esm2015/lib/shared/token.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
export class Token {
    constructor() {
        this.access_token = '';
        this.expiresIn = 0;
    }
    /**
     * @return {?}
     */
    get authenticated() {
        return !this.expired;
    }
    /**
     * @return {?}
     */
    get expires_in() {
        return this.expiresIn;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set expires_in(value) {
        this.expiresIn = value;
        this.expiresInDate = new Date();
    }
    /**
     * @return {?}
     */
    get expirationDate() {
        const /** @type {?} */ t = new Date(this.expiresInDate);
        t.setSeconds(t.getSeconds() + this.expiresIn);
        return t;
    }
    /**
     * @return {?}
     */
    get expired() {
        const /** @type {?} */ now = new Date();
        return this.access_token && this.expiresIn && this.expiresInDate ? now > this.expirationDate : true;
    }
    /**
     * @return {?}
     */
    toString() {
        return this.expired ? '' : this.access_token;
    }
}
function Token_tsickle_Closure_declarations() {
    /** @type {?} */
    Token.prototype.access_token;
    /** @type {?} */
    Token.prototype.expiresIn;
    /** @type {?} */
    Token.prototype.expiresInDate;
}
