// @caiu/library/esm2015/lib/shared/user.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Token } from './token';
import { build } from './utils';
export class User {
    constructor() {
        this.email = '';
        this.firstName = '';
        this.lastName = '';
        this.fullName = '';
        this.password = '';
        this.confirmPassword = '';
        this.userName = '';
    }
}
function User_tsickle_Closure_declarations() {
    /** @type {?} */
    User.prototype.email;
    /** @type {?} */
    User.prototype.firstName;
    /** @type {?} */
    User.prototype.lastName;
    /** @type {?} */
    User.prototype.fullName;
    /** @type {?} */
    User.prototype.password;
    /** @type {?} */
    User.prototype.confirmPassword;
    /** @type {?} */
    User.prototype.userName;
}
export class CurrentUser extends User {
    constructor() {
        super(...arguments);
        this.token = new Token();
    }
    /**
     * @param {?} data
     * @return {?}
     */
    static Build(data) {
        const /** @type {?} */ token = data && data.token ? build(Token, data.token) : new Token();
        return build(CurrentUser, data, { token });
    }
    /**
     * @return {?}
     */
    get authenticated() {
        return this.token && this.token.authenticated;
    }
    /**
     * @return {?}
     */
    get userRoles() {
        return new UserRoles(this);
    }
}
function CurrentUser_tsickle_Closure_declarations() {
    /** @type {?} */
    CurrentUser.prototype.token;
}
export class UserRoles {
    /**
     * @param {?=} user
     */
    constructor(user = new CurrentUser()) {
        this.user = user;
    }
}
function UserRoles_tsickle_Closure_declarations() {
    /** @type {?} */
    UserRoles.prototype.user;
}
export class Login {
    constructor() {
        this.grant_type = 'password';
        this.userName = '';
        this.password = '';
    }
}
function Login_tsickle_Closure_declarations() {
    /** @type {?} */
    Login.prototype.grant_type;
    /** @type {?} */
    Login.prototype.userName;
    /** @type {?} */
    Login.prototype.password;
}
export class ResetPassword {
    constructor() {
        this.passwordResetCode = '';
        this.password = '';
        this.confirmPassword = '';
    }
}
function ResetPassword_tsickle_Closure_declarations() {
    /** @type {?} */
    ResetPassword.prototype.passwordResetCode;
    /** @type {?} */
    ResetPassword.prototype.password;
    /** @type {?} */
    ResetPassword.prototype.confirmPassword;
}
