// @caiu/library/esm2015/lib/shared/window.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
export class Window {
    constructor() {
        this.windowHeight = 0;
        this.windowWidth = 0;
    }
    /**
     * @return {?}
     */
    get windowWidthSmall() {
        return this.windowWidth < 1200;
    }
    /**
     * @return {?}
     */
    get windowWidthMedium() {
        return !this.windowWidthSmall && this.windowWidth < 1300;
    }
    /**
     * @return {?}
     */
    get windowWidthLarge() {
        return this.windowWidth >= 1300;
    }
}
function Window_tsickle_Closure_declarations() {
    /** @type {?} */
    Window.prototype.windowHeight;
    /** @type {?} */
    Window.prototype.windowWidth;
}
export class WindowResize {
    constructor() {
        this.windowHeight = 0;
        this.windowWidth = 0;
    }
}
function WindowResize_tsickle_Closure_declarations() {
    /** @type {?} */
    WindowResize.prototype.windowHeight;
    /** @type {?} */
    WindowResize.prototype.windowWidth;
}
