// @caiu/library/esm5/lib/http/http.commands.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { HttpService } from './http.service';
import { HttpGetPayload, HttpPostPayload, HttpPutPayload, HttpDeletePayload } from './http.models';
import { Command } from '../shared/decorators';
import * as i0 from "@angular/core";
import * as i1 from "./http.service";
/**
 * @template T
 */
var HttpCommands = /** @class */ (function () {
    function HttpCommands(http) {
        this.http = http;
    }
    /**
     * @param {?} payload
     * @return {?}
     */
    HttpCommands.prototype.delete$ = /**
     * @param {?} payload
     * @return {?}
     */
    function (payload) {
        return this.http.delete(payload.path, payload.headers, payload.options);
    };
    /**
     * @param {?} payload
     * @return {?}
     */
    HttpCommands.prototype.get$ = /**
     * @param {?} payload
     * @return {?}
     */
    function (payload) {
        return this.http.get(payload.path, payload.headers, payload.options);
    };
    /**
     * @param {?} payload
     * @return {?}
     */
    HttpCommands.prototype.post$ = /**
     * @param {?} payload
     * @return {?}
     */
    function (payload) {
        return this.http.post(payload.path, payload.model, payload.headers, payload.options);
    };
    /**
     * @param {?} payload
     * @return {?}
     */
    HttpCommands.prototype.postFormUrlEncoded$ = /**
     * @param {?} payload
     * @return {?}
     */
    function (payload) {
        return this.http.postFormUrlEncoded(payload.path, payload.model, payload.headers, payload.options);
    };
    /**
     * @param {?} payload
     * @return {?}
     */
    HttpCommands.prototype.put$ = /**
     * @param {?} payload
     * @return {?}
     */
    function (payload) {
        return this.http.put(payload.path, payload.model, payload.headers, payload.options);
    };
    HttpCommands.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] },
    ];
    /** @nocollapse */
    HttpCommands.ctorParameters = function () { return [
        { type: HttpService }
    ]; };
    /** @nocollapse */ HttpCommands.ngInjectableDef = i0.defineInjectable({ factory: function HttpCommands_Factory() { return new HttpCommands(i0.inject(i1.HttpService)); }, token: HttpCommands, providedIn: "root" });
    tslib_1.__decorate([
        Command(),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", [HttpDeletePayload]),
        tslib_1.__metadata("design:returntype", Observable)
    ], HttpCommands.prototype, "delete$", null);
    tslib_1.__decorate([
        Command(),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", [HttpGetPayload]),
        tslib_1.__metadata("design:returntype", Observable)
    ], HttpCommands.prototype, "get$", null);
    tslib_1.__decorate([
        Command(),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", [HttpPostPayload]),
        tslib_1.__metadata("design:returntype", Observable)
    ], HttpCommands.prototype, "post$", null);
    tslib_1.__decorate([
        Command(),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", [HttpPostPayload]),
        tslib_1.__metadata("design:returntype", Observable)
    ], HttpCommands.prototype, "postFormUrlEncoded$", null);
    tslib_1.__decorate([
        Command(),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", [HttpPutPayload]),
        tslib_1.__metadata("design:returntype", Observable)
    ], HttpCommands.prototype, "put$", null);
    return HttpCommands;
}());
export { HttpCommands };
function HttpCommands_tsickle_Closure_declarations() {
    /** @type {?} */
    HttpCommands.prototype.http;
}
