// @caiu/library/esm2015/lib/components/datepicker/datepicker.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { ChangeDetectionStrategy, Component, EventEmitter, forwardRef, Input, Output, ViewChild, } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { MatDatepicker } from '@angular/material';
import { DateHelper } from '../../shared/date';
export const /** @type {?} */ DATEPICKER_ACCESSOR = {
    provide: NG_VALUE_ACCESSOR,
    useExisting: forwardRef(() => DatepickerComponent),
    multi: true
};
export class DatepickerComponent {
    constructor() {
        this.placeholder = 'Choose a date';
        this.startView = 'month';
        this.selectedChanged = new EventEmitter();
    }
    /**
     * @return {?}
     */
    get value() {
        return this._value;
    }
    /**
     * @param {?} val
     * @return {?}
     */
    set value(val) {
        this._value = val;
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnChange(fn) {
        this.onModelChange = fn;
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnTouched(fn) {
        this.onTouch = fn;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    writeValue(value) {
        this.value = value;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    onChange(value) {
        this.value = value;
        if (this.onModelChange) {
            this.onModelChange(value);
        }
    }
    /**
     * @param {?} input
     * @return {?}
     */
    onBlur(input) {
        const /** @type {?} */ date = new Date(input.value);
        if (DateHelper.IsValidDate(date)) {
            this.changeSelected(date);
        }
        else {
            input.value = DateHelper.FormatDate(this.value);
        }
    }
    /**
     * @return {?}
     */
    get id() {
        return this.datepicker.id;
    }
    /**
     * @return {?}
     */
    get opened() {
        return this.datepicker.opened;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.startAt = this.value;
    }
    /**
     * @param {?} date
     * @return {?}
     */
    changeSelected(date) {
        this.onChange(date);
        this.selectedChanged.emit(date);
    }
    /**
     * @return {?}
     */
    close() {
        this.datepicker.close();
    }
    /**
     * @return {?}
     */
    open() {
        this.datepicker.open();
    }
}
DatepickerComponent.decorators = [
    { type: Component, args: [{
                selector: 'iu-datepicker',
                template: `<mat-form-field>
  <input matInput [placeholder]="placeholder" [matDatepicker]="picker" (blur)="onBlur(pickerInput)" (change)="onChange(value)"
    [min]="min" [max]="max" [value]="value" #pickerInput>
  <mat-datepicker-toggle matSuffix [for]="picker"></mat-datepicker-toggle>
  <mat-datepicker #picker [startAt]="startAt" [startView]="startView" [touchUi]="touchUi" (selectedChanged)="changeSelected($event)"></mat-datepicker>
</mat-form-field>`,
                styles: [`.mat-input-container{float:left}`],
                changeDetection: ChangeDetectionStrategy.OnPush,
                providers: [DATEPICKER_ACCESSOR]
            },] },
];
/** @nocollapse */
DatepickerComponent.ctorParameters = () => [];
DatepickerComponent.propDecorators = {
    min: [{ type: Input }],
    max: [{ type: Input }],
    placeholder: [{ type: Input }],
    startAt: [{ type: Input }],
    startView: [{ type: Input }],
    touchUi: [{ type: Input }],
    selectedChanged: [{ type: Output }],
    datepicker: [{ type: ViewChild, args: ['picker',] }],
    value: [{ type: Input }]
};
function DatepickerComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    DatepickerComponent.prototype.min;
    /** @type {?} */
    DatepickerComponent.prototype.max;
    /** @type {?} */
    DatepickerComponent.prototype.placeholder;
    /** @type {?} */
    DatepickerComponent.prototype.startAt;
    /** @type {?} */
    DatepickerComponent.prototype.startView;
    /** @type {?} */
    DatepickerComponent.prototype.touchUi;
    /** @type {?} */
    DatepickerComponent.prototype.selectedChanged;
    /** @type {?} */
    DatepickerComponent.prototype.datepicker;
    /** @type {?} */
    DatepickerComponent.prototype.onModelChange;
    /** @type {?} */
    DatepickerComponent.prototype.onTouch;
    /** @type {?} */
    DatepickerComponent.prototype._value;
    /** @type {?} */
    DatepickerComponent.prototype.focused;
}
