// @caiu/library/esm5/lib/router/actions.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var RouterActions = /** @class */ (function () {
    function RouterActions() {
    }
    /**
     * @param {?} routeName
     * @param {?=} parent
     * @return {?}
     */
    RouterActions.activate = /**
     * @param {?} routeName
     * @param {?=} parent
     * @return {?}
     */
    function (routeName, parent) {
        var /** @type {?} */ payload = {
            routeName: routeName,
            parent: parent
        };
        return {
            type: RouterActions.ACTIVATE,
            payload: payload
        };
    };
    /**
     * @param {?} r
     * @return {?}
     */
    RouterActions.navigate = /**
     * @param {?} r
     * @return {?}
     */
    function (r) {
        return {
            type: RouterActions.NAVIGATE,
            payload: r
        };
    };
    /**
     * @return {?}
     */
    RouterActions.navigateSuccess = /**
     * @return {?}
     */
    function () {
        return {
            type: RouterActions.NAVIGATE_SUCCESS
        };
    };
    /**
     * @return {?}
     */
    RouterActions.navigateError = /**
     * @return {?}
     */
    function () {
        return {
            type: RouterActions.NAVIGATE_ERROR
        };
    };
    /**
     * @param {?} e
     * @return {?}
     */
    RouterActions.navCancel = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        return {
            type: RouterActions.NAVIGATION_CANCEL,
            payload: e
        };
    };
    /**
     * @param {?} e
     * @return {?}
     */
    RouterActions.navEnd = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        return {
            type: RouterActions.NAVIGATION_END,
            payload: e
        };
    };
    /**
     * @param {?} e
     * @return {?}
     */
    RouterActions.navError = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        return {
            type: RouterActions.NAVIGATION_ERROR,
            payload: e
        };
    };
    /**
     * @param {?} e
     * @return {?}
     */
    RouterActions.navStart = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        return {
            type: RouterActions.NAVIGATION_START,
            payload: e
        };
    };
    /**
     * @param {?} e
     * @return {?}
     */
    RouterActions.routesRecognized = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        return {
            type: RouterActions.ROUTES_RECOGNIZED,
            payload: e
        };
    };
    RouterActions.ACTIVATE = '[Router] Activate Route';
    RouterActions.NAVIGATE = '[Router] Navigate';
    RouterActions.NAVIGATE_SUCCESS = '[Router] Navigate Success';
    RouterActions.NAVIGATE_ERROR = '[Router] Navigate Error';
    RouterActions.NAVIGATION_CANCEL = '[Router] Cancel Navigation';
    RouterActions.NAVIGATION_END = '[Router] End Navigation';
    RouterActions.NAVIGATION_ERROR = '[Router] Error Navigation';
    RouterActions.NAVIGATION_START = '[Router] Start Navigation';
    RouterActions.ROUTES_RECOGNIZED = '[Router] Routes Recognized';
    return RouterActions;
}());
export { RouterActions };
function RouterActions_tsickle_Closure_declarations() {
    /** @type {?} */
    RouterActions.ACTIVATE;
    /** @type {?} */
    RouterActions.NAVIGATE;
    /** @type {?} */
    RouterActions.NAVIGATE_SUCCESS;
    /** @type {?} */
    RouterActions.NAVIGATE_ERROR;
    /** @type {?} */
    RouterActions.NAVIGATION_CANCEL;
    /** @type {?} */
    RouterActions.NAVIGATION_END;
    /** @type {?} */
    RouterActions.NAVIGATION_ERROR;
    /** @type {?} */
    RouterActions.NAVIGATION_START;
    /** @type {?} */
    RouterActions.ROUTES_RECOGNIZED;
}
