// @caiu/library/esm5/lib/router/router.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Injectable, EventEmitter } from '@angular/core';
import { Router, ActivatedRoute, NavigationCancel, NavigationEnd, NavigationError, NavigationStart, RoutesRecognized } from '@angular/router';
import { Store } from '@ngrx/store';
import { from } from 'rxjs';
import { RouterActions } from './actions';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
import * as i2 from "@ngrx/store";
var RouterService = /** @class */ (function () {
    function RouterService(_router, _route, store) {
        var _this = this;
        this._router = _router;
        this._route = _route;
        this.store = store;
        this.navigationCancel = new EventEmitter();
        this.navigationEnd = new EventEmitter();
        this.navigationError = new EventEmitter();
        this.navigationStart = new EventEmitter();
        this.routesRecognized = new EventEmitter();
        _router.events.subscribe(function (e) {
            if (e instanceof NavigationCancel) {
                _this.onNavigationCancel(e);
            }
            else if (e instanceof NavigationEnd) {
                _this.onNavigationEnd(e);
            }
            else if (e instanceof NavigationError) {
                _this.onNavigationError(e);
            }
            else if (e instanceof NavigationStart) {
                _this.onNavigationStart(e);
            }
            else if (e instanceof RoutesRecognized) {
                _this.onRoutesRecognized(e);
            }
        });
    }
    /**
     * @param {?} e
     * @return {?}
     */
    RouterService.prototype.onNavigationCancel = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        this.store.dispatch(RouterActions.navCancel(e));
    };
    /**
     * @param {?} e
     * @return {?}
     */
    RouterService.prototype.onNavigationEnd = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        this.store.dispatch(RouterActions.navEnd(e));
    };
    /**
     * @param {?} e
     * @return {?}
     */
    RouterService.prototype.onNavigationError = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        this.store.dispatch(RouterActions.navError(e));
    };
    /**
     * @param {?} e
     * @return {?}
     */
    RouterService.prototype.onNavigationStart = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        this.store.dispatch(RouterActions.navStart(e));
    };
    /**
     * @param {?} e
     * @return {?}
     */
    RouterService.prototype.onRoutesRecognized = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        this.store.dispatch(RouterActions.routesRecognized(e));
    };
    Object.defineProperty(RouterService.prototype, "activatedRoute", {
        get: /**
         * @return {?}
         */
        function () {
            return this._route;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RouterService.prototype, "routeData$", {
        get: /**
         * @return {?}
         */
        function () {
            return this.activatedRoute.data;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RouterService.prototype, "routeParams$", {
        get: /**
         * @return {?}
         */
        function () {
            return this.activatedRoute.params;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RouterService.prototype, "router", {
        get: /**
         * @return {?}
         */
        function () {
            return this._router;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RouterService.prototype, "url$", {
        get: /**
         * @return {?}
         */
        function () {
            return this.store.select(function (s) { return s.router.url; });
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} commands
     * @param {?=} extras
     * @return {?}
     */
    RouterService.prototype.navigate = /**
     * @param {?} commands
     * @param {?=} extras
     * @return {?}
     */
    function (commands, extras) {
        var /** @type {?} */ promise = this.router.navigate(commands, extras);
        return from(promise);
    };
    /**
     * @param {?} url
     * @param {?=} extras
     * @return {?}
     */
    RouterService.prototype.navigateByUrl = /**
     * @param {?} url
     * @param {?=} extras
     * @return {?}
     */
    function (url, extras) {
        var /** @type {?} */ promise = this.router.navigateByUrl(url, extras);
        return from(promise);
    };
    RouterService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] },
    ];
    /** @nocollapse */
    RouterService.ctorParameters = function () { return [
        { type: Router },
        { type: ActivatedRoute },
        { type: Store }
    ]; };
    /** @nocollapse */ RouterService.ngInjectableDef = i0.defineInjectable({ factory: function RouterService_Factory() { return new RouterService(i0.inject(i1.Router), i0.inject(i1.ActivatedRoute), i0.inject(i2.Store)); }, token: RouterService, providedIn: "root" });
    return RouterService;
}());
export { RouterService };
function RouterService_tsickle_Closure_declarations() {
    /** @type {?} */
    RouterService.prototype.navigationCancel;
    /** @type {?} */
    RouterService.prototype.navigationEnd;
    /** @type {?} */
    RouterService.prototype.navigationError;
    /** @type {?} */
    RouterService.prototype.navigationStart;
    /** @type {?} */
    RouterService.prototype.routesRecognized;
    /** @type {?} */
    RouterService.prototype._router;
    /** @type {?} */
    RouterService.prototype._route;
    /** @type {?} */
    RouterService.prototype.store;
}
