// @caiu/library/esm5/lib/router/selectors.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { of, combineLatest } from 'rxjs';
import { map, distinctUntilChanged } from 'rxjs/operators';
import { Breadcrumbs } from './models';
import { QueryModel } from '../shared/models';
import { toArray, toInt } from '../shared/utils';
/**
 * @param {?} store
 * @return {?}
 */
export function routeSelector(store) {
    return store.select('route');
}
/**
 * @param {?} store
 * @return {?}
 */
export function breadcrumbsSelector(store) {
    return routeSelector(store).pipe(map(function (r) {
        var /** @type {?} */ outlets = r.activatedOutlets;
        return new Breadcrumbs(Object.keys(outlets).map(function (key) { return outlets[key]; }));
    }));
}
/**
 * @param {?} store
 * @return {?}
 */
export function routeNameSelector(store) {
    return routeSelector(store).pipe(map(function (r) { return r.routeName; }), distinctUntilChanged());
}
/**
 * @param {?} store
 * @return {?}
 */
export function routeParamsSelector(store) {
    return routeSelector(store).pipe(map(function (r) { return r.params; }), distinctUntilChanged());
}
/**
 * @param {?} store
 * @param {?} key
 * @return {?}
 */
export function lazyRouteParamsSelector(store, key) {
    return routeParamsSelector(store).pipe(map(function (params) { return params[key]; }), distinctUntilChanged());
}
/**
 * @param {?} store
 * @param {?} key
 * @param {?=} defaultValue
 * @return {?}
 */
export function routeParamSelector(store, key, defaultValue) {
    if (defaultValue === void 0) { defaultValue = null; }
    return combineLatest(of(defaultValue), lazyRouteParamsSelector(store, key), function (x, y) {
        return y || x;
    });
}
/**
 * @param {?} store
 * @param {?} key
 * @return {?}
 */
export function routeParamArraySelector(store, key) {
    return routeParamSelector(store, key, []).pipe(map(function (x) { return toArray(x); }));
}
/**
 * @param {?} store
 * @param {?} key
 * @return {?}
 */
export function routeParamIdSelector(store, key) {
    return routeParamIntSelector(store, key);
}
/**
 * @param {?} store
 * @param {?} key
 * @return {?}
 */
export function routeParamIntSelector(store, key) {
    return routeParamSelector(store, key, 0).pipe(map(function (x) { return toInt(x); }));
}
/**
 * @param {?} store
 * @return {?}
 */
export function urlSelector(store) {
    return routeSelector(store).pipe(map(function (r) { return r.url; }), distinctUntilChanged());
}
/**
 * @param {?} store
 * @return {?}
 */
export function querySelector(store) {
    var /** @type {?} */ skip$ = routeParamIntSelector(store, 'skip');
    var /** @type {?} */ take$ = routeParamIntSelector(store, 'take');
    var /** @type {?} */ term$ = routeParamSelector(store, 'term');
    return combineLatest(skip$, take$, term$, function (skip, take, term) { return (Object.assign(new QueryModel(), {
        skip: skip,
        take: take,
        term: term
    })); });
}
