// @caiu/library/esm2015/lib/components/accordion/accordion.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { trigger, state, style, transition, animate } from '@angular/animations';
export class AccordionComponent {
    constructor() {
        this.active = true;
        this.opened = true;
        this.showArrow = false;
        this.closeStart = new EventEmitter();
        this.closeDone = new EventEmitter();
        this.openStart = new EventEmitter();
        this.openDone = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @return {?}
     */
    toggle() {
        this.opened = !this.opened;
    }
    /**
     * @return {?}
     */
    open() {
        this.opened = true;
    }
    /**
     * @return {?}
     */
    close() {
        this.opened = false;
    }
    /**
     * @param {?} e
     * @return {?}
     */
    onStart(e) {
        if (this.opened) {
            this.openStart.emit();
        }
        else {
            this.closeStart.emit();
        }
    }
    /**
     * @param {?} e
     * @return {?}
     */
    onDone(e) {
        if (this.opened) {
            this.openDone.emit();
        }
        else {
            this.closeDone.emit();
        }
    }
    /**
     * @return {?}
     */
    onClickTrigger() {
        if (this.active) {
            this.toggle();
        }
    }
}
AccordionComponent.decorators = [
    { type: Component, args: [{
                selector: 'iu-accordion',
                template: `<div class="accordion">
  <div class="trigger-wrapper" (click)="onClickTrigger()" [class.active]="active">
    <ng-content select=".accordion-trigger"></ng-content>
  </div>
  <div class="content-wrapper" [@toggle]="opened ? 'show' : 'hide'" (@toggle.start)="onStart($event)" (@toggle.done)="onDone($event)">
    <ng-content select=".accordion-content"></ng-content>
  </div>
</div>`,
                styles: [`.trigger-wrapper{position:relative;padding:0}.trigger-wrapper.active{cursor:pointer}.content-wrapper{overflow:hidden;clear:both}.mat-icon{position:absolute;right:2%;top:25%;color:#fff;font-size:32px;height:50%;font-weight:700}`],
                animations: [
                    trigger('toggle', [
                        state('*', style({ height: '0px' })),
                        state('show', style({ height: '*' })),
                        state('hide', style({ height: '0px' })),
                        transition('show <=> hide', [
                            animate('300ms ease-out')
                        ])
                    ])
                ]
            },] },
];
/** @nocollapse */
AccordionComponent.ctorParameters = () => [];
AccordionComponent.propDecorators = {
    active: [{ type: Input }],
    opened: [{ type: Input }],
    showArrow: [{ type: Input }],
    closeStart: [{ type: Output }],
    closeDone: [{ type: Output }],
    openStart: [{ type: Output }],
    openDone: [{ type: Output }]
};
function AccordionComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    AccordionComponent.prototype.active;
    /** @type {?} */
    AccordionComponent.prototype.opened;
    /** @type {?} */
    AccordionComponent.prototype.showArrow;
    /** @type {?} */
    AccordionComponent.prototype.closeStart;
    /** @type {?} */
    AccordionComponent.prototype.closeDone;
    /** @type {?} */
    AccordionComponent.prototype.openStart;
    /** @type {?} */
    AccordionComponent.prototype.openDone;
}
