// @caiu/library/esm5/lib/components/accordion/accordion.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { trigger, state, style, transition, animate } from '@angular/animations';
var AccordionComponent = /** @class */ (function () {
    function AccordionComponent() {
        this.active = true;
        this.opened = true;
        this.showArrow = false;
        this.closeStart = new EventEmitter();
        this.closeDone = new EventEmitter();
        this.openStart = new EventEmitter();
        this.openDone = new EventEmitter();
    }
    /**
     * @return {?}
     */
    AccordionComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    AccordionComponent.prototype.toggle = /**
     * @return {?}
     */
    function () {
        this.opened = !this.opened;
    };
    /**
     * @return {?}
     */
    AccordionComponent.prototype.open = /**
     * @return {?}
     */
    function () {
        this.opened = true;
    };
    /**
     * @return {?}
     */
    AccordionComponent.prototype.close = /**
     * @return {?}
     */
    function () {
        this.opened = false;
    };
    /**
     * @param {?} e
     * @return {?}
     */
    AccordionComponent.prototype.onStart = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        if (this.opened) {
            this.openStart.emit();
        }
        else {
            this.closeStart.emit();
        }
    };
    /**
     * @param {?} e
     * @return {?}
     */
    AccordionComponent.prototype.onDone = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        if (this.opened) {
            this.openDone.emit();
        }
        else {
            this.closeDone.emit();
        }
    };
    /**
     * @return {?}
     */
    AccordionComponent.prototype.onClickTrigger = /**
     * @return {?}
     */
    function () {
        if (this.active) {
            this.toggle();
        }
    };
    AccordionComponent.decorators = [
        { type: Component, args: [{
                    selector: 'iu-accordion',
                    template: "<div class=\"accordion\">\n  <div class=\"trigger-wrapper\" (click)=\"onClickTrigger()\" [class.active]=\"active\">\n    <ng-content select=\".accordion-trigger\"></ng-content>\n  </div>\n  <div class=\"content-wrapper\" [@toggle]=\"opened ? 'show' : 'hide'\" (@toggle.start)=\"onStart($event)\" (@toggle.done)=\"onDone($event)\">\n    <ng-content select=\".accordion-content\"></ng-content>\n  </div>\n</div>",
                    styles: [".trigger-wrapper{position:relative;padding:0}.trigger-wrapper.active{cursor:pointer}.content-wrapper{overflow:hidden;clear:both}.mat-icon{position:absolute;right:2%;top:25%;color:#fff;font-size:32px;height:50%;font-weight:700}"],
                    animations: [
                        trigger('toggle', [
                            state('*', style({ height: '0px' })),
                            state('show', style({ height: '*' })),
                            state('hide', style({ height: '0px' })),
                            transition('show <=> hide', [
                                animate('300ms ease-out')
                            ])
                        ])
                    ]
                },] },
    ];
    /** @nocollapse */
    AccordionComponent.ctorParameters = function () { return []; };
    AccordionComponent.propDecorators = {
        active: [{ type: Input }],
        opened: [{ type: Input }],
        showArrow: [{ type: Input }],
        closeStart: [{ type: Output }],
        closeDone: [{ type: Output }],
        openStart: [{ type: Output }],
        openDone: [{ type: Output }]
    };
    return AccordionComponent;
}());
export { AccordionComponent };
function AccordionComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    AccordionComponent.prototype.active;
    /** @type {?} */
    AccordionComponent.prototype.opened;
    /** @type {?} */
    AccordionComponent.prototype.showArrow;
    /** @type {?} */
    AccordionComponent.prototype.closeStart;
    /** @type {?} */
    AccordionComponent.prototype.closeDone;
    /** @type {?} */
    AccordionComponent.prototype.openStart;
    /** @type {?} */
    AccordionComponent.prototype.openDone;
}
