// @caiu/library/esm5/lib/lookup/lookup.models.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import * as tslib_1 from "tslib";
import { build } from '../shared/utils';
var Lookup = /** @class */ (function () {
    function Lookup() {
        this.typeName = '';
        this.values = [];
        this._key = '';
    }
    Object.defineProperty(Lookup.prototype, "key", {
        get: /**
         * @return {?}
         */
        function () {
            return this._key || this.typeName;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._key = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Lookup.prototype, "name", {
        get: /**
         * @return {?}
         */
        function () {
            return this.key;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.key = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Lookup.prototype, "names", {
        get: /**
         * @return {?}
         */
        function () {
            return this.values.map(function (x) { return x.name; });
        },
        enumerable: true,
        configurable: true
    });
    return Lookup;
}());
export { Lookup };
function Lookup_tsickle_Closure_declarations() {
    /** @type {?} */
    Lookup.prototype.expires;
    /** @type {?} */
    Lookup.prototype.typeName;
    /** @type {?} */
    Lookup.prototype.values;
    /** @type {?} */
    Lookup.prototype._key;
}
var LookupValue = /** @class */ (function () {
    function LookupValue() {
        this.id = 0;
        this.active = true;
        this.description = '';
        this.name = '';
        this.sort = 0;
        this.tooltip = '';
        this._label = '';
    }
    Object.defineProperty(LookupValue.prototype, "value", {
        get: /**
         * @return {?}
         */
        function () {
            return this._value || this.id;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._value = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LookupValue.prototype, "label", {
        get: /**
         * @return {?}
         */
        function () {
            return this._label || this.description;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._label = value;
        },
        enumerable: true,
        configurable: true
    });
    return LookupValue;
}());
export { LookupValue };
function LookupValue_tsickle_Closure_declarations() {
    /** @type {?} */
    LookupValue.prototype.id;
    /** @type {?} */
    LookupValue.prototype.active;
    /** @type {?} */
    LookupValue.prototype.description;
    /** @type {?} */
    LookupValue.prototype.name;
    /** @type {?} */
    LookupValue.prototype.sort;
    /** @type {?} */
    LookupValue.prototype.tooltip;
    /** @type {?} */
    LookupValue.prototype._label;
    /** @type {?} */
    LookupValue.prototype._value;
}
var Lookups = /** @class */ (function () {
    function Lookups() {
    }
    /**
     * @param {?} state
     * @param {?} lookups
     * @return {?}
     */
    Lookups.AddLookups = /**
     * @param {?} state
     * @param {?} lookups
     * @return {?}
     */
    function (state, lookups) {
        return lookups.reduce(function (acc, lookup) {
            return build(Lookups, acc, (_a = {}, _a[lookup.key] = build(Lookup, lookup), _a));
            var _a;
        }, state);
    };
    /**
     * @param {?} state
     * @param {?} lookup
     * @return {?}
     */
    Lookups.AddLookup = /**
     * @param {?} state
     * @param {?} lookup
     * @return {?}
     */
    function (state, lookup) {
        return build(Lookups, state, (_a = {},
            _a[lookup.key] = build(Lookup, lookup),
            _a));
        var _a;
    };
    /**
     * @param {?} values
     * @param {?} names
     * @return {?}
     */
    Lookups.BuildLookupValues = /**
     * @param {?} values
     * @param {?} names
     * @return {?}
     */
    function (values, names) {
        return names.reduce(function (acc, name) {
            var /** @type {?} */ value = values.find(function (x) { return x.name === name; });
            return value ? tslib_1.__spread(acc, [value]) : tslib_1.__spread(acc);
        }, []);
    };
    /**
     * @param {?} values
     * @param {?} name
     * @return {?}
     */
    Lookups.FindLookupIdByName = /**
     * @param {?} values
     * @param {?} name
     * @return {?}
     */
    function (values, name) {
        var /** @type {?} */ lkp = build(LookupValue, values.find(function (x) { return x.name === name; }));
        return lkp.id;
    };
    /**
     * @param {?} values
     * @param {?} id
     * @return {?}
     */
    Lookups.FindLookupNameById = /**
     * @param {?} values
     * @param {?} id
     * @return {?}
     */
    function (values, id) {
        var /** @type {?} */ lkp = build(LookupValue, values.find(function (x) { return x.id === id; }));
        return lkp.name;
    };
    /**
     * @param {?} state
     * @param {?} removeKey
     * @return {?}
     */
    Lookups.RemoveLookup = /**
     * @param {?} state
     * @param {?} removeKey
     * @return {?}
     */
    function (state, removeKey) {
        var /** @type {?} */ newState = new Lookups();
        var /** @type {?} */ keys = Object.keys(state).filter(function (key) { return key !== removeKey; });
        keys.forEach(function (key) {
            newState[key] = Object.assign({}, state[key]);
        });
        return newState;
    };
    return Lookups;
}());
export { Lookups };
var AddLookupPayload = /** @class */ (function () {
    function AddLookupPayload() {
        this.key = '';
    }
    return AddLookupPayload;
}());
export { AddLookupPayload };
function AddLookupPayload_tsickle_Closure_declarations() {
    /** @type {?} */
    AddLookupPayload.prototype.key;
    /** @type {?} */
    AddLookupPayload.prototype.value;
}
