// @caiu/library/esm5/lib/lookup/lookup.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Injectable, Inject } from '@angular/core';
import { Store } from '@ngrx/store';
import { map } from 'rxjs/operators';
import { LookupActions } from './lookup.actions';
import { AddLookupPayload } from './lookup.models';
import { LOOKUP_PATH, LOOKUP_KEYS } from './lookup.module';
import { HttpActions } from '../http/http.actions';
import { build } from '../shared/utils';
import * as i0 from "@angular/core";
import * as i1 from "@ngrx/store";
import * as i2 from "./lookup.module";
var LookupService = /** @class */ (function () {
    function LookupService(store, lookupPath, lookupKeys) {
        if (lookupPath === void 0) { lookupPath = 'lookup'; }
        if (lookupKeys === void 0) { lookupKeys = []; }
        this.store = store;
        this._keys = [];
        this.path = 'lookup';
        this.path = lookupPath;
        this.keys = lookupKeys;
    }
    Object.defineProperty(LookupService.prototype, "keys", {
        get: /**
         * @return {?}
         */
        function () {
            return this._keys;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._keys = value;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} keys
     * @param {?=} values
     * @return {?}
     */
    LookupService.prototype.load = /**
     * @param {?} keys
     * @param {?=} values
     * @return {?}
     */
    function (keys, values) {
        if (values) {
            this.loadValues(values);
        }
        this.loadKeys(keys);
    };
    /**
     * @param {?} keys
     * @return {?}
     */
    LookupService.prototype.loadKeys = /**
     * @param {?} keys
     * @return {?}
     */
    function (keys) {
        this.keys = keys;
        this.store.dispatch(HttpActions.get("" + this.path + this.buildQueryString(), LookupActions.ADD_KEYS));
    };
    /**
     * @param {?} values
     * @return {?}
     */
    LookupService.prototype.loadValues = /**
     * @param {?} values
     * @return {?}
     */
    function (values) {
        this.store.dispatch(LookupActions.addValues(values));
    };
    /**
     * @param {?} key
     * @return {?}
     */
    LookupService.prototype.lookup = /**
     * @param {?} key
     * @return {?}
     */
    function (key) {
        return this.store.select('lookup').pipe(map(function (s) { return s[key]; }));
    };
    /**
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    LookupService.prototype.addKey = /**
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    function (key, value) {
        this.store.dispatch(LookupActions.addKey(build(AddLookupPayload, {
            key: key,
            value: value
        })));
    };
    /**
     * @param {?} key
     * @return {?}
     */
    LookupService.prototype.removeKey = /**
     * @param {?} key
     * @return {?}
     */
    function (key) {
        this.store.dispatch(LookupActions.removeKey(key));
    };
    /**
     * @return {?}
     */
    LookupService.prototype.buildQueryString = /**
     * @return {?}
     */
    function () {
        var /** @type {?} */ query = this.keys.length > 0 ? "?keys=" + this.keys[0] : '';
        this.keys.filter(function (key, index) { return index !== 0; })
            .forEach(function (key) {
            query += "&keys=" + key;
        });
        return query;
    };
    LookupService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] },
    ];
    /** @nocollapse */
    LookupService.ctorParameters = function () { return [
        { type: Store },
        { type: undefined, decorators: [{ type: Inject, args: [LOOKUP_PATH,] }] },
        { type: Array, decorators: [{ type: Inject, args: [LOOKUP_KEYS,] }] }
    ]; };
    /** @nocollapse */ LookupService.ngInjectableDef = i0.defineInjectable({ factory: function LookupService_Factory() { return new LookupService(i0.inject(i1.Store), i0.inject(i2.LOOKUP_PATH), i0.inject(i2.LOOKUP_KEYS)); }, token: LookupService, providedIn: "root" });
    return LookupService;
}());
export { LookupService };
function LookupService_tsickle_Closure_declarations() {
    /** @type {?} */
    LookupService.prototype._keys;
    /** @type {?} */
    LookupService.prototype.path;
    /** @type {?} */
    LookupService.prototype.store;
}
