// @caiu/library/esm5/lib/lookup/lookup.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { NgModule, InjectionToken } from '@angular/core';
import { Store } from '@ngrx/store';
import { LookupService } from './lookup.service';
export var /** @type {?} */ LOOKUP_PATH = new InjectionToken('API lookup endpoint');
export var /** @type {?} */ LOOKUP_KEYS = new InjectionToken('API lookup keys');
/**
 * @param {?} store
 * @param {?=} LOOKUP_PATH
 * @param {?=} LOOKUP_KEYS
 * @return {?}
 */
export function lookupServiceFactory(store, LOOKUP_PATH, LOOKUP_KEYS) {
    if (LOOKUP_PATH === void 0) { LOOKUP_PATH = 'lookup'; }
    if (LOOKUP_KEYS === void 0) { LOOKUP_KEYS = []; }
    return new LookupService(store, LOOKUP_PATH, LOOKUP_KEYS);
}
/**
 * @param {?=} path
 * @param {?=} keys
 * @return {?}
 */
export function provideLookup(path, keys) {
    if (path === void 0) { path = ''; }
    if (keys === void 0) { keys = []; }
    return [
        { provide: LookupService, useFactory: lookupServiceFactory, deps: [Store, LOOKUP_PATH, LOOKUP_KEYS] },
        { provide: LOOKUP_PATH, useValue: path },
        { provide: LOOKUP_KEYS, useValue: keys },
    ];
}
var LookupModule = /** @class */ (function () {
    function LookupModule() {
    }
    /**
     * @return {?}
     */
    LookupModule.forRoot = /**
     * @return {?}
     */
    function () {
        return {
            ngModule: LookupModule,
            providers: provideLookup('lookup', [])
        };
    };
    /**
     * @param {?=} path
     * @return {?}
     */
    LookupModule.forRootWithPath = /**
     * @param {?=} path
     * @return {?}
     */
    function (path) {
        if (path === void 0) { path = ''; }
        return {
            ngModule: LookupModule,
            providers: provideLookup(path, [])
        };
    };
    /**
     * @param {?} keys
     * @return {?}
     */
    LookupModule.forRootWithKeys = /**
     * @param {?} keys
     * @return {?}
     */
    function (keys) {
        return {
            ngModule: LookupModule,
            providers: provideLookup('lookup', keys)
        };
    };
    /**
     * @param {?=} path
     * @param {?=} keys
     * @return {?}
     */
    LookupModule.forRootWithPathAndKeys = /**
     * @param {?=} path
     * @param {?=} keys
     * @return {?}
     */
    function (path, keys) {
        if (path === void 0) { path = ''; }
        return {
            ngModule: LookupModule,
            providers: provideLookup(path, keys)
        };
    };
    LookupModule.decorators = [
        { type: NgModule, args: [{
                    imports: [],
                    declarations: []
                },] },
    ];
    return LookupModule;
}());
export { LookupModule };
