// @caiu/library/esm2015/lib/components/editor/editor-window/editor-window.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, Inject, ViewEncapsulation } from '@angular/core';
import { MAT_DIALOG_DATA, MatDialogRef } from '@angular/material';
import { DialogAction, DialogModel } from '../../dialog/dialog.model';
import { build } from '../../../shared/utils';
export class EditorWindowComponent {
    /**
     * @param {?} data
     * @param {?} dialogRef
     */
    constructor(data, dialogRef) {
        this.data = data;
        this.dialogRef = dialogRef;
        this.actionWidth = '100%';
    }
    /**
     * @return {?}
     */
    get actions() {
        return [
            build(DialogAction, { value: this.data, label: 'Save and Close' }),
        ];
    }
    /**
     * @return {?}
     */
    get dialog() {
        return build(DialogModel, {
            title: this.title,
            actions: this.actions
        });
    }
    /**
     * @return {?}
     */
    get title() {
        return 'Editor';
    }
    /**
     * @return {?}
     */
    close() {
        this.dialogRef.close(this.data);
    }
    /**
     * @param {?} data
     * @return {?}
     */
    update(data) {
        this.data = data;
    }
}
EditorWindowComponent.decorators = [
    { type: Component, args: [{
                selector: 'iu-editor-window',
                template: `<span id="editor-window">
  <mat-toolbar color="primary" mat-dialog-title>
    <span class="dialog-title">
      {{ title }}
      <button mat-button [mat-dialog-close]="null">
        <mat-icon>close</mat-icon>
      </button>
    </span>
  </mat-toolbar>
  <mat-dialog-content>
    <div id="expanded-editor">
      <iu-editor [value]="data" [expanded]="true" [height]="380" (changes)="update($event)"></iu-editor>
    </div>
  </mat-dialog-content>
  <mat-dialog-actions>
    <button mat-raised-button (click)="close()" [style.width]="actionWidth" color="accent">Save and Close</button>
  </mat-dialog-actions>
</span>
`,
                styles: [`#editor-window #expanded-editor{margin:0;padding:0;height:60vh}#editor-window .mat-toolbar{margin:-24px -24px 0;width:calc(100% + 48px);min-height:32px!important}#editor-window .mat-toolbar .mat-toolbar-row{height:48px}#editor-window .mat-toolbar .dialog-title{position:relative;height:calc(100% - 8px);width:100%;margin-top:8px}#editor-window .mat-toolbar .dialog-title button{cursor:pointer;position:absolute;right:0;top:-5px;border-radius:100%;height:42px;width:42px;min-width:36px}#editor-window .mat-toolbar .dialog-title button:hover,#editor-window .mat-toolbar .dialog-title button:hover mat-icon{font-weight:700}#editor-window .mat-toolbar .dialog-title mat-icon{position:absolute;top:9px;right:8px;z-index:1}#editor-window .mat-toolbar-row{height:48px}#editor-window .mat-dialog-actions{margin-left:-24px;margin-right:-24px;color:#fff;padding:0}#editor-window .mat-dialog-actions button{padding:12px 0}#editor-window .mat-dialog-actions .mat-accent{color:#fff!important}`],
                encapsulation: ViewEncapsulation.None
            },] },
];
/** @nocollapse */
EditorWindowComponent.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [MAT_DIALOG_DATA,] }] },
    { type: MatDialogRef }
];
function EditorWindowComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    EditorWindowComponent.prototype.actionWidth;
    /** @type {?} */
    EditorWindowComponent.prototype.data;
    /** @type {?} */
    EditorWindowComponent.prototype.dialogRef;
}
