// @caiu/library/esm5/lib/components/grid/grid.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, ContentChildren, ElementRef, Input, QueryList, Renderer2, ViewChild, ViewEncapsulation, Output, EventEmitter } from '@angular/core';
import { Grid } from './grid.model';
import { GridColumnComponent } from './grid-column/grid-column.component';
import { str2int } from '../../shared/utils';
var GridComponent = /** @class */ (function () {
    function GridComponent(_elementRef, _renderer) {
        this._elementRef = _elementRef;
        this._renderer = _renderer;
        this.advancedFilter = false;
        this.autoCalculateHeight = true;
        this.buttonColor = '#fff';
        this.buttonCount = 5;
        this.buttonClass = 'toolbar-button';
        this.pagerInfo = true;
        this.pagerType = 'numeric';
        this.pageSizes = true;
        this.previousNext = true;
        this.export = false;
        this.fileName = 'export';
        this.filter = { logic: 'and', filters: [] };
        this.filterable = true;
        this.pageSize = 20;
        this.pager = {
            pageSizes: [10, 20, 50, 100]
        };
        this.pdfHeader = '';
        this.skip = 0;
        this.sortable = { mode: 'multiple' };
        this.sort = [];
        this.total = 0;
        this.changes = new EventEmitter();
        this.columns = [];
        this.filtering = true;
    }
    Object.defineProperty(GridComponent.prototype, "color", {
        /** The color of the toolbar. Can be primary, accent, or warn. */
        get: /**
         * The color of the toolbar. Can be primary, accent, or warn.
         * @return {?}
         */
        function () {
            return this._color;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._updateColor(value);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} newColor
     * @return {?}
     */
    GridComponent.prototype._updateColor = /**
     * @param {?} newColor
     * @return {?}
     */
    function (newColor) {
        this._setElementColor(this._color, false);
        this._setElementColor(newColor, true);
        this._color = newColor;
    };
    /**
     * @param {?} color
     * @param {?} isAdd
     * @return {?}
     */
    GridComponent.prototype._setElementColor = /**
     * @param {?} color
     * @param {?} isAdd
     * @return {?}
     */
    function (color, isAdd) {
        if (color != null && color !== '') {
            var /** @type {?} */ element = this._elementRef.nativeElement;
            if (isAdd) {
                this._renderer.addClass(element, "mat-" + color);
            }
            else {
                this._renderer.removeClass(element, "mat-" + color);
            }
        }
    };
    Object.defineProperty(GridComponent.prototype, "activeRows", {
        get: /**
         * @return {?}
         */
        function () {
            return this.pageRows(this.filterRows(this.sortRows(this.rows)));
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GridComponent.prototype, "computedRowHeight", {
        get: /**
         * @return {?}
         */
        function () {
            return this.rowHeightMax || 32;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GridComponent.prototype, "dynamicHeight", {
        get: /**
         * @return {?}
         */
        function () {
            return this.autoCalculateHeight ? (this.activeRows.length === 0 || !this.numericHeight) ?
                null : Math.max(96 + this.headerHeight + this.pagerHeight + this.toolbarHeight, Math.min(this.computedRowHeight * this.activeRows.length + this.headerHeight + this.pagerHeight + this.toolbarHeight, this.numericHeight))
                : this.numericHeight;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GridComponent.prototype, "excelFileName", {
        get: /**
         * @return {?}
         */
        function () {
            return this.fileName + ".xlsx";
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GridComponent.prototype, "filters", {
        get: /**
         * @return {?}
         */
        function () {
            return this.filter.filters;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GridComponent.prototype, "gridData", {
        get: /**
         * @return {?}
         */
        function () {
            return {
                data: this.activeRows,
                total: this.total ? this.total : this.rows.length
            };
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GridComponent.prototype, "headerHeight", {
        get: /**
         * @return {?}
         */
        function () {
            return this.filterable ? 66 : 27;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GridComponent.prototype, "numericHeight", {
        get: /**
         * @return {?}
         */
        function () {
            return typeof (this.height) === 'string' ? str2int(this.height) : this.height;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GridComponent.prototype, "pagerHeight", {
        get: /**
         * @return {?}
         */
        function () {
            return this.pageable ? 46 : 0;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GridComponent.prototype, "pagerSettings", {
        get: /**
         * @return {?}
         */
        function () {
            return this.pager;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GridComponent.prototype, "pdfFileName", {
        get: /**
         * @return {?}
         */
        function () {
            return this.fileName + ".pdf";
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GridComponent.prototype, "rows", {
        get: /**
         * @return {?}
         */
        function () {
            return this.model.data ? this.model.data.toArray() : [];
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GridComponent.prototype, "showFilters", {
        get: /**
         * @return {?}
         */
        function () {
            return this.filterable && this.filtering;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GridComponent.prototype, "take", {
        get: /**
         * @return {?}
         */
        function () {
            return this.pageSize || 20;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GridComponent.prototype, "toolbarHeight", {
        get: /**
         * @return {?}
         */
        function () {
            return this.export ? 53 : 0;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    GridComponent.prototype.ngOnChanges = /**
     * @return {?}
     */
    function () {
        this.model.update(this);
    };
    /**
     * @return {?}
     */
    GridComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.pageable = {
            buttonCount: this.buttonCount,
            info: this.pagerInfo,
            type: this.pagerType,
            pageSizes: this.pageSizes,
            previousNext: this.previousNext
        };
    };
    /**
     * @return {?}
     */
    GridComponent.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.contentChildren.forEach(function (x) {
            _this.columns.push(x);
        });
    };
    /**
     * @param {?} state
     * @return {?}
     */
    GridComponent.prototype.dataStateChange = /**
     * @param {?} state
     * @return {?}
     */
    function (state) {
        this.filterChange(state['filter']);
        this.pageChange({ skip: state['skip'], take: state['take'] });
        this.sortChange(state['sort']);
        this.changes.emit(state);
    };
    /**
     * @param {?} component
     * @return {?}
     */
    GridComponent.prototype.save = /**
     * @param {?} component
     * @return {?}
     */
    function (component) {
        var /** @type {?} */ options = component.workbookOptions();
        var /** @type {?} */ rows = options.sheets[0].rows;
        var /** @type {?} */ altIdx = 0;
        rows.forEach(function (row) {
            if (row.type === 'data') {
                if (altIdx % 2 !== 0) {
                    row.cells.forEach(function (cell) {
                        cell.background = '#aabbcc';
                    });
                }
                altIdx++;
            }
        });
        component.save(options);
    };
    /**
     * @param {?} e
     * @return {?}
     */
    GridComponent.prototype.filterChange = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        this.filter = e;
    };
    /**
     * @param {?} e
     * @return {?}
     */
    GridComponent.prototype.pageChange = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        this.skip = e.skip;
        this.pageSize = str2int(e.take.toString());
    };
    /**
     * @param {?} sort
     * @return {?}
     */
    GridComponent.prototype.sortChange = /**
     * @param {?} sort
     * @return {?}
     */
    function (sort) {
        this.sort = sort;
    };
    /**
     * @param {?} rows
     * @return {?}
     */
    GridComponent.prototype.filterRows = /**
     * @param {?} rows
     * @return {?}
     */
    function (rows) {
        return Grid.FilterRows(rows, this.filters);
    };
    /**
     * @param {?} rows
     * @return {?}
     */
    GridComponent.prototype.pageRows = /**
     * @param {?} rows
     * @return {?}
     */
    function (rows) {
        return this.pager ? Grid.PageRows(rows, this.skip, this.take) : rows;
    };
    /**
     * @param {?} rows
     * @return {?}
     */
    GridComponent.prototype.sortRows = /**
     * @param {?} rows
     * @return {?}
     */
    function (rows) {
        return Grid.SortRows(rows, this.sort);
    };
    GridComponent.decorators = [
        { type: Component, args: [{
                    selector: 'iu-grid',
                    template: "<mat-card class=\"grid mat-primary\" [class.mat-card]=\"!export\">\n  <kendo-grid [data]=\"gridData\" [filter]=\"filter\" [filterable]=\"showFilters\" [groupable]=\"groupable\" [height]=\"dynamicHeight\"\n    [pageSize]=\"pageSize\" [pageable]=\"pagerSettings\" [rowHeight]=\"rowHeight\" [scrollable]=\"scrollable\" [selectable]=\"selectable\"\n    [skip]=\"skip\" [sortable]=\"sortable\" [group]=\"group\" [sort]=\"sort\" (dataStateChange)=\"dataStateChange($event)\" #kendoGridInstance=\"kendoGrid\">\n\n    <!-- Toolbar Template -->\n    <ng-template kendoGridToolbarTemplate *ngIf=\"export\">\n      <button mat-raised-button class=\"toolbar-button pdf-button\" kendoGridPDFCommand>\n        <mat-icon>picture_as_pdf</mat-icon>Export to PDF</button>\n      <button mat-raised-button class=\"toolbar-button excel-button\" (click)=\"save(excelexport)\">\n        <mat-icon>import_export</mat-icon>Export To Excel</button>\n    </ng-template>\n\n    <!-- Grid Columns -->\n    <kendo-grid-column *ngFor=\"let col of columns\" [class]=\"col.class\" [editable]=\"col.editable\" [editor]=\"col.editor\" [field]=\"col.colField\"\n      [filter]=\"col.filter\" [filterable]=\"col.filterable\" [footerClass]=\"col.footerClass\" [footerStyle]=\"col.footerStyle\" [format]=\"col.format\"\n      [headerClass]=\"col.headerClass\" [headerStyle]=\"col.headerStyle\" [hidden]=\"col.hidden\" [locked]=\"col.locked\" [media]=\"col.media\"\n      [sortable]=\"col.sortable\" [style]=\"col.style\" [title]=\"col.colTitle\" [width]=\"col.width\">\n      <ng-template *ngIf=\"col.templateRef\" kendoGridCellTemplate let-dataItem>\n        <iu-grid-cell [templateRef]=\"col.templateRef\" [dataItem]=\"dataItem\"></iu-grid-cell>\n      </ng-template>\n    </kendo-grid-column>\n\n    <!-- Export to PDF -->\n    <kendo-grid-pdf *ngIf=\"export\" [fileName]=\"pdfFileName\" [allPages]=\"true\" paperSize=\"A4\" [repeatHeaders]=\"true\" [landscape]=\"true\">\n      <kendo-grid-pdf-margin top=\"2cm\" left=\"1cm\" right=\"1cm\" bottom=\"2cm\"></kendo-grid-pdf-margin>\n      <ng-template kendoGridPDFTemplate let-pageNum=\"pageNum\" let-totalPages=\"totalPages\">\n        <div class=\"page-template\">\n          <div class=\"header\">\n            <div style=\"float: right\">Page {{ pageNum }} of {{ totalPages }}</div>\n            {{ pdfHeader }}\n          </div>\n          <div class=\"footer\">\n            Page {{ pageNum }} of {{ totalPages }}\n          </div>\n        </div>\n      </ng-template>\n    </kendo-grid-pdf>\n\n    <!-- Export to Excel -->\n    <kendo-excelexport [data]=\"activeRows\" [fileName]=\"excelFileName\" [filterable]=\"true\" #excelexport>\n      <kendo-excelexport-column *ngFor=\"let col of columns\" [field]=\"col.colField\" [title]=\"col.colTitle\">\n      </kendo-excelexport-column>\n    </kendo-excelexport>\n\n  </kendo-grid>\n</mat-card>\n",
                    styles: [".mat-card.grid{padding:0}.grid .k-grid-header{padding-right:13px}.grid th.k-grid-header:first-child{padding-left:6px}.grid .mat-raised-button{color:#fff}.k-grid-header .k-header>.k-link{font-size:16px;padding:2px 12px}.k-grid td{border:none;padding:6px 12px}.k-widget{border:none}.k-filter-row>th{padding:4px;max-width:190px;border-width:0 0 1px 1px;border-style:solid}.k-input.k-textbox,.k-textbox{height:calc(1em + (4px * 2) + (1px * 2))!important;min-width:80px}.k-filtercell .k-filtercell-wrapper,.k-filtercell>span{padding-right:calc(1.5em + 8px + 2px + 4px)}.k-filtercell .k-dropdown-operator{display:none}.k-picker-wrap.k-state-default{border-bottom-width:1px}.page-template{font-family:\"DejaVu Sans\",Arial,sans-serif;position:absolute;width:100%;height:100%;top:0;left:0}.page-template .header{position:absolute;top:30px;left:30px;right:30px;border-bottom:1px solid #888;color:#888}.page-template .footer{position:absolute;bottom:30px;left:30px;right:30px;border-top:1px solid #888;text-align:center;color:#888}.toolbar-button{color:#fff;background-image:none;vertical-align:middle}"],
                    encapsulation: ViewEncapsulation.None
                },] },
    ];
    /** @nocollapse */
    GridComponent.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 }
    ]; };
    GridComponent.propDecorators = {
        model: [{ type: Input }],
        advancedFilter: [{ type: Input }],
        autoCalculateHeight: [{ type: Input }],
        buttonColor: [{ type: Input }],
        buttonCount: [{ type: Input }],
        buttonClass: [{ type: Input }],
        pagerInfo: [{ type: Input }],
        pagerType: [{ type: Input }],
        pageSizes: [{ type: Input }],
        previousNext: [{ type: Input }],
        detailRowHeight: [{ type: Input }],
        export: [{ type: Input }],
        fileName: [{ type: Input }],
        filter: [{ type: Input }],
        filterable: [{ type: Input }],
        groupable: [{ type: Input }],
        height: [{ type: Input }],
        pageSize: [{ type: Input }],
        pageable: [{ type: Input }],
        pager: [{ type: Input }],
        pdfHeader: [{ type: Input }],
        rowHeight: [{ type: Input }],
        rowHeightMax: [{ type: Input }],
        scrollable: [{ type: Input }],
        selectable: [{ type: Input }],
        skip: [{ type: Input }],
        sortable: [{ type: Input }],
        group: [{ type: Input }],
        rowClass: [{ type: Input }],
        sort: [{ type: Input }],
        total: [{ type: Input }],
        changes: [{ type: Output }],
        grid: [{ type: ViewChild, args: ['kendoGridInstance',] }],
        contentChildren: [{ type: ContentChildren, args: [GridColumnComponent,] }],
        color: [{ type: Input }]
    };
    return GridComponent;
}());
export { GridComponent };
function GridComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    GridComponent.prototype.model;
    /** @type {?} */
    GridComponent.prototype.advancedFilter;
    /** @type {?} */
    GridComponent.prototype.autoCalculateHeight;
    /** @type {?} */
    GridComponent.prototype.buttonColor;
    /** @type {?} */
    GridComponent.prototype.buttonCount;
    /** @type {?} */
    GridComponent.prototype.buttonClass;
    /** @type {?} */
    GridComponent.prototype.pagerInfo;
    /** @type {?} */
    GridComponent.prototype.pagerType;
    /** @type {?} */
    GridComponent.prototype.pageSizes;
    /** @type {?} */
    GridComponent.prototype.previousNext;
    /** @type {?} */
    GridComponent.prototype.detailRowHeight;
    /** @type {?} */
    GridComponent.prototype.export;
    /** @type {?} */
    GridComponent.prototype.fileName;
    /** @type {?} */
    GridComponent.prototype.filter;
    /** @type {?} */
    GridComponent.prototype.filterable;
    /** @type {?} */
    GridComponent.prototype.groupable;
    /** @type {?} */
    GridComponent.prototype.height;
    /** @type {?} */
    GridComponent.prototype.pageSize;
    /** @type {?} */
    GridComponent.prototype.pageable;
    /** @type {?} */
    GridComponent.prototype.pager;
    /** @type {?} */
    GridComponent.prototype.pdfHeader;
    /** @type {?} */
    GridComponent.prototype.rowHeight;
    /** @type {?} */
    GridComponent.prototype.rowHeightMax;
    /** @type {?} */
    GridComponent.prototype.scrollable;
    /** @type {?} */
    GridComponent.prototype.selectable;
    /** @type {?} */
    GridComponent.prototype.skip;
    /** @type {?} */
    GridComponent.prototype.sortable;
    /** @type {?} */
    GridComponent.prototype.group;
    /** @type {?} */
    GridComponent.prototype.rowClass;
    /** @type {?} */
    GridComponent.prototype.sort;
    /** @type {?} */
    GridComponent.prototype.total;
    /** @type {?} */
    GridComponent.prototype.changes;
    /** @type {?} */
    GridComponent.prototype.grid;
    /** @type {?} */
    GridComponent.prototype.contentChildren;
    /** @type {?} */
    GridComponent.prototype.columns;
    /** @type {?} */
    GridComponent.prototype._color;
    /** @type {?} */
    GridComponent.prototype.filtering;
    /** @type {?} */
    GridComponent.prototype._elementRef;
    /** @type {?} */
    GridComponent.prototype._renderer;
}
