// @caiu/library/esm5/lib/components/grid/grid.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { orderBy } from '@progress/kendo-data-query';
import { Collection } from '../../shared/collection';
import { DateHelper } from '../../shared/date';
import { QueryModel, StyleModel } from '../../shared/models';
/**
 * @template T
 */
var /**
 * @template T
 */
Grid = /** @class */ (function () {
    function Grid(data) {
        this.data = data;
        this.query = new QueryModel();
        this.style = new StyleModel();
        this.filterable = true;
        this.sortable = { mode: 'multiple' };
    }
    /**
     * @param {?} acc
     * @param {?} operator
     * @param {?} field
     * @param {?} value
     * @return {?}
     */
    Grid.ApplyFilter = /**
     * @param {?} acc
     * @param {?} operator
     * @param {?} field
     * @param {?} value
     * @return {?}
     */
    function (acc, operator, field, value) {
        switch (operator) {
            case 'contains':
                return acc.filter(function (row) { return Grid.Contains(row[field], value); });
            case 'gte':
                return acc.filter(function (row) { return Grid.MatchDate(row[field], value); });
        }
    };
    /**
     * @template T
     * @param {?} items
     * @return {?}
     */
    Grid.Build = /**
     * @template T
     * @param {?} items
     * @return {?}
     */
    function (items) {
        var /** @type {?} */ data = Object.assign(new Collection(), { items: items });
        return new Grid(data);
    };
    /**
     * @param {?} field
     * @param {?=} dir
     * @return {?}
     */
    Grid.BuildSort = /**
     * @param {?} field
     * @param {?=} dir
     * @return {?}
     */
    function (field, dir) {
        if (dir === void 0) { dir = 'asc'; }
        return /** @type {?} */ ({ dir: dir, field: field });
    };
    /**
     * @param {?} str
     * @param {?} substr
     * @return {?}
     */
    Grid.Contains = /**
     * @param {?} str
     * @param {?} substr
     * @return {?}
     */
    function (str, substr) {
        if (typeof (str) === 'string' && typeof (substr) === 'string') {
            return str.toLowerCase().includes(substr.toLowerCase());
        }
        return true;
    };
    /**
     * @param {?} rows
     * @param {?} filters
     * @return {?}
     */
    Grid.FilterRows = /**
     * @param {?} rows
     * @param {?} filters
     * @return {?}
     */
    function (rows, filters) {
        return filters.reduce(function (acc, filter) {
            var /** @type {?} */ operator = filter['operator'];
            var /** @type {?} */ field = filter['field'];
            var /** @type {?} */ value = filter['value'];
            return Grid.ApplyFilter(rows, operator, field, value);
        }, rows);
    };
    /**
     * @param {?} date1
     * @param {?} date2
     * @return {?}
     */
    Grid.MatchDate = /**
     * @param {?} date1
     * @param {?} date2
     * @return {?}
     */
    function (date1, date2) {
        return DateHelper.IsSameDay(date1, date2);
    };
    /**
     * @param {?} rows
     * @param {?} skip
     * @param {?} take
     * @return {?}
     */
    Grid.PageRows = /**
     * @param {?} rows
     * @param {?} skip
     * @param {?} take
     * @return {?}
     */
    function (rows, skip, take) {
        return rows.filter(function (row, index) { return index >= skip && index < skip + take; });
    };
    /**
     * @param {?} rows
     * @param {?} sort
     * @return {?}
     */
    Grid.SortRows = /**
     * @param {?} rows
     * @param {?} sort
     * @return {?}
     */
    function (rows, sort) {
        return orderBy(rows, sort);
    };
    /**
     * @param {?} grid
     * @return {?}
     */
    Grid.prototype.update = /**
     * @param {?} grid
     * @return {?}
     */
    function (grid) {
    };
    return Grid;
}());
/**
 * @template T
 */
export { Grid };
function Grid_tsickle_Closure_declarations() {
    /** @type {?} */
    Grid.prototype.query;
    /** @type {?} */
    Grid.prototype.style;
    /** @type {?} */
    Grid.prototype.detailRowHeight;
    /** @type {?} */
    Grid.prototype.filter;
    /** @type {?} */
    Grid.prototype.filterable;
    /** @type {?} */
    Grid.prototype.groupable;
    /** @type {?} */
    Grid.prototype.height;
    /** @type {?} */
    Grid.prototype.pageSize;
    /** @type {?} */
    Grid.prototype.pageable;
    /** @type {?} */
    Grid.prototype.rowHeight;
    /** @type {?} */
    Grid.prototype.scrollable;
    /** @type {?} */
    Grid.prototype.selectable;
    /** @type {?} */
    Grid.prototype.skip;
    /** @type {?} */
    Grid.prototype.sortable;
    /** @type {?} */
    Grid.prototype.group;
    /** @type {?} */
    Grid.prototype.rowClass;
    /** @type {?} */
    Grid.prototype.sort;
    /** @type {?} */
    Grid.prototype.data;
}
