// @caiu/library/esm2015/lib/forms/actions.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
export class FormActions {
    /**
     * @param {?} control
     * @return {?}
     */
    static addForm(control) {
        return {
            type: FormActions.ADD,
            payload: control
        };
    }
    /**
     * @param {?} control
     * @param {?} changes
     * @return {?}
     */
    static changeStatus(control, changes) {
        return {
            type: FormActions.STATUS_CHANGES,
            payload: {
                control: control,
                changes: changes
            }
        };
    }
    /**
     * @param {?} control
     * @param {?} changes
     * @return {?}
     */
    static changeValue(control, changes) {
        return {
            type: FormActions.VALUE_CHANGES,
            payload: {
                control: control,
                changes: changes
            }
        };
    }
    /**
     * @param {?} route
     * @return {?}
     */
    static navigateToForm(route) {
        return {
            type: FormActions.NAVIGATE,
            payload: route
        };
    }
    /**
     * @param {?} forms
     * @return {?}
     */
    static resetForms(forms) {
        return {
            type: FormActions.RESET,
            payload: forms
        };
    }
}
FormActions.ADD = '[forms] Add Form';
FormActions.RESET = '[forms] Reset Forms';
FormActions.NAVIGATE = '[forms] Navigate to Form';
FormActions.STATUS_CHANGES = '[form] Status Changes';
FormActions.VALUE_CHANGES = '[form] Value Changes';
function FormActions_tsickle_Closure_declarations() {
    /** @type {?} */
    FormActions.ADD;
    /** @type {?} */
    FormActions.RESET;
    /** @type {?} */
    FormActions.NAVIGATE;
    /** @type {?} */
    FormActions.STATUS_CHANGES;
    /** @type {?} */
    FormActions.VALUE_CHANGES;
}
