// @caiu/library/esm5/lib/components/file-upload/file-preview/file-preview.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, Input, EventEmitter, Output } from '@angular/core';
import { FileUpload } from '../file-upload.model';
var FilePreviewComponent = /** @class */ (function () {
    function FilePreviewComponent() {
        this.file = new FileUpload();
        this.ordered = true;
        this.first = false;
        this.last = false;
        this.moveUp = new EventEmitter();
        this.moveDown = new EventEmitter();
        this.remove = new EventEmitter();
    }
    Object.defineProperty(FilePreviewComponent.prototype, "cursorStyle", {
        get: /**
         * @return {?}
         */
        function () {
            return this.ordered ? 'move' : 'default';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FilePreviewComponent.prototype, "graphic", {
        get: /**
         * @return {?}
         */
        function () {
            if (this.file.loading) {
                return 'cloud_upload';
            }
            else if (this.file.isImage) {
                return 'image';
            }
            else {
                switch (this.file.extension) {
                    case 'pdf':
                        return 'picture_as_pdf';
                    default:
                        return 'cloud_done';
                }
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FilePreviewComponent.prototype, "showImage", {
        get: /**
         * @return {?}
         */
        function () {
            return this.graphic === 'image';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FilePreviewComponent.prototype, "showUpArrow", {
        get: /**
         * @return {?}
         */
        function () {
            return this.ordered && !this.first;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FilePreviewComponent.prototype, "showDownArrow", {
        get: /**
         * @return {?}
         */
        function () {
            return this.ordered && !this.last;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    FilePreviewComponent.prototype.onMoveUp = /**
     * @return {?}
     */
    function () {
        this.moveUp.emit(this.file);
    };
    /**
     * @return {?}
     */
    FilePreviewComponent.prototype.onMoveDown = /**
     * @return {?}
     */
    function () {
        this.moveDown.emit(this.file);
    };
    /**
     * @return {?}
     */
    FilePreviewComponent.prototype.onRemove = /**
     * @return {?}
     */
    function () {
        this.remove.emit(this.file);
    };
    FilePreviewComponent.decorators = [
        { type: Component, args: [{
                    selector: 'iu-file-preview',
                    template: "<mat-card class=\"file-preview\" [style.cursor]=\"cursorStyle\">\n  <div class=\"graphic\">\n    <img *ngIf=\"showImage\" [src]=\"file.src\" [style.height]=\"height\" [style.width]=\"width\">\n    <mat-icon *ngIf=\"!showImage\" class=\"icon-graphic\">{{ graphic }}</mat-icon>\n  </div>\n  <mat-progress-bar mode=\"indeterminate\" *ngIf=\"file.loading\"></mat-progress-bar>\n  <div class=\"file-info\">\n    <span class=\"file-name\">{{ file.name }}</span>\n  </div>\n  <mat-icon *ngIf=\"showUpArrow\" class=\"up-arrow\" (click)=\"onMoveUp()\">keyboard_arrow_up</mat-icon>\n  <mat-icon *ngIf=\"showDownArrow\" class=\"down-arrow\" (click)=\"onMoveDown()\">keyboard_arrow_down</mat-icon>\n  <mat-icon class=\"remove-upload\" (click)=\"onRemove()\">cancel</mat-icon>\n</mat-card>\n",
                    styles: [".file-preview{height:80px;margin:2px 0;position:relative}.mat-card{padding:0}.file-preview img{max-height:72px}.graphic{position:absolute;left:50px;height:72px;max-width:200px;text-align:center;padding-left:20px;overflow:hidden}.icon-graphic{font-size:72px;height:72px;width:72px}.mat-progress-bar{position:absolute;left:0;top:73px}.file-info{position:absolute;left:250px;height:72px;max-width:300px;text-align:center}.file-name{margin:18px 6px;display:inline-block;line-height:36px;font-size:16px;height:36px;font-weight:700;color:#000}.arrow,.down-arrow,.up-arrow{cursor:pointer}.up-arrow{position:absolute;left:6px;top:0;font-size:36px;height:36px;width:36px;font-weight:700}.down-arrow{position:absolute;left:6px;top:36px;font-size:36px;height:36px;width:36px;font-weight:700}.remove-upload{cursor:pointer;position:absolute;right:10px;top:20px;font-size:40px;height:40px;width:40px}.down-arrow:hover,.up-arrow:hover{color:#fff}"]
                },] },
    ];
    /** @nocollapse */
    FilePreviewComponent.ctorParameters = function () { return []; };
    FilePreviewComponent.propDecorators = {
        file: [{ type: Input }],
        height: [{ type: Input }],
        width: [{ type: Input }],
        ordered: [{ type: Input }],
        first: [{ type: Input }],
        last: [{ type: Input }],
        moveUp: [{ type: Output }],
        moveDown: [{ type: Output }],
        remove: [{ type: Output }]
    };
    return FilePreviewComponent;
}());
export { FilePreviewComponent };
function FilePreviewComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    FilePreviewComponent.prototype.file;
    /** @type {?} */
    FilePreviewComponent.prototype.height;
    /** @type {?} */
    FilePreviewComponent.prototype.width;
    /** @type {?} */
    FilePreviewComponent.prototype.ordered;
    /** @type {?} */
    FilePreviewComponent.prototype.first;
    /** @type {?} */
    FilePreviewComponent.prototype.last;
    /** @type {?} */
    FilePreviewComponent.prototype.moveUp;
    /** @type {?} */
    FilePreviewComponent.prototype.moveDown;
    /** @type {?} */
    FilePreviewComponent.prototype.remove;
}
