// @caiu/library/esm5/lib/store/actions.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var AppActions = /** @class */ (function () {
    function AppActions() {
    }
    /**
     * @return {?}
     */
    AppActions.noAction = /**
     * @return {?}
     */
    function () {
        return {
            type: AppActions.NO_ACTION
        };
    };
    AppActions.INIT_STORE = '[Storage] Init Store';
    AppActions.NO_ACTION = '[App] No Action';
    return AppActions;
}());
export { AppActions };
function AppActions_tsickle_Closure_declarations() {
    /** @type {?} */
    AppActions.INIT_STORE;
    /** @type {?} */
    AppActions.NO_ACTION;
}
var ConfigActions = /** @class */ (function () {
    function ConfigActions() {
    }
    /**
     * @param {?} config
     * @return {?}
     */
    ConfigActions.initialize = /**
     * @param {?} config
     * @return {?}
     */
    function (config) {
        return {
            type: ConfigActions.INITIALIZE,
            payload: config
        };
    };
    ConfigActions.INITIALIZE = '[Config] Initialize Configuration';
    return ConfigActions;
}());
export { ConfigActions };
function ConfigActions_tsickle_Closure_declarations() {
    /** @type {?} */
    ConfigActions.INITIALIZE;
}
var SidenavActions = /** @class */ (function () {
    function SidenavActions() {
    }
    /**
     * @return {?}
     */
    SidenavActions.close = /**
     * @return {?}
     */
    function () {
        return {
            type: SidenavActions.CLOSE
        };
    };
    /**
     * @return {?}
     */
    SidenavActions.open = /**
     * @return {?}
     */
    function () {
        return {
            type: SidenavActions.OPEN
        };
    };
    /**
     * @return {?}
     */
    SidenavActions.toggle = /**
     * @return {?}
     */
    function () {
        return {
            type: SidenavActions.TOGGLE
        };
    };
    SidenavActions.CLOSE = '[Sidenav] Close';
    SidenavActions.OPEN = '[Sidenav] Open';
    SidenavActions.TOGGLE = '[Sidenav] Toggle';
    return SidenavActions;
}());
export { SidenavActions };
function SidenavActions_tsickle_Closure_declarations() {
    /** @type {?} */
    SidenavActions.CLOSE;
    /** @type {?} */
    SidenavActions.OPEN;
    /** @type {?} */
    SidenavActions.TOGGLE;
}
var WindowActions = /** @class */ (function () {
    function WindowActions() {
    }
    /**
     * @param {?} payload
     * @return {?}
     */
    WindowActions.resize = /**
     * @param {?} payload
     * @return {?}
     */
    function (payload) {
        return {
            type: WindowActions.RESIZE,
            payload: payload
        };
    };
    WindowActions.RESIZE = '[Window] Resize';
    return WindowActions;
}());
export { WindowActions };
function WindowActions_tsickle_Closure_declarations() {
    /** @type {?} */
    WindowActions.RESIZE;
}
