// @caiu/library/esm5/lib/store/models.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @record
 */
export function Action() { }
function Action_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    Action.prototype.actionType;
    /** @type {?|undefined} */
    Action.prototype.payload;
}
/**
 * @record
 */
export function ActionWithKey() { }
function ActionWithKey_tsickle_Closure_declarations() {
    /** @type {?} */
    ActionWithKey.prototype.key;
}
/** @enum {number} */
var ActionTypes = {
    DEFAULT: 0,
    GET: 1,
    GET_SUCCESS: 2,
    GET_ERROR: 3,
    POST: 4,
    POST_SUCCESS: 5,
    POST_ERROR: 6,
    PUT: 7,
    PUT_SUCCESS: 8,
    PUT_ERROR: 9,
    DELETE: 10,
    DELETE_SUCCESS: 11,
    DELETE_ERROR: 12,
    SAVE: 13,
};
export { ActionTypes };
ActionTypes[ActionTypes.DEFAULT] = "DEFAULT";
ActionTypes[ActionTypes.GET] = "GET";
ActionTypes[ActionTypes.GET_SUCCESS] = "GET_SUCCESS";
ActionTypes[ActionTypes.GET_ERROR] = "GET_ERROR";
ActionTypes[ActionTypes.POST] = "POST";
ActionTypes[ActionTypes.POST_SUCCESS] = "POST_SUCCESS";
ActionTypes[ActionTypes.POST_ERROR] = "POST_ERROR";
ActionTypes[ActionTypes.PUT] = "PUT";
ActionTypes[ActionTypes.PUT_SUCCESS] = "PUT_SUCCESS";
ActionTypes[ActionTypes.PUT_ERROR] = "PUT_ERROR";
ActionTypes[ActionTypes.DELETE] = "DELETE";
ActionTypes[ActionTypes.DELETE_SUCCESS] = "DELETE_SUCCESS";
ActionTypes[ActionTypes.DELETE_ERROR] = "DELETE_ERROR";
ActionTypes[ActionTypes.SAVE] = "SAVE";
/**
 * @record
 * @template T
 */
export function ActionWithPayload() { }
function ActionWithPayload_tsickle_Closure_declarations() {
    /** @type {?} */
    ActionWithPayload.prototype.payload;
}
/**
 * @record
 */
export function Reducers() { }
function Reducers_tsickle_Closure_declarations() {
    /* TODO: handle strange member:
    [key: string]: ActionReducer<any>;
    */
}
/**
 * @record
 */
export function UnsafeAction() { }
function UnsafeAction_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    UnsafeAction.prototype.payload;
}
