// @caiu/library/esm2015/lib/components/grid/grid.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { orderBy } from '@progress/kendo-data-query';
import { Collection } from '../../shared/collection';
import { DateHelper } from '../../shared/date';
import { QueryModel, StyleModel } from '../../shared/models';
/**
 * @template T
 */
export class Grid {
    /**
     * @param {?} data
     */
    constructor(data) {
        this.data = data;
        this.query = new QueryModel();
        this.style = new StyleModel();
        this.filterable = true;
        this.sortable = { mode: 'multiple' };
    }
    /**
     * @param {?} acc
     * @param {?} operator
     * @param {?} field
     * @param {?} value
     * @return {?}
     */
    static ApplyFilter(acc, operator, field, value) {
        switch (operator) {
            case 'contains':
                return acc.filter(row => Grid.Contains(row[field], value));
            case 'gte':
                return acc.filter(row => Grid.MatchDate(row[field], value));
        }
    }
    /**
     * @template T
     * @param {?} items
     * @return {?}
     */
    static Build(items) {
        const /** @type {?} */ data = Object.assign(new Collection(), { items });
        return new Grid(data);
    }
    /**
     * @param {?} field
     * @param {?=} dir
     * @return {?}
     */
    static BuildSort(field, dir = 'asc') {
        return /** @type {?} */ ({ dir, field });
    }
    /**
     * @param {?} str
     * @param {?} substr
     * @return {?}
     */
    static Contains(str, substr) {
        if (typeof (str) === 'string' && typeof (substr) === 'string') {
            return str.toLowerCase().includes(substr.toLowerCase());
        }
        return true;
    }
    /**
     * @param {?} rows
     * @param {?} filters
     * @return {?}
     */
    static FilterRows(rows, filters) {
        return filters.reduce((acc, filter) => {
            const /** @type {?} */ operator = filter['operator'];
            const /** @type {?} */ field = filter['field'];
            const /** @type {?} */ value = filter['value'];
            return Grid.ApplyFilter(rows, operator, field, value);
        }, rows);
    }
    /**
     * @param {?} date1
     * @param {?} date2
     * @return {?}
     */
    static MatchDate(date1, date2) {
        return DateHelper.IsSameDay(date1, date2);
    }
    /**
     * @param {?} rows
     * @param {?} skip
     * @param {?} take
     * @return {?}
     */
    static PageRows(rows, skip, take) {
        return rows.filter((row, index) => index >= skip && index < skip + take);
    }
    /**
     * @param {?} rows
     * @param {?} sort
     * @return {?}
     */
    static SortRows(rows, sort) {
        return orderBy(rows, sort);
    }
    /**
     * @param {?} grid
     * @return {?}
     */
    update(grid) {
    }
}
function Grid_tsickle_Closure_declarations() {
    /** @type {?} */
    Grid.prototype.query;
    /** @type {?} */
    Grid.prototype.style;
    /** @type {?} */
    Grid.prototype.detailRowHeight;
    /** @type {?} */
    Grid.prototype.filter;
    /** @type {?} */
    Grid.prototype.filterable;
    /** @type {?} */
    Grid.prototype.groupable;
    /** @type {?} */
    Grid.prototype.height;
    /** @type {?} */
    Grid.prototype.pageSize;
    /** @type {?} */
    Grid.prototype.pageable;
    /** @type {?} */
    Grid.prototype.rowHeight;
    /** @type {?} */
    Grid.prototype.scrollable;
    /** @type {?} */
    Grid.prototype.selectable;
    /** @type {?} */
    Grid.prototype.skip;
    /** @type {?} */
    Grid.prototype.sortable;
    /** @type {?} */
    Grid.prototype.group;
    /** @type {?} */
    Grid.prototype.rowClass;
    /** @type {?} */
    Grid.prototype.sort;
    /** @type {?} */
    Grid.prototype.data;
}
