// @caiu/library/esm2015/lib/http/http.models.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @record
 */
export function HttpAction() { }
function HttpAction_tsickle_Closure_declarations() {
    /** @type {?} */
    HttpAction.prototype.type;
    /** @type {?} */
    HttpAction.prototype.payload;
}
/**
 * @record
 * @template T
 */
export function HttpActionPayload() { }
function HttpActionPayload_tsickle_Closure_declarations() {
    /** @type {?} */
    HttpActionPayload.prototype.path;
    /** @type {?|undefined} */
    HttpActionPayload.prototype.model;
    /** @type {?|undefined} */
    HttpActionPayload.prototype.options;
    /** @type {?|undefined} */
    HttpActionPayload.prototype.headers;
    /** @type {?|undefined} */
    HttpActionPayload.prototype.onSuccess;
    /** @type {?|undefined} */
    HttpActionPayload.prototype.onError;
}
/**
 * @template T
 */
export class HttpDeletePayload {
    constructor() {
        this.path = '';
        this.options = new HttpOptions();
        this.headers = undefined;
        this.onSuccess = '';
    }
}
function HttpDeletePayload_tsickle_Closure_declarations() {
    /** @type {?} */
    HttpDeletePayload.prototype.path;
    /** @type {?} */
    HttpDeletePayload.prototype.options;
    /** @type {?} */
    HttpDeletePayload.prototype.headers;
    /** @type {?} */
    HttpDeletePayload.prototype.onSuccess;
    /** @type {?} */
    HttpDeletePayload.prototype.onSuccessPayload;
    /** @type {?} */
    HttpDeletePayload.prototype.onError;
}
/**
 * @template T
 */
export class HttpGetPayload {
    constructor() {
        this.path = '';
        this.options = new HttpOptions();
        this.headers = undefined;
        this.onSuccess = '';
    }
}
function HttpGetPayload_tsickle_Closure_declarations() {
    /** @type {?} */
    HttpGetPayload.prototype.path;
    /** @type {?} */
    HttpGetPayload.prototype.options;
    /** @type {?} */
    HttpGetPayload.prototype.headers;
    /** @type {?} */
    HttpGetPayload.prototype.onSuccess;
    /** @type {?} */
    HttpGetPayload.prototype.onError;
}
/**
 * @template T
 */
export class HttpPutPayload {
    constructor() {
        this.path = '';
        this.model = null;
        this.options = new HttpOptions();
        this.headers = undefined;
        this.onSuccess = '';
    }
}
function HttpPutPayload_tsickle_Closure_declarations() {
    /** @type {?} */
    HttpPutPayload.prototype.path;
    /** @type {?} */
    HttpPutPayload.prototype.model;
    /** @type {?} */
    HttpPutPayload.prototype.options;
    /** @type {?} */
    HttpPutPayload.prototype.headers;
    /** @type {?} */
    HttpPutPayload.prototype.onSuccess;
    /** @type {?} */
    HttpPutPayload.prototype.onError;
}
/**
 * @template T
 */
export class HttpPostPayload {
    constructor() {
        this.path = '';
        this.model = null;
        this.options = new HttpOptions();
        this.headers = undefined;
        this.onSuccess = '';
    }
}
function HttpPostPayload_tsickle_Closure_declarations() {
    /** @type {?} */
    HttpPostPayload.prototype.path;
    /** @type {?} */
    HttpPostPayload.prototype.model;
    /** @type {?} */
    HttpPostPayload.prototype.options;
    /** @type {?} */
    HttpPostPayload.prototype.headers;
    /** @type {?} */
    HttpPostPayload.prototype.onSuccess;
    /** @type {?} */
    HttpPostPayload.prototype.onError;
}
/** @enum {number} */
const HTTP_REQUEST_STATUS = {
    PENDING: 0,
    ERROR: 1,
    SUCCESS: 2,
};
export { HTTP_REQUEST_STATUS };
HTTP_REQUEST_STATUS[HTTP_REQUEST_STATUS.PENDING] = "PENDING";
HTTP_REQUEST_STATUS[HTTP_REQUEST_STATUS.ERROR] = "ERROR";
HTTP_REQUEST_STATUS[HTTP_REQUEST_STATUS.SUCCESS] = "SUCCESS";
/** @enum {number} */
const HTTP_STATUS_CODES = {
    CONTINUE: 100,
    SWITCHING_PROTOCOLS: 101,
    OK: 200,
    CREATED: 201,
    ACCEPTED: 202,
    NON_AUTHORITATIVE_INFORMATION: 203,
    NO_CONTENT: 204,
    RESET_CONTENT: 205,
    PARTIAL_CONTENT: 206,
    MULTIPLE_CHOICES: 300,
    MOVED_PERMANENTLY: 301,
    FOUND: 302,
    SEE_OTHER: 303,
    NOT_MODIFIED: 304,
    USE_PROXY: 305,
    TEMPORARY_REDIRECT: 307,
    BAD_REQUEST: 400,
    UNAUTHORIZED: 401,
    PAYMENT_REQUIRED: 402,
    FORBIDDEN: 403,
    NOT_FOUND: 404,
    METHOD_NOT_ALLOWED: 405,
    NOT_ACCEPTABLE: 406,
    PROXY_AUTHENTICATION_REQUIRED: 407,
    REQUEST_TIMEOUT: 408,
    CONFLICT: 409,
    GONE: 410,
    LENGTH_REQUIRED: 411,
    PRECONDITION_FAILED: 412,
    REQUEST_ENTITY_TOO_LARGE: 413,
    REQUEST_URI_TOO_LONG: 414,
    UNSUPPORTED_MEDIA_TYPE: 415,
    REQUESTED_RANGE_NOT_SATISFIABLE: 416,
    EXPECTATION_FAILED: 417,
    UNPROCESSABLE_ENTITY: 422,
    TOO_MANY_REQUESTS: 429,
    INTERNAL_SERVER_ERROR: 500,
    NOT_IMPLEMENTED: 501,
    BAD_GATEWAY: 502,
    SERVICE_UNAVAILABLE: 503,
    GATEWAY_TIMEOUT: 504,
    HTTP_VERSION_NOT_SUPPORTED: 505,
};
export { HTTP_STATUS_CODES };
HTTP_STATUS_CODES[HTTP_STATUS_CODES.CONTINUE] = "CONTINUE";
HTTP_STATUS_CODES[HTTP_STATUS_CODES.SWITCHING_PROTOCOLS] = "SWITCHING_PROTOCOLS";
HTTP_STATUS_CODES[HTTP_STATUS_CODES.OK] = "OK";
HTTP_STATUS_CODES[HTTP_STATUS_CODES.CREATED] = "CREATED";
HTTP_STATUS_CODES[HTTP_STATUS_CODES.ACCEPTED] = "ACCEPTED";
HTTP_STATUS_CODES[HTTP_STATUS_CODES.NON_AUTHORITATIVE_INFORMATION] = "NON_AUTHORITATIVE_INFORMATION";
HTTP_STATUS_CODES[HTTP_STATUS_CODES.NO_CONTENT] = "NO_CONTENT";
HTTP_STATUS_CODES[HTTP_STATUS_CODES.RESET_CONTENT] = "RESET_CONTENT";
HTTP_STATUS_CODES[HTTP_STATUS_CODES.PARTIAL_CONTENT] = "PARTIAL_CONTENT";
HTTP_STATUS_CODES[HTTP_STATUS_CODES.MULTIPLE_CHOICES] = "MULTIPLE_CHOICES";
HTTP_STATUS_CODES[HTTP_STATUS_CODES.MOVED_PERMANENTLY] = "MOVED_PERMANENTLY";
HTTP_STATUS_CODES[HTTP_STATUS_CODES.FOUND] = "FOUND";
HTTP_STATUS_CODES[HTTP_STATUS_CODES.SEE_OTHER] = "SEE_OTHER";
HTTP_STATUS_CODES[HTTP_STATUS_CODES.NOT_MODIFIED] = "NOT_MODIFIED";
HTTP_STATUS_CODES[HTTP_STATUS_CODES.USE_PROXY] = "USE_PROXY";
HTTP_STATUS_CODES[HTTP_STATUS_CODES.TEMPORARY_REDIRECT] = "TEMPORARY_REDIRECT";
HTTP_STATUS_CODES[HTTP_STATUS_CODES.BAD_REQUEST] = "BAD_REQUEST";
HTTP_STATUS_CODES[HTTP_STATUS_CODES.UNAUTHORIZED] = "UNAUTHORIZED";
HTTP_STATUS_CODES[HTTP_STATUS_CODES.PAYMENT_REQUIRED] = "PAYMENT_REQUIRED";
HTTP_STATUS_CODES[HTTP_STATUS_CODES.FORBIDDEN] = "FORBIDDEN";
HTTP_STATUS_CODES[HTTP_STATUS_CODES.NOT_FOUND] = "NOT_FOUND";
HTTP_STATUS_CODES[HTTP_STATUS_CODES.METHOD_NOT_ALLOWED] = "METHOD_NOT_ALLOWED";
HTTP_STATUS_CODES[HTTP_STATUS_CODES.NOT_ACCEPTABLE] = "NOT_ACCEPTABLE";
HTTP_STATUS_CODES[HTTP_STATUS_CODES.PROXY_AUTHENTICATION_REQUIRED] = "PROXY_AUTHENTICATION_REQUIRED";
HTTP_STATUS_CODES[HTTP_STATUS_CODES.REQUEST_TIMEOUT] = "REQUEST_TIMEOUT";
HTTP_STATUS_CODES[HTTP_STATUS_CODES.CONFLICT] = "CONFLICT";
HTTP_STATUS_CODES[HTTP_STATUS_CODES.GONE] = "GONE";
HTTP_STATUS_CODES[HTTP_STATUS_CODES.LENGTH_REQUIRED] = "LENGTH_REQUIRED";
HTTP_STATUS_CODES[HTTP_STATUS_CODES.PRECONDITION_FAILED] = "PRECONDITION_FAILED";
HTTP_STATUS_CODES[HTTP_STATUS_CODES.REQUEST_ENTITY_TOO_LARGE] = "REQUEST_ENTITY_TOO_LARGE";
HTTP_STATUS_CODES[HTTP_STATUS_CODES.REQUEST_URI_TOO_LONG] = "REQUEST_URI_TOO_LONG";
HTTP_STATUS_CODES[HTTP_STATUS_CODES.UNSUPPORTED_MEDIA_TYPE] = "UNSUPPORTED_MEDIA_TYPE";
HTTP_STATUS_CODES[HTTP_STATUS_CODES.REQUESTED_RANGE_NOT_SATISFIABLE] = "REQUESTED_RANGE_NOT_SATISFIABLE";
HTTP_STATUS_CODES[HTTP_STATUS_CODES.EXPECTATION_FAILED] = "EXPECTATION_FAILED";
HTTP_STATUS_CODES[HTTP_STATUS_CODES.UNPROCESSABLE_ENTITY] = "UNPROCESSABLE_ENTITY";
HTTP_STATUS_CODES[HTTP_STATUS_CODES.TOO_MANY_REQUESTS] = "TOO_MANY_REQUESTS";
HTTP_STATUS_CODES[HTTP_STATUS_CODES.INTERNAL_SERVER_ERROR] = "INTERNAL_SERVER_ERROR";
HTTP_STATUS_CODES[HTTP_STATUS_CODES.NOT_IMPLEMENTED] = "NOT_IMPLEMENTED";
HTTP_STATUS_CODES[HTTP_STATUS_CODES.BAD_GATEWAY] = "BAD_GATEWAY";
HTTP_STATUS_CODES[HTTP_STATUS_CODES.SERVICE_UNAVAILABLE] = "SERVICE_UNAVAILABLE";
HTTP_STATUS_CODES[HTTP_STATUS_CODES.GATEWAY_TIMEOUT] = "GATEWAY_TIMEOUT";
HTTP_STATUS_CODES[HTTP_STATUS_CODES.HTTP_VERSION_NOT_SUPPORTED] = "HTTP_VERSION_NOT_SUPPORTED";
export class HttpOptions {
    constructor() {
        this.prependBaseUrl = true;
    }
}
function HttpOptions_tsickle_Closure_declarations() {
    /** @type {?} */
    HttpOptions.prototype.prependBaseUrl;
}
