// @caiu/library/esm5/lib/errors/errors.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { NgModule, ErrorHandler } from '@angular/core';
import { ErrorEffects } from './errors.effects';
import { ErrorsService } from './errors.service';
import { GlobalErrorsService } from './global-errors.service';
var ErrorsModule = /** @class */ (function () {
    function ErrorsModule() {
    }
    /**
     * @return {?}
     */
    ErrorsModule.forRoot = /**
     * @return {?}
     */
    function () {
        return {
            ngModule: ErrorsModule,
            providers: [
                ErrorsService,
                ErrorEffects,
                {
                    provide: ErrorHandler,
                    useClass: GlobalErrorsService
                }
            ]
        };
    };
    ErrorsModule.decorators = [
        { type: NgModule, args: [{
                    imports: [],
                    declarations: []
                },] },
    ];
    return ErrorsModule;
}());
export { ErrorsModule };
