// @caiu/library/esm2015/lib/events/events.actions.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Event, UpdateValuePayload } from './events.models';
import { build, guid } from '../shared/utils';
export class EventActions {
    /**
     * @param {?} onAction
     * @param {?=} handler
     * @param {?=} value
     * @return {?}
     */
    static addEvent(onAction, handler, value = null) {
        const /** @type {?} */ e = build(Event, {
            onAction,
            handler,
            value,
            key: guid()
        });
        return {
            type: EventActions.ADD_EVENT,
            payload: e,
            key: e.key
        };
    }
    /**
     * @param {?} key
     * @return {?}
     */
    static removeEvent(key) {
        return {
            type: EventActions.REMOVE_EVENT,
            payload: key
        };
    }
    /**
     * @param {?} action
     * @return {?}
     */
    static triggerEvent(action) {
        return {
            type: EventActions.TRIGGER_EVENT,
            payload: action
        };
    }
    /**
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    static updateValue(key, value) {
        const /** @type {?} */ payload = build(UpdateValuePayload, {
            key,
            value
        });
        return {
            type: EventActions.UPDATE_VALUE,
            payload: payload
        };
    }
}
EventActions.ADD_EVENT = '[Events] Add Event';
EventActions.REMOVE_EVENT = '[Events] Remove Event';
EventActions.TRIGGER_EVENT = '[Events] Trigger Event';
EventActions.UPDATE_VALUE = '[Events] Update Value';
function EventActions_tsickle_Closure_declarations() {
    /** @type {?} */
    EventActions.ADD_EVENT;
    /** @type {?} */
    EventActions.REMOVE_EVENT;
    /** @type {?} */
    EventActions.TRIGGER_EVENT;
    /** @type {?} */
    EventActions.UPDATE_VALUE;
}
