// @caiu/library/esm2015/lib/events/events.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Store } from '@ngrx/store';
import { of } from 'rxjs';
import { combineLatest, skip, take } from 'rxjs/operators';
import { EventActions } from './events.actions';
import { eventValueSelector } from './events.selectors';
import * as i0 from "@angular/core";
import * as i1 from "@ngrx/store";
export class EventsService {
    /**
     * @param {?} store
     */
    constructor(store) {
        this.store = store;
        this.eventIndex = 0;
        this.events = [];
    }
    /**
     * @return {?}
     */
    get eventKey() {
        this.eventIndex++;
        return `event-${this.eventIndex}`;
    }
    /**
     * @param {?} actionType
     * @param {?=} handler
     * @return {?}
     */
    addEvent(actionType, handler) {
        const /** @type {?} */ f = handler ? handler : (action) => true;
        const /** @type {?} */ actionWithKey = EventActions.addEvent(actionType, f, false);
        this.store.dispatch(actionWithKey);
        return actionWithKey.key;
    }
    /**
     * @param {?} action
     * @param {?=} handler
     * @return {?}
     */
    dispatch(action, handler) {
        return this.dispatchHttpAction(action, handler);
    }
    /**
     * @param {?} action
     * @param {?=} handler
     * @return {?}
     */
    dispatchHttpAction(action, handler) {
        const /** @type {?} */ successAction = action.payload.onSuccess;
        const /** @type {?} */ errorAction = typeof (action.payload.onError) === 'string' ? action.payload.onError : null;
        const /** @type {?} */ success = this.takeOne(successAction, handler);
        const /** @type {?} */ obs = errorAction ? success.pipe(combineLatest(this.takeOne(errorAction, handler), (x, y) => true)) : success;
        this.store.dispatch(action);
        return obs;
    }
    /**
     * @param {?} eventKey
     * @return {?}
     */
    removeEvent(eventKey) {
        this.store.dispatch(EventActions.removeEvent(eventKey));
    }
    /**
     * @param {?} action
     * @param {?} handler
     * @return {?}
     */
    subscribe(action, handler) {
        return this.take(1, action).subscribe(handler);
    }
    /**
     * @param {?=} n
     * @param {?=} actionType
     * @param {?=} handler
     * @return {?}
     */
    take(n = 1, actionType, handler) {
        const /** @type {?} */ key = this.addEvent(actionType, handler);
        return eventValueSelector(this.store, key).pipe(skip(1), take(n));
    }
    /**
     * @param {?=} n
     * @param {?=} actions
     * @param {?=} handler
     * @return {?}
     */
    takeFrom(n = 1, actions, handler) {
        const /** @type {?} */ keys = actions.map(action => this.addEvent(action, handler));
        return keys.map(key => eventValueSelector(this.store, key))
            .reduce((acc, obs) => {
            return acc.pipe(combineLatest(obs, (x, y) => true));
        }, of(false)).pipe(skip(1), take(n));
    }
    /**
     * @param {?} actionType
     * @param {?=} handler
     * @return {?}
     */
    takeOne(actionType, handler) {
        return this.take(1, actionType, handler);
    }
}
EventsService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] },
];
/** @nocollapse */
EventsService.ctorParameters = () => [
    { type: Store }
];
/** @nocollapse */ EventsService.ngInjectableDef = i0.defineInjectable({ factory: function EventsService_Factory() { return new EventsService(i0.inject(i1.Store)); }, token: EventsService, providedIn: "root" });
function EventsService_tsickle_Closure_declarations() {
    /** @type {?} */
    EventsService.prototype.eventIndex;
    /** @type {?} */
    EventsService.prototype.events;
    /** @type {?} */
    EventsService.prototype.store;
}
