// @caiu/library/esm2015/lib/events/events.models.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Collection } from '../shared/collection';
import { build } from '../shared/utils';
export class Event {
    constructor() {
        this.emittedCount = 0;
        this.key = '';
        this.onAction = '';
        this.value = null;
    }
}
function Event_tsickle_Closure_declarations() {
    /** @type {?} */
    Event.prototype.emittedCount;
    /** @type {?} */
    Event.prototype.handler;
    /** @type {?} */
    Event.prototype.key;
    /** @type {?} */
    Event.prototype.onAction;
    /** @type {?} */
    Event.prototype.value;
}
export class Events extends Collection {
    /**
     * @return {?}
     */
    get actions() {
        return this.toArray().reduce((acc, event) => {
            const /** @type {?} */ addValues = !event.onAction ?
                [] : typeof event.onAction === 'string' ?
                acc.indexOf(event.onAction) === -1 ? [event.onAction] : []
                : event.onAction.filter(x => acc.indexOf(x) === -1);
            return [...acc, ...addValues];
        }, []);
    }
    /**
     * @return {?}
     */
    get eventsByAction() {
        return this.toArray().reduce((acc, event) => {
            if (typeof event.onAction === 'string') {
                const /** @type {?} */ val = acc[event.onAction] || [];
                acc[event.onAction] = [...val, event];
            }
            else {
                event.onAction.forEach(actionType => {
                    const /** @type {?} */ val = acc[actionType] || [];
                    acc[actionType] = [...val, event];
                });
            }
            return acc;
        }, {});
    }
    /**
     * @param {?} payload
     * @return {?}
     */
    addEvent(payload) {
        return build(Events, this.addItem(payload, payload.key));
    }
    /**
     * @param {?} actionType
     * @return {?}
     */
    getActionEvents(actionType) {
        return this.eventsByAction[actionType];
    }
    /**
     * @param {?} payload
     * @return {?}
     */
    removeEvent(payload) {
        return build(Events, this.removeItem(payload));
    }
    /**
     * @param {?} payload
     * @return {?}
     */
    replaceEvents(payload) {
        const /** @type {?} */ events = this.copyItems();
        payload.forEach(e => {
            events[e.key] = e;
        });
        return build(Events, this, { items: events });
    }
    /**
     * @param {?} payload
     * @return {?}
     */
    triggerEvent(payload) {
        const /** @type {?} */ events = this.getActionEvents(payload.type)
            .map(e => build(Event, e, { value: e.handler ? e.handler(payload) : payload }));
        return this.replaceEvents(events);
    }
    /**
     * @param {?} payload
     * @return {?}
     */
    updateValue(payload) {
        const /** @type {?} */ event = build(Event, this.items[payload.key], { value: payload.value });
        return build(Events, this.updateItem(event, payload.key, Event));
    }
}
export class UpdateValuePayload {
}
function UpdateValuePayload_tsickle_Closure_declarations() {
    /** @type {?} */
    UpdateValuePayload.prototype.key;
    /** @type {?} */
    UpdateValuePayload.prototype.value;
}
