// @caiu/library/esm2015/lib/events/events.effects.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { Actions, Effect } from '@ngrx/effects';
import { Store } from '@ngrx/store';
import { Observable } from 'rxjs';
import { filter, map } from 'rxjs/operators';
import { EventActions } from './events.actions';
import { Events } from './events.models';
export class EventEffects {
    /**
     * @param {?} actions$
     * @param {?} store
     */
    constructor(actions$, store) {
        this.actions$ = actions$;
        this.store = store;
        this.events = new Events();
        /**
         * Trigger event if action is a registered trigger.
         */
        this.onTriggerEvent = this.actions$.pipe(filter(action => this.events.actions.indexOf(action.type) !== -1), map(this.trigger));
        this.store.select(s => s['events'])
            .subscribe(events => {
            this.events = events;
        });
    }
    /**
     * @param {?} action
     * @return {?}
     */
    trigger(action) {
        return EventActions.triggerEvent(action);
    }
}
EventEffects.decorators = [
    { type: Injectable },
];
/** @nocollapse */
EventEffects.ctorParameters = () => [
    { type: Actions },
    { type: Store }
];
tslib_1.__decorate([
    Effect(),
    tslib_1.__metadata("design:type", Observable)
], EventEffects.prototype, "onTriggerEvent", void 0);
function EventEffects_tsickle_Closure_declarations() {
    /** @type {?} */
    EventEffects.prototype.events;
    /**
     * Trigger event if action is a registered trigger.
     * @type {?}
     */
    EventEffects.prototype.onTriggerEvent;
    /** @type {?} */
    EventEffects.prototype.actions$;
    /** @type {?} */
    EventEffects.prototype.store;
}
